// @vtb/ui-kit3/src/components/TabMenu/style.ts
import styled from 'styled-components';

import { typography } from '../Typography';

import type { Dimension } from './constants';
import {
  BADGE_MARGIN,
  ICON_MARGIN,
  ICON_SIZE_L,
  ICON_SIZE_M,
  LINE_HEIGHT,
  TAB_HEIGHT_L,
  TAB_HEIGHT_M,
  TAB_PADDING_L,
  TAB_PADDING_M,
} from './constants';

export const Wrapper = styled.div<{ underline?: boolean; mobile?: boolean }>`
  position: relative;
  display: flex;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  box-shadow: inset 0 -${LINE_HEIGHT} 0 0 ${({ theme, underline }) => (underline ? theme.color.background.tertiary : 'transparent')};
  overflow-x: ${({ mobile }) => (mobile ? 'scroll' : 'visible')};

  &::-webkit-scrollbar {
    width: 0 !important;
    height: 0 !important;
  }

  overflow: -moz-scrollbars-none;
  -ms-overflow-style: none;
  scrollbar-width: none;
  scrollbar-height: none;
`;

export const TabsWrapper = styled.div`
  display: flex;
`;

export const Underline = styled.div<{
  left: number;
  width: number;
}>`
  position: absolute;
  bottom: 0;
  ${({ theme, left, width }) => `
    display: ${width ? 'block' : 'none'};
    left: ${left}px;
    width: ${width}px;
    background-color: ${theme.color.basic.primary};
  `}
  height: ${LINE_HEIGHT};
  transition: all 0.2s ease-out;
`;

export const TabContent = styled.div`
  height: 100%;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
`;

export const TabContentWrapper = styled.div<{ dimension: Dimension }>`
  display: inline-flex;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: ${({ dimension }) => (dimension === 'l' ? TAB_PADDING_L : TAB_PADDING_M)};
`;

export const Tab = styled.button<{ dimension: Dimension }>`
  position: relative;
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  height: ${({ dimension }) => (dimension === 'm' ? TAB_HEIGHT_M : TAB_HEIGHT_L)}px;
  padding: 0;
  background: transparent;
  appearance: none;
  border: none;
  color: ${({ theme }) => theme.color.text.primary};
  ${({ dimension }) => (dimension === 'm' ? typography['Additional/S'] : typography['Additional/L'])}
  user-select: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;

  &:focus {
    &:before {
      position: absolute;
      content: '';
      border: ${LINE_HEIGHT} solid ${({ theme }) => theme.color.basic.hover};
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
    }
  }
  &:focus,
  & > ${TabContentWrapper} {
    outline: none;
  }

  & svg {
    display: inline-block;
    flex: 1 0 auto;
    fill: ${({ theme }) => theme.color.basic.tertiary};
    width: ${({ dimension }) => (dimension === 'm' ? ICON_SIZE_M : ICON_SIZE_L)}px;
    height: ${({ dimension }) => (dimension === 'm' ? ICON_SIZE_M : ICON_SIZE_L)}px;
    margin-right: ${ICON_MARGIN};
  }
  & [data-badge] {
    margin-left: ${BADGE_MARGIN};
  }

  &:hover:not(:disabled) {
    color: ${({ theme }) => theme.color.basic.hover};
    & svg {
      fill: ${({ theme }) => theme.color.basic.hover};
    }
    [data-badge] {
      background: ${({ theme }) => theme.color.basic.hover};
      color: ${({ theme }) => theme.color.text.staticWhite};
    }
  }
  &:active:not(:disabled) {
    color: ${({ theme }) => theme.color.basic.press};
    & svg {
      fill: ${({ theme }) => theme.color.basic.press};
    }
    [data-badge] {
      background: ${({ theme }) => theme.color.basic.press};
      color: ${({ theme }) => theme.color.text.staticWhite};
    }
  }
  &:disabled {
    color: ${({ theme }) => theme.color.text.tertiary};
    cursor: default;
    & svg {
      fill: ${({ theme }) => theme.color.text.tertiary};
    }
  }
`;
