// @vtb/ui-kit3/src/components/input/TextInput/TextInput.stories.tsx
import React, { ChangeEvent, useState } from 'react';
import { Meta, Story } from '@storybook/react';
import { TextInput } from './index';
import { withDesign } from 'storybook-addon-designs';
import { INPUT_DIMENSIONS_VALUES } from '~/components/input/inputTypes';

export default {
  title: 'Input/TextInput',
  component: TextInput,
  decorators: [withDesign],
  parameters: {
    design: {
      type: 'figma',
      url:
        'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=65740%3A76866',
    },
  },
  argTypes: {
    dimension: {
      options: INPUT_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },

    disabled: {
      type: 'boolean',
    },

    readOnly: {
      options: [undefined, true],
      control: { type: 'radio' },
    },
    disableCopying: {
      options: [undefined, true],
      control: { type: 'radio' },
    },
    placeholder: {
      type: 'string',
    },
    onChange: {
      action: 'onChange',
    },
  },
} as Meta;

const Template: Story<React.ComponentProps<typeof TextInput>> = (props) => {
  const cleanProps = (Object.keys(props) as Array<keyof typeof props>).reduce((acc, key) => {
    if (props[key] !== undefined) acc[key] = props[key];

    return acc;
  }, {} as Record<any, any>);

  const [localValue, setValue] = useState<string>(String(props.value) ?? '');

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.currentTarget.value;
    setValue(inputValue);
    props.onChange?.(e);
  };
  return <TextInput {...cleanProps} value={localValue} onChange={handleChange} />;
};

export const TextInputStory = Template.bind({});
TextInputStory.args = {
  value: 'Привет!',
  placeholder: 'Placeholder',
};
TextInputStory.storyName = 'Базовый input компонент';
