// @vtb/ui-kit3/src/components/Hint/styled/HintWrapper.ts
import styled, { css } from 'styled-components';

import { TYPOGRAPHY } from '../../Typography';

export type Dimension = 's' | 'm' | 'l';

const HINT_PADDING = '16px';
const HINT_BORDER_RADIUS = '4px';
const HINT_WIDTH_S = '280px';
const HINT_WIDTH_M = '384px';
const HINT_WIDTH_L = '488px';
const HINT_FONT_SIZE = '14px';
const HINT_LINE_HEIGHT = '20px';
const HINT_FONT_WEIGHT = 400;

const sizes = css<{ dimension: Dimension; isMobile: boolean }>`
  width: ${({ dimension }) => {
    if (dimension === 's') return HINT_WIDTH_S;
    if (dimension === 'm') return HINT_WIDTH_M;
    return HINT_WIDTH_L;
  }};
`;

export const HintWrapper = styled.div<{
  dimension: Dimension;
  isMobile: boolean;
}>`
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  justify-content: space-between;
  z-index: ${({ theme }) => theme.zIndex.hint};
  ${({ isMobile }) => !isMobile && sizes}
  ${({ theme }) => `
    background-color: ${theme.color.background.primary};
    color: ${theme.color.text.primary};
  `}
  font-family: ${TYPOGRAPHY.fontFamily};
  font-size: ${HINT_FONT_SIZE};
  font-weight: ${HINT_FONT_WEIGHT};
  line-height: ${HINT_LINE_HEIGHT};
  border-radius: ${HINT_BORDER_RADIUS};
  box-shadow: 0 4px 8px rgba(138, 150, 168, 0.4), 0 -2px 4px rgba(138, 150, 168, 0.08);
  box-sizing: border-box;
`;

export const HintDialog = styled.div`
  display: flex;
  padding: ${HINT_PADDING};
`;
