// @vtb/ui-kit3/src/components/RadioButton/RadioButton.stories.tsx
import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import styled from 'styled-components';

import { RadioButton, RadioButtonProps } from '../RadioButton';
import { Button } from '../Button';

const Separator = styled.div`
  height: 20px;
  width: 100%;
`;

const Desc = styled.div`
  font-family: 'VTB Group UI';
  font-size: 16px;
  line-height: 24px;
`;

const Description = () => (
  <Desc>
    Радиальные кнопки применяются, когда есть список опций, из которых пользователь может выбрать только один вариант
  </Desc>
);

export default {
  title: 'Example/RadioButton',
  decorators: [withDesign],
  component: RadioButton,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=64810%3A2',
    },
    componentSubtitle: <Description />,
    actions: {
      argTypesRegex: 'onChange',
    },
    layout: 'centered',
  },
  argTypes: {
    dimension: {
      options: ['m', 's'],
      control: { type: 'radio' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    checked: {
      control: { type: 'boolean' },
    },
  },
} as Meta;

const Template1: Story<RadioButtonProps> = () => {
  return (
    <>
      <RadioButton value={1}>Dimension - m</RadioButton>
      <Separator />
      <RadioButton value={1} dimension="s">
        Dimension - s
      </RadioButton>
    </>
  );
};

const Template2: Story<RadioButtonProps> = () => {
  return (
    <>
      <RadioButton value={1} checked>
        Checked
      </RadioButton>
      <Separator />
      <RadioButton value={1} checked={false}>
        Not checked
      </RadioButton>
      <Separator />
      <RadioButton value={1} checked disabled>
        Checked disabled
      </RadioButton>
      <Separator />
      <RadioButton value={1} checked={false} disabled>
        Not checked disabled
      </RadioButton>
    </>
  );
};

const Template3: Story<RadioButtonProps> = ({ dimension, disabled, checked: argChecked }: RadioButtonProps) => {
  const [checked, setChecked] = useState(argChecked);
  return (
    <>
      <RadioButton
        value={1}
        checked={checked}
        onChange={(event) => setChecked(event.currentTarget.checked)}
        dimension={dimension}
        disabled={disabled}
      >
        Some text
      </RadioButton>
      <Separator />
      <Button onClick={() => setChecked(false)}>Сбросить состояние радиокнопки</Button>
    </>
  );
};

const Template4: Story<RadioButtonProps> = () => {
  return (
    <>
      <RadioButton value={1} extraText="Add text">
        Dimension - m
      </RadioButton>
      <Separator />
      <RadioButton value={1} dimension="s" extraText="Add text">
        Dimension - s
      </RadioButton>
    </>
  );
};

export const Playground = Template3.bind({});
Playground.args = {};
Playground.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};

export const RadioSize = Template1.bind({});
RadioSize.args = {};
RadioSize.storyName = 'RadioButton. Размеры';

export const RadioState = Template2.bind({});
RadioState.args = {};
RadioState.storyName = 'RadioButton. Состояния';

export const RadioExtraText = Template4.bind({});
RadioExtraText.args = {};
RadioExtraText.storyName = 'RadioButton. Дополнительный текст';
