// @vtb-ermo/router/src/routes/upzp/routes.ts
export const upzpRoutes = {
    UPZP: {
        params: {
            relativePath: 'upzp',
        },
        MANAGE_TASKS: {
            params: {
                relativePath: 'manage-task',
            },
            ALL_TASKS: {
                params: {
                    relativePath: 'all-tasks',
                },
            },
            COMPLETED_TASKS: {
                params: {
                    relativePath: 'completed-tasks',
                },
            },
            HANDBOOK: {
                params: {
                    relativePath: 'handbook',
                },
            },
        },
        WORKLOAD_AND_CAPACITY: {
            params: {
                relativePath: 'workload-and-capacity',
            },
        },
        MY_TEAM: {
            params: {
                relativePath: 'my-team',
            },
        },
        TASKS_PROCESSING: {
            params: {
                relativePath: 'tasks-processing',
            },
            CLIENTS_CARD: {
                params: {
                    relativePath: 'clients-card',
                }
            }
        },
    },
};
