// @vtb/ui-kit3/src/components/RadioButton/styled/Label.ts
import styled from 'styled-components';

import { typography } from '../../Typography';
import type { Dimension } from '../constants';
import { TEXT_PADDING_M, TEXT_PADDING_S } from '../constants';

export const Label = styled.span<{
  dimension: Dimension;
  disabled?: boolean;
}>`
  display: inline-flex;
  flex-direction: column;
  box-sizing: content-box;
  ${({ dimension }) => (dimension === 's' ? typography['Additional/XS'] : typography['Additional/M'])}
  padding-left: ${({ dimension }) => (dimension === 's' ? TEXT_PADDING_S : TEXT_PADDING_M)};

  fieldset[data-dimension='s'] & {
    ${typography['Additional/XS']}
    padding-left: ${TEXT_PADDING_S};
  }

  color: ${({ disabled, theme }) => (disabled ? theme.color.text.tertiary : theme.color.basic.secondary)};
  fieldset:disabled & {
    color: ${({ theme }) => theme.color.text.tertiary};
  }
  margin: auto 0;
  user-select: none;
`;

export const Hint = styled.span<{ dimension: Dimension; disabled?: boolean }>`
  margin-top: 6px;
  ${({ dimension }) => (dimension === 's' ? typography['Caption/XS'] : typography['Additional/XS'])}

  fieldset[data-dimension='s'] & {
    ${typography['Caption/XS']}
  }

  color: ${({ disabled, theme }) => (disabled ? theme.color.text.tertiary : theme.color.text.secondary)};
  fieldset:disabled & {
    color: ${({ theme }) => theme.color.text.tertiary};
  }
`;
