// @vtb/ui-kit3/src/components/RadioButton/styled/Span.ts
import styled from 'styled-components';

import type { Dimension } from '../constants';
import { BORDER_WIDTH_DEFAULT, DIMENSION_M, DIMENSION_S } from '../constants';

export const Span = styled.span<{ dimension: Dimension; disabled?: boolean }>`
  display: inline-block;
  position: absolute;
  margin: 0;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  content: '';
  box-sizing: border-box;
  ${({ theme, dimension }) => `
    background-color: ${theme.color.background.primary};
    width: ${dimension === 's' ? DIMENSION_S : DIMENSION_M};
    height: ${dimension === 's' ? DIMENSION_S : DIMENSION_M};
  `}
  fieldset[data-dimension='s'] & {
    width: ${DIMENSION_S};
    height: ${DIMENSION_S};
  }
  border: ${BORDER_WIDTH_DEFAULT} solid
    ${({ theme, disabled }) => (disabled ? theme.color.text.tertiary : theme.color.basic.tertiary)};
  fieldset:disabled & {
    border: ${BORDER_WIDTH_DEFAULT} solid ${({ theme }) => theme.color.text.tertiary};
  }
  border-radius: 50%;
  transition: all 0.25s ease-in-out;
`;
