// @vtb/ui-kit3/dist/ui-kit3.cjs.js
"use strict";var e=Object.defineProperty,t=Object.defineProperties,r=Object.getOwnPropertyDescriptors,n=Object.getOwnPropertySymbols,o=Object.prototype.hasOwnProperty,i=Object.prototype.propertyIsEnumerable,l=(t,r,n)=>r in t?e(t,r,{enumerable:!0,configurable:!0,writable:!0,value:n}):t[r]=n,a=(e,t)=>{for(var r in t||(t={}))o.call(t,r)&&l(e,r,t[r]);if(n)for(var r of n(t))i.call(t,r)&&l(e,r,t[r]);return e},s=(e,n)=>t(e,r(n)),c=(e,t)=>{var r={};for(var l in e)o.call(e,l)&&t.indexOf(l)<0&&(r[l]=e[l]);if(null!=e&&n)for(var l of n(e))t.indexOf(l)<0&&i.call(e,l)&&(r[l]=e[l]);return r};Object.defineProperty(exports,"__esModule",{value:!0}),exports[Symbol.toStringTag]="Module";var d=require("react"),u=require("styled-components"),p=require("react-dom"),f=require("@openvtb/fonts/VTBGroupUI-Medium/VTBGroupUI-Medium.eot"),h=require("@openvtb/fonts/VTBGroupUI-Medium/VTBGroupUI-Medium.woff2"),m=require("@openvtb/fonts/VTBGroupUI-Medium/VTBGroupUI-Medium.woff"),b=require("@openvtb/fonts/VTBGroupUI-Medium/VTBGroupUI-Medium.ttf"),g=require("@openvtb/fonts/VTBGroupUI-Medium/VTBGroupUI-Medium.svg"),x=require("@openvtb/fonts/VTBGroupUI-Regular/VTBGroupUI-Regular.eot"),v=require("@openvtb/fonts/VTBGroupUI-Regular/VTBGroupUI-Regular.woff2"),y=require("@openvtb/fonts/VTBGroupUI-Regular/VTBGroupUI-Regular.woff"),$=require("@openvtb/fonts/VTBGroupUI-Regular/VTBGroupUI-Regular.ttf"),w=require("@openvtb/fonts/VTBGroupUI-Regular/VTBGroupUI-Regular.svg");function E(e){return e&&"object"==typeof e&&"default"in e?e:{default:e}}function k(e){if(e&&e.__esModule)return e;var t={__proto__:null,[Symbol.toStringTag]:"Module"};return e&&Object.keys(e).forEach((function(r){if("default"!==r){var n=Object.getOwnPropertyDescriptor(e,r);Object.defineProperty(t,r,n.get?n:{enumerable:!0,get:function(){return e[r]}})}})),t.default=e,Object.freeze(t)}var C=E(d),D=k(d),S=E(u),F=E(p),M=E(f),_=E(h),R=E(m),A=E(b),O=E(g),B=E(x),T=E(v),z=E(y),I=E($),L=E(w);const P={spinner:100,tooltip:100,notification:99,hint:98,modal:97,dropdown:96},N={color:{basic:{primary:"#0D69F2",secondary:"#2B313B",tertiary:"#8A96A8",hover:"#0A54C2",press:"#083F91",disable:"#C4CAD4",primaryLight:"#6EA5F7"},background:{primary:"#FFFFFF",secondary:"#F1F2F4",tertiary:"#E2E5E9",tertiaryHover:"#D3D8DE",card:"#FFFFFF",cardHover:"#FFFFFF",surfaceDanger:"#FEF0F1",surfaceWarn:"#FFF7F0",surfaceSuccess:"#F2FCF5",inversion:"#2B313B",shadow:"#8A96A8"},text:{primary:"#2B313B",secondary:"#626F84",tertiary:"#A7B0BE",inversion:"#FFFFFF",staticWhite:"#FFFFFF",staticBlack:"#2B313B"},status:{warn:"#F59300",hoverWarn:"#E57300",pressWarn:"#CC6600",success:"#05A13A",hoverSuccess:"#219246",pressSuccess:"#186832",danger:"#DA0B20",hoverDanger:"#A90919",pressDanger:"#790612"},special:{blue:"#3F7DFE",softBlue:"#E6EEFF",lightBlue:"#0ABBFB",softLightBlue:"#E6F8FF",turquoise:"#4AD2CA",softTurquoise:"#EAFAF9",green:"#5ECF70",softGreen:"#EBF9EE",yellow:"#FFAF37",softYellow:"#FFF5E5",red:"#FF525C",softRed:"#FFE5E7",magenta:"#E052BD",softMagenta:"#FBE9F7",purple:"#BF67C7",softPurple:"#F7ECF8",violet:"#8742F0",softViolet:"#F0E7FD"}},shadow:{ClickableActive:"\n  box-shadow: 0 2px 4px rgba(138, 150, 168, 0.4), 0px -2px 2px rgba(138, 150, 168, 0.08);\n",ClickableDefault:"\n  box-shadow: 0 4px 8px rgba(138, 150, 168, 0.4), 0px -2px 4px rgba(138, 150, 168, 0.08);\n",ClickableHover:"\n  box-shadow: 0 8px 16px rgba(138, 150, 168, 0.4), 0px -2px 8px rgba(138, 150, 168, 0.08);\n",NonClickable:"\n  box-shadow: 0 4px 16px rgba(138, 150, 168, 0.4), 0 0 6px rgba(138, 150, 168, 0.08);\n"},zIndex:P,name:"light"},V={color:{basic:{primary:"#4789EB",secondary:"#E2E5E9",tertiary:"#576375",hover:"#1A6BE5",press:"#1455B8",disable:"#414A58",primaryLight:"#75A6F0"},background:{primary:"#1A1E23",secondary:"#252A32",tertiary:"#2B313B",tertiaryHover:"#323943",card:"#21252C",cardHover:"#2B313B",surfaceDanger:"#252A32",surfaceWarn:"#252A32",surfaceSuccess:"#252A32",inversion:"#E2E5E9",shadow:"#000000"},text:{primary:"#E2E5E9",secondary:"#8A96A8",tertiary:"#4C5767",inversion:"#1A1E23",staticWhite:"#FFFFFF",staticBlack:"#2B313B"},status:{warn:"#FF9F19",hoverWarn:"#E58600",pressWarn:"#CC7700",success:"#4AB56E",hoverSuccess:"#339956",pressSuccess:"#267340",danger:"#E15664",hoverDanger:"#CC3342",pressDanger:"#A32935"},special:{blue:"#3F7DFE",softBlue:"#CCDCFF",lightBlue:"#0ABBFB",softLightBlue:"#CCF1FF",turquoise:"#4AD2CA",softTurquoise:"#D6F5F3",green:"#5ECF70",softGreen:"#D8F3DC",yellow:"#FFAF37",softYellow:"#FFEBCC",red:"#FF525C",softRed:"#FFCCCF",magenta:"#E052BD",softMagenta:"#F7D4EF",purple:"#BF67C7",softPurple:"#EFDAF1",violet:"#8742F0",softViolet:"#E1D0FB"}},shadow:{ClickableActive:"\n  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.4), 0px -2px 2px rgba(0, 0, 0, 0.08);;\n",ClickableDefault:"\n  box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.4), 0px -2px 4px rgba(0, 0, 0, 0.08);\n",ClickableHover:"\n  box-shadow: 0px 8px 16px rgba(0, 0, 0, 0.4), 0px -2px 8px rgba(0, 0, 0, 0.08);\n",NonClickable:"\n  box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.4), 0px 0px 6px rgba(0, 0, 0, 0.08);\n"},zIndex:P,name:"dark"},j="'VTB Group UI', sans-serif",H={fontFamily:j,fontSize:{12:"12px",14:"14px",16:"16px",18:"18px",22:"22px",28:"28px",32:"32px",36:"36px",42:"42px"},fontWeight:{normal:400,medium:500,bold:600},lineHeight:{16:"16px",20:"20px",24:"24px",28:"28px",36:"36px",40:"40px",44:"44px",52:"52px"}},W={"Main/XXL":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 42px;
    line-height: 52px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Main/XL":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 36px;
    line-height: 44px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Main/L":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 32px;
    line-height: 40px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Main/M":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 28px;
    line-height: 36px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Main/S":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 22px;
    line-height: 28px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Main/XS-bold":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 18px;
    line-height: 24px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Main/XS":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 18px;
    line-height: 24px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Additional/L-bold":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 16px;
    line-height: 24px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Additional/L":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 16px;
    line-height: 24px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Additional/M":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 16px;
    line-height: 20px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Additional/S":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 14px;
    line-height: 20px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Additional/S-bold":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: 500;
    font-size: 14px;
    line-height: 20px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Additional/XS":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 14px;
    line-height: 16px;
    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Button/M":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 16px;
    line-height: 24px;
    /* or 150% */

    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Button/S":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 14px;
    line-height: 20px;
    /* or 143% */

    font-feature-settings: 'tnum' on, 'lnum' on;
  `,"Caption/XS":u.css`
    font-family: ${j};
    font-style: normal;
    font-weight: normal;
    font-size: 12px;
    line-height: 16px;
    /* identical to box height, or 133% */

    font-feature-settings: 'tnum' on, 'lnum' on;
  `},U=({color:e,fontSize:t,fontWeight:r,lineHeight:n})=>u.css`
  color: ${e||"inherit"};
  font-family: ${H.fontFamily};
  font-size: ${H.fontSize[t]};
  font-weight: ${H.fontWeight[r]};
  line-height: ${H.lineHeight[n]};
`,Y=S.default.div`
  ${({color:e})=>U({color:e,fontSize:42,fontWeight:"bold",lineHeight:52})};
`,q=S.default.div`
  ${({color:e})=>U({color:e,fontSize:36,fontWeight:"bold",lineHeight:44})};
`,G=S.default.h1`
  ${({color:e})=>U({color:e,fontSize:32,fontWeight:"bold",lineHeight:40})};
`,K=S.default.h2`
  ${({color:e})=>U({color:e,fontSize:28,fontWeight:"bold",lineHeight:36})};
`,X=S.default.h3`
  ${({color:e})=>U({color:e,fontSize:22,fontWeight:"bold",lineHeight:28})};
`,Q=S.default.h4`
  ${({color:e})=>U({color:e,fontSize:18,fontWeight:"bold",lineHeight:24})};
`,Z=S.default.h5`
  ${({color:e})=>U({color:e,fontSize:18,fontWeight:"normal",lineHeight:24})};
`,J=S.default.h6`
  ${({color:e})=>U({color:e,fontSize:16,fontWeight:"bold",lineHeight:24})};
`,ee=S.default.div`
  ${({color:e})=>U({color:e,fontSize:16,fontWeight:"normal",lineHeight:24})};
`,te=S.default.div`
  ${({color:e})=>U({color:e,fontSize:16,fontWeight:"normal",lineHeight:20})};
`,re=S.default.div`
  ${({color:e})=>U({color:e,fontSize:14,fontWeight:"bold",lineHeight:20})};
`,ne=S.default.div`
  ${({color:e})=>U({color:e,fontSize:14,fontWeight:"normal",lineHeight:20})};
`,oe=S.default.div`
  ${({color:e})=>U({color:e,fontSize:14,fontWeight:"normal",lineHeight:16})};
`,ie=S.default.div`
  ${({color:e})=>U({color:e,fontSize:16,fontWeight:"medium",lineHeight:24})};
`,le=S.default.div`
  ${({color:e})=>U({color:e,fontSize:14,fontWeight:"medium",lineHeight:20})};
`,ae=S.default.div`
  ${({color:e})=>U({color:e,fontSize:12,fontWeight:"normal",lineHeight:16})};
`,se=u.css`
  background: ${({appearance:e,theme:t})=>{switch(e){case"info":return t.color.basic.primary;case"warning":return t.color.status.warn;case"success":return t.color.status.success;case"error":return t.color.status.danger;case"grey":return t.color.background.shadow;case"dark":return t.color.basic.secondary;case"white":case"whiteInactive":case"whiteDisable":return t.color.background.primary;case"light":case"lightInactive":case"lightDisable":default:return t.color.background.tertiary}}};
`,ce=u.css`
  color: ${({appearance:e,theme:t})=>{switch(e){case"info":case"warning":case"success":case"error":case"grey":case"dark":return t.color.text.staticWhite;case"lightInactive":case"lightDisable":case"whiteInactive":return t.color.text.secondary;case"whiteDisable":return t.color.text.tertiary;case"white":case"light":default:return t.color.text.primary}}};
`,de=S.default.div`
  position: relative;
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: ${({dimension:e})=>"s"===e?"0 5px":"0 7px"};
  height: ${({dimension:e})=>"s"===e?"16px":"20px"};
  border-radius: ${({dimension:e})=>"s"===e?"8px":"10px"};
  ${se}
  ${ce}
  user-select: none;
  ${({dimension:e})=>"s"===e?W["Caption/XS"]:W["Additional/S"]}
`,ue=e=>{var t=e,{children:r,dimension:n="m",appearance:o="light"}=t,i=c(t,["children","dimension","appearance"]);const l=`Количество ${r}`;return C.default.createElement(de,a({dimension:n,appearance:o,"aria-label":l},i),r)},pe=[0,1,2,3,4,5,6],fe="ru",he=e=>e.charAt(0).toUpperCase()+e.slice(1).toLowerCase(),me={ru:{code:"ru",options:{weekStartsOn:1,firstWeekContainsDate:1}},enUS:{code:"en-US",options:{weekStartsOn:0,firstWeekContainsDate:1}},de:{code:"de",options:{weekStartsOn:1,firstWeekContainsDate:4}}},be=e=>{const t=new Date(Date.UTC(e.getFullYear(),e.getMonth(),e.getDate(),e.getHours(),e.getMinutes(),e.getSeconds(),e.getMilliseconds()));return t.setUTCFullYear(e.getFullYear()),e.getTime()-t.getTime()},ge=e=>(e.setHours(0,0,0,0),e),xe=(e,t)=>{const r=new Date(e.getTime());return isNaN(t)?new Date(NaN):t?(r.setDate(r.getDate()+t),r):r},ve=(e,t)=>ye(e,12*t),ye=(e,t)=>{const r=new Date(e.getTime());if(isNaN(t))return new Date(NaN);if(!t)return r;const n=r.getDate(),o=new Date(r.getTime());o.setMonth(r.getMonth()+t+1,0);return n>=o.getDate()?o:(r.setFullYear(o.getFullYear(),o.getMonth(),n),r)},$e=(e,t)=>e.getTime()>t.getTime(),we=(e,t)=>e.getTime()<t.getTime(),Ee=(e,t,r)=>{let n=!1;const o=ge(t),i=((l=r).setHours(23,59,59,999),l);var l;try{n=((e,t)=>{const r=e.getTime(),n=t.start.getTime(),o=t.end.getTime();if(!(n<=o))throw new RangeError("Invalid interval");return r>=n&&r<=o})(e,{start:o,end:i})}catch(a){n=!1}return n},ke=(e,t)=>{const r=ge(e),n=ge(t),o=r.getTime()-be(r),i=n.getTime()-be(n);return Math.round((o-i)/864e5)},Ce=(e,t)=>12*(e.getFullYear()-t.getFullYear())+(e.getMonth()-t.getMonth()),De=(e,t)=>e.getFullYear()-t.getFullYear(),Se=(e,t)=>{return e&&t?(r=t,e.getTime()===r.getTime()):!e&&!t;var r},Fe=(e,t,r)=>new Intl.DateTimeFormat(me[r].code,t).format(e||new Date),Me=(e,t,r)=>t&&ke(e,t)<0||r&&ke(e,r)>0,_e=(e,t)=>e&&t?((e,t)=>{const r=ge(e),n=ge(t);return r.getTime()===n.getTime()})(e,t):!e&&!t,Re=(e,t)=>{return e&&t?(n=t,(r=e).getFullYear()===n.getFullYear()&&r.getMonth()===n.getMonth()):!e&&!t;var r,n},Ae=(e,t)=>{const r=e.getFullYear(),n=e.getDate(),o=new Date(0);o.setFullYear(r,t,15),o.setHours(0,0,0,0);const i=(e=>{const t=e.getFullYear(),r=e.getMonth(),n=new Date(0);return n.setFullYear(t,r+1,0),n.setHours(0,0,0,0),n.getDate()})(o),l=new Date(e);return l.setMonth(t,Math.min(n,i)),l},Oe=(e,t)=>{if(isNaN(e))return new Date(NaN);const r=new Date(e);return r.setFullYear(t),r},Be=(e,t)=>ve(e,-t),Te=e=>{const t=new Date(e);return t.setDate(1),t.setHours(0,0,0,0),t},ze=(e,t)=>{const r=xe(e,6);return Re(e,t)||Re(r,t)},Ie=(e,t)=>{const r=me[t],n=r&&r.options&&r.options.weekStartsOn,o=new Date(e.getTime()),i=o.getDay(),l=(i<n?7:0)+i-n;return o.setDate(o.getDate()-l),o.setHours(0,0,0,0),o},Le=(e,t)=>ye(e,-t),Pe=e=>!isNaN(e),Ne=N,Ve=S.default(ne)`
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  width: ${"60px"};
  height: ${"36px"};
  padding: ${"8px 0"};
  margin-bottom: ${"16px"};
  border: 1px solid ${({today:e,theme:t})=>e?t.color.basic.secondary:"transparent"};
  border-radius: ${"18px"};
  background: transparent;
  color: ${({theme:e})=>e.color.text.primary};
  user-select: none;
  cursor: pointer;

  ${({disabled:e,theme:t})=>e&&`\n      cursor: default;\n      color: ${t.color.text.tertiary};\n    `}

  ${({disabled:e,theme:t})=>!e&&`\n      &:hover {\n        border: 1px solid ${t.color.basic.primary};\n      }\n    `}

  ${({disabled:e,theme:t,selected:r})=>!e&&r&&`\n      border: 1px solid ${t.color.basic.primary};\n      color: ${t.color.text.staticWhite};\n      background: ${t.color.basic.primary};\n      &:hover {\n        border: 1px solid ${t.color.basic.hover};\n        background: ${t.color.basic.hover};\n      }\n    `}
`;Ve.defaultProps={theme:Ne};const je=({viewDate:e,startDate:t,selected:r,validator:n,onClick:o})=>{const{start:i,end:l}=((e,t)=>{const r=Math.ceil(e.getFullYear()/t)*t;return{start:r-(t-1),end:r}})(e,20),a=Array(l-i+1).fill(0).map(((e,t)=>i+t));return C.default.createElement(C.default.Fragment,null,a.map((i=>{const l=!!(null==n?void 0:n.invalidYear(i));return C.default.createElement(Ve,{key:i,today:i===(new Date).getFullYear(),selected:!(!r&&!t)&&i===(r||t).getFullYear(),disabled:l,onMouseDown:t=>{t.preventDefault();const r=(e=>{const t=new Date(0);return t.setFullYear(e.getFullYear(),0,1),t.setHours(0,0,0,0),t})(Oe(e,i));!l&&o&&o(r,t)}},i)})))};var He,We;function Ue(){return(Ue=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Ye(e){return D.createElement("svg",Ue({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),He||(He=D.createElement("path",{d:"M14.508 18.754c.18 0 .35-.07.48-.21.24-.27.22-.68-.04-.92l-6.369-5.74 6.41-5.55a.65.65 0 00.04-.92.658.658 0 00-.92-.04l-6.9 5.99c-.31.28-.31.76 0 1.04l6.86 6.18c.13.12.28.17.44.17z"})))}function qe(){return(qe=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Ge(e){return D.createElement("svg",qe({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),We||(We=D.createElement("path",{d:"M9.474 18.754c-.18 0-.35-.07-.48-.21a.658.658 0 01.04-.92l6.37-5.74-6.41-5.55a.65.65 0 01-.04-.92c.24-.26.65-.28.92-.04l6.9 5.99c.31.28.31.76 0 1.04l-6.86 6.18c-.13.12-.28.17-.44.17z"})))}const Ke=S.default.div`
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: ${"32px"};
  height: ${"32px"};
  padding: ${"4px"};
  border-radius: 50%;
  background-color: transparent;
  cursor: pointer;
  ${({disabled:e})=>e&&"\n      pointer-events: none;\n      cursor: default;\n    "}

  & svg {
    fill: ${({disabled:e,theme:t})=>e?t.color.basic.disable:t.color.basic.tertiary};
  }

  ${({disabled:e,theme:t})=>`\n    &:hover,\n    &:focus {\n      background-color: ${e?"transparent":t.color.background.secondary};\n      & svg {\n        fill: ${t.color.basic.hover};\n      }\n    }\n    &:active {\n      background-color: ${e?"transparent":t.color.background.secondary};\n      & svg {\n        fill: ${t.color.basic.press};\n      }\n    }\n  `}
`;Ke.defaultProps={theme:Ne};const Xe=d.memo((({onMouseDown:e,disabled:t,type:r})=>C.default.createElement(Ke,{onMouseDown:t=>{null==t||t.preventDefault(),e(t)},disabled:t},"left"===r&&C.default.createElement(Ye,{width:24,height:24}),"right"===r&&C.default.createElement(Ge,{width:24,height:24})))),Qe=S.default(ne)`
  position: relative;
  display: inline-block;
  width: ${"36px"};
  height: ${"36px"};
  padding: ${"8px 0"};
  margin-bottom: ${"4px"};
  color: ${({theme:e})=>e.color.text.primary};
  box-sizing: border-box;
  user-select: none;
  cursor: ${({disabled:e})=>e?"default":"pointer"};

  &:after {
    content: '';
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    border: 1px solid
      ${({theme:e,today:t,selected:r})=>t&&!r?e.color.basic.secondary:"transparent"};
    border-radius: 50%;
  }

  ${({disabled:e,theme:t,outsideMonth:r})=>(e||r)&&`\n      color: ${t.color.text.tertiary};\n    `}

  ${({disabled:e,theme:t})=>!e&&`\n      &:hover {\n        &:after {\n          border: 1px solid ${t.color.basic.primary};\n          background: white;\n          z-index: -1;\n        }\n      }\n    `}

  ${({disabled:e,theme:t,selected:r,inSelectingRange:n})=>!e&&r&&`\n      color: ${t.color.text.staticWhite};\n      background: ${n?t.color.basic.hover:t.color.basic.primary};\n      border-radius: 50%;\n      &:hover {\n        background: ${t.color.basic.hover};\n      }\n      &:active {\n        background: ${t.color.basic.press};\n      }\n    `}

  ${({disabled:e,inRange:t,theme:r,corners:n,selected:o,isActiveDate:i})=>!e&&t&&`\n      &:before {\n        z-index: -1;\n        content: '';\n        position: absolute;\n        width: 100%;\n        left: 0;\n        right: 0;\n        top: 0;\n        bottom: 0;\n        background: ${r.color.background.tertiary};\n        ${n&&Object.keys(n).map((e=>n[e]?o||i?`border-${e}-radius: 50%;`:`border-${e}-radius: 4px;`:"")).join("")}\n      }\n    `}
`;Qe.defaultProps={theme:Ne};const Ze=({day:e,month:t,startDate:r,endDate:n,selected:o,activeDate:i,range:l,localeName:a=fe,validator:s,filterDate:c,onMouseEnter:d,onClick:u})=>{const p=!!(null==s?void 0:s.invalidValue(e))||c&&!c(e),f=void 0!==t&&t!==e.getMonth(),h=!p&&!!l&&!!r&&!!i&&!n&&($e(i,r)||Se(i,r))&&Ee(e,r,i),m=!!r&&!!n&&Ee(e,r,n),b=!!r&&_e(e,r),g=!!r&&!!n&&_e(e,n),x=h&&_e(e,r),v=h&&_e(e,i),y={};if(r){const t=_e(e,Ie(e,a)),r=_e(e,((e,t)=>{const r=me[t],n=r&&r.options&&r.options.weekStartsOn,o=new Date(e.getTime()),i=o.getDay(),l=6+(i<n?-7:0)-(i-n);return o.setDate(o.getDate()+l),o.setHours(23,59,59,999),o})(e,a)),n=b||x,o=g||v;y["top-left"]=n||t&&!o,y["bottom-left"]=n||t&&!o,y["top-right"]=o||r&&!n,y["bottom-right"]=o||r&&!n}return C.default.createElement(Qe,{today:_e(e,new Date),selected:_e(e,o)||b||g,inSelectingRange:h,isActiveDate:_e(e,i),corners:y,inRange:m||h,disabled:p,outsideMonth:f,onMouseEnter:t=>!p&&d&&d(e,t),onMouseDown:t=>{t.preventDefault(),!p&&u&&u(e,t)}},e.getDate())},Je=S.default(ne)`
  display: inline-block;
  user-select: none;
  width: ${"36px"};
  height: ${"36px"};
  padding: ${"8px 0"};
  margin-bottom: ${"4px"};
  color: ${({theme:e})=>e.color.text.primary};
  box-sizing: border-box;
`;Je.defaultProps={theme:Ne};const et=({date:e,localeName:t})=>{const r=Ie(e,t);return C.default.createElement("div",null,pe.map((e=>{const n=xe(r,e),o=he(Fe(n,{weekday:"short"},t).slice(0,2));return C.default.createElement(Je,{key:n.valueOf()},o)})))},tt=S.default.div`
  white-space: nowrap;
`,rt=({localeName:e=fe,day:t,month:r,startDate:n,endDate:o,selected:i,activeDate:l,range:a,validator:s,filterDate:c,onMouseEnter:d,onClick:u})=>{const p=Ie(t,e);return C.default.createElement(tt,null,pe.map((t=>{const f=xe(p,t);return C.default.createElement(Ze,{key:f.valueOf(),day:f,month:r,startDate:n,endDate:o,selected:i,activeDate:l,range:a,localeName:e,validator:s,filterDate:c,onMouseEnter:(e,t)=>((e,t)=>d&&d(e,t))(f,t),onClick:(e,t)=>((e,t)=>u&&u(e,t))(f,t)})})))},nt=({localeName:e=fe,day:t,startDate:r,endDate:n,selected:o,activeDate:i,range:l,validator:a,filterDate:s,onMouseEnter:c,onMouseLeave:d,onClick:u})=>{const p=[],f=(e,t)=>c&&c(e,t),h=(e,t)=>u&&u(e,t);let m=Ie(Te(t),e);do{p.push(m),m=xe(m,7*1)}while(ze(m,t));return C.default.createElement("div",{onMouseLeave:()=>d&&d()},p.map((c=>C.default.createElement(rt,{key:c.valueOf(),day:c,month:t.getMonth(),startDate:r,endDate:n,selected:o,activeDate:i,range:l,localeName:e,validator:a,filterDate:s,onMouseEnter:f,onClick:h}))))},ot=S.default(Ve)`
  width: ${"84px"};
  margin-bottom: ${"32px"};
`,it=["Январь","Февраль","Март","Апрель","Май","Июнь","Июль","Август","Сентябрь","Октябрь","Ноябрь","Декабрь"],lt=({viewDate:e,startDate:t,selected:r,validator:n,onClick:o})=>C.default.createElement(C.default.Fragment,null,it.map(((i,l)=>{const a=!!(null==n?void 0:n.invalidMonth(l,e.getFullYear()));return C.default.createElement(ot,{key:i,today:l===(new Date).getMonth(),selected:!(!r&&!t)&&l===(r||t).getMonth(),disabled:a,onMouseDown:t=>{t.preventDefault();const r=Te(Ae(e,l));!a&&o&&o(r,t)}},i)}))),at="12px",st="28px",ct=S.default.div`
  display: flex;
  align-items: center;
  background-color: transparent;
  border: 0 none;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
  margin-bottom: ${({monthsView:e,yearsView:t})=>e||t?st:at};
`,dt=S.default(J)`
  display: flex;
  align-items: center;
  box-sizing: border-box;
  height: ${"32px"};
  padding: ${"4px 8px"};
  margin: 0;
  border-radius: ${"16px"};

  ${({theme:e,view:t})=>`\n    color: ${e.color.basic.primary};\n    background: ${t?e.color.background.tertiary:"transparent"};\n\n    &:hover,\n    &:focus {\n      color: ${e.color.basic.hover};\n      background: ${e.color.background.secondary};\n    }\n    &:active {\n      color: ${e.color.basic.press};\n      background: ${e.color.background.secondary};\n    }\n  `}
`,ut=S.default(dt)`
  width: ${"54px"};
`;ut.defaultProps={theme:Ne};const pt=S.default(dt)``;pt.defaultProps={theme:Ne};const ft=S.default.div`
  display: flex;
  justify-content: center;
  width: 100%;
`;function ht(e,t,r){(r.current=e,t)&&("function"!=typeof t?t.current=e:function(e){return"function"==typeof e}(t)&&t(e))}const mt=e=>{const t=[];if(e){let r=e;for(;r;){const e=window.getComputedStyle(r);"auto"!==e.getPropertyValue("overflow")&&"scroll"!==e.getPropertyValue("overflow")&&"auto"!==e.getPropertyValue("overflow-x")&&"scroll"!==e.getPropertyValue("overflow-x")&&"auto"!==e.getPropertyValue("overflow-y")&&"scroll"!==e.getPropertyValue("overflow-y")||t.push(r),r=r.parentElement}}return t},bt=S.default.div`
  display: inline-block;
  position: relative;
  cursor: pointer;
`,gt=S.default.div`
  position: absolute;
  top: 0;
  left: 0;
  z-index: ${({theme:e})=>e.zIndex.tooltip};
  width: auto;
  background-color: ${({theme:e})=>e.color.background.primary};
  color: ${({theme:e})=>e.color.text.primary};
  font-family: ${H.fontFamily};
  font-size: 14px;
  font-weight: 400;
  line-height: 16px;
  border-radius: ${"4px"};
  ${e=>e.theme.shadow.ClickableDefault}
  padding: ${"4px 8px"};
`,xt={bottom:{check:(e,t)=>{const r=window.innerHeight-e.bottom>8+t.height,n=e.left+e.width/2>t.width/2,o=window.innerWidth-e.right+e.width/2>t.width/2;return r&&n&&o},getStyles:(e,t)=>`\n      translate3d(${Math.floor(e.left+e.width/2-t.width/2)}px, ${Math.floor(e.bottom+8+window.pageYOffset)}px, 0px)\n    `},top:{check:(e,t)=>{const r=e.top>8+t.height,n=e.left+e.width/2>t.width/2,o=window.innerWidth-e.right+e.width/2>t.width/2;return r&&n&&o},getStyles:(e,t)=>`\n      translate3d(${Math.floor(e.left+e.width/2-t.width/2)}px, ${Math.floor(e.top-t.height-8+window.pageYOffset)}px, 0px)\n    `},left:{check:(e,t)=>{const r=e.left>8+t.width,n=e.top>(t.height-e.height)/2,o=window.innerHeight-e.bottom>(t.height-e.height)/2;return r&&o&&n},getStyles:(e,t)=>`\n      translate3d(${Math.floor(e.left-t.width-8)}px, ${Math.floor(e.top+e.height/2-t.height/2+window.pageYOffset)}px, 0px)\n    `},right:{check:(e,t)=>{const r=window.innerWidth-e.right>8+t.width,n=e.top>(t.height-e.height)/2,o=window.innerHeight-e.bottom>(t.height-e.height)/2;return r&&o&&n},getStyles:(e,t)=>`\n    translate3d(${Math.floor(e.right+8)}px, ${Math.floor(e.top+e.height/2-t.height/2+window.pageYOffset)}px, 0px)\n    `}};const vt=e=>{var t=e,{renderContent:r,container:n,withDelay:o,tooltipRef:i,children:l,className:s}=t,u=c(t,["renderContent","container","withDelay","tooltipRef","children","className"]);const p=d.useRef(null),f=d.useRef(null),h=n||document.body;let m,b;const[g,x]=d.useState(!1),v=()=>x(!1),y=()=>{if(p.current&&f.current){const e=function(e,t){const r=e.getBoundingClientRect(),n=t.getBoundingClientRect(),o=Object.entries(xt).filter((e=>e[1].check(r,n)));return o.length?o[0][1]:xt.bottom}(p.current,f.current),t=p.current.getBoundingClientRect(),r=f.current.getBoundingClientRect(),n=e.getStyles(t,r);f.current.style.transform=n}};d.useEffect((()=>(window.addEventListener("resize",v),window.addEventListener("scroll",v),!m&&p.current&&(m=mt(p.current),null==m||m.forEach((e=>e.addEventListener("scroll",v)))),()=>{window.removeEventListener("resize",v),window.removeEventListener("scroll",v),null==m||m.forEach((e=>e.removeEventListener("scroll",v)))}))),d.useEffect((()=>{y()}),[r(),p]);const $=()=>{b=window.setTimeout((()=>{x(!0),y()}),o?1500:0)},w=()=>{clearTimeout(b),v()};return C.default.createElement(bt,{onMouseEnter:$,onMouseLeave:w,onFocus:$,onBlur:w,ref:p,className:s},l,g&&F.default.createPortal(C.default.createElement(gt,a({role:"tooltip",ref:e=>ht(e,i,f)},u),r()),h))},yt=({viewDate:e,minDate:t,maxDate:r,yearsView:n,monthsView:o,localeName:i,onChangeYearsView:l,onChangeMonthsView:a,onNext:s,onPrevious:c})=>{const d=!!t&&Ce(t,Le(e,1))>0,u=!!r&&Ce(ye(e,1),r)>0,p=!!t&&De(t,Be(e,1))>0,f=!!r&&De(ve(e,1),r)>0,h=n?p:d,m=n?f:u;return C.default.createElement(ct,{yearsView:n,monthsView:o},!o&&C.default.createElement(vt,{renderContent:()=>n?"Назад":"Предыдущий месяц"},C.default.createElement(Xe,{onMouseDown:c,disabled:h,type:"left"})),C.default.createElement(ft,null,C.default.createElement(vt,{renderContent:()=>o?"Вернуться":"Выбор месяца"},C.default.createElement(pt,{view:o,onMouseDown:e=>{e.preventDefault(),a(e),n&&l(e)}},he(Fe(e,{month:"long"},i)))),C.default.createElement(vt,{renderContent:()=>n?"Вернуться":"Выбор года"},C.default.createElement(ut,{view:n,onMouseDown:e=>{e.preventDefault(),l(e),o&&a(e)}},e.getFullYear()))),!o&&C.default.createElement(vt,{renderContent:()=>n?"Вперед":"Следующий месяц"},C.default.createElement(Xe,{onMouseDown:s,disabled:m,type:"right"})))},$t=S.default.div`
  position: relative;
  box-sizing: border-box;
  text-align: center;
  border: 0 none;
  z-index: 0; /* to fix range rounded corners fill */

  padding: ${({yearsView:e,monthsView:t})=>e?"20px 12px 16px":t?"20px 16px 4px":"20px 12px 12px"};

  width: ${284}px;
  background: ${({theme:e})=>e.color.background.primary};
  box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.12), 0px 3.2px 8px rgba(0, 0, 0, 0.14);
  font-family: 'VTB Group UI', sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
`;$t.defaultProps={theme:Ne};const wt=C.default.forwardRef(((e,t)=>{var r=e,{startDate:n,endDate:o,validator:i,minDate:l,maxDate:u,selected:p,range:f,tooltipContainer:h,localeName:m=fe,filterDate:b,onChange:g}=r,x=c(r,["startDate","endDate","validator","minDate","maxDate","selected","range","tooltipContainer","localeName","filterDate","onChange"]);const[v,y]=d.useState((()=>{const e=new Date;return p||(l&&we(e,l)?l:u&&$e(e,u)?u:e)})()),[$,w]=d.useState(null),[E,k]=d.useState(!1),[D,S]=d.useState(!1);d.useEffect((()=>{f&&n&&(R(n.getFullYear()),A(n.getMonth()))}),[]);const F=()=>i||((e,t)=>({invalidValue:r=>r?Pe(r)?Me(r,e,t)?"Дата вне диапазона":null:"Дата не валидна":"Неверный формат даты",invalidRange:(r,n)=>{return r||n?Pe(r)||Pe(n)?r?n?Pe(r)?Pe(n)?(i=n,(null===(o=r)&&null===i||void 0===o&&void 0===i?0:null==o&&null!=i?-1:null==i&&null!=o?1:o&&i?o.getTime()-i.getTime():0)>=0?"Конечная дата меньше начальной":Me(r,e,t)?"Начальная дата вне диапазона":Me(n,e,t)?"Конечная дата вне диапазона":null):"Конечная дата не валидна":"Начальная дата не валидна":"Неверный формат конечной даты":"Неверный формат начальной даты":"Даты не валидны":"Неверный формат дат";var o,i},invalidYear:r=>Me(new Date(r,11,31),e)||Me(new Date(r,0,1),null,t)?"Дата вне диапазона":null,invalidMonth:(r,n)=>{const o=(e=>{const t=e.getMonth();return e.setFullYear(e.getFullYear(),t+1,0),e.setHours(23,59,59,999),e})(new Date(n,r,1)).getDate();return Me(new Date(n,r,o),e)||Me(new Date(n,r,1),null,t)?"Дата вне диапазона":null}}))(l,u),M=()=>k((e=>!e)),_=()=>S((e=>!e)),R=e=>y((t=>Oe(t,e))),A=e=>y((t=>Ae(t,e))),O=()=>y((e=>ve(e,E?20:1))),B=()=>y((e=>Be(e,E?20:1))),T=()=>y((e=>ye(e,1))),z=()=>y((e=>Le(e,1))),I=(e,t)=>w(e),L=()=>w(null),P=(e,t)=>{let r=e;var i,l;!f&&Se(p,r)||(i=r,r=(l=p)?(i.setMilliseconds(l.getMilliseconds()),i.setSeconds(l.getSeconds()),i.setMinutes(l.getMinutes()),i.setHours(l.getHours()),i):i,f?(n||o?n&&!o?we(r,n)?g([r,null],t):g([n,r],t):!n&&o&&(we(r,o)?g([r,o],t):g([r,null],t)):g([r,null],t),n&&o&&g([r,null],t)):g(r,t))},N=(e,t)=>{R(e.getFullYear()),k(!1)},V=(e,t)=>{A(e.getMonth()),S(!1)};return C.default.createElement($t,s(a({yearsView:E,monthsView:D},x),{ref:t}),C.default.createElement(yt,{viewDate:v,minDate:l,maxDate:u,yearsView:E,monthsView:D,localeName:m,onChangeYearsView:M,onChangeMonthsView:_,onNext:E?O:T,onPrevious:E?B:z,tooltipContainer:h}),E&&C.default.createElement(je,{viewDate:v,startDate:n,endDate:o,selected:p,localeName:m,range:f,validator:F(),onClick:N}),D&&C.default.createElement(lt,{viewDate:v,startDate:n,endDate:o,selected:p,localeName:m,range:f,validator:F(),onClick:V}),!E&&!D&&C.default.createElement(C.default.Fragment,null,C.default.createElement(et,{date:v,localeName:m}),C.default.createElement(nt,{day:v,startDate:n,endDate:o,selected:p,activeDate:$,range:f,localeName:m,validator:F(),filterDate:b,onMouseEnter:I,onMouseLeave:L,onClick:P})))}));wt.defaultProps={localeName:fe};const Et=S.default.span`
  display: inline-block;
  position: absolute;
  margin: 0;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  content: '';
  box-sizing: border-box;
  ${({theme:e,dimension:t})=>`\n    background-color: ${e.color.background.primary};\n    width: ${"s"===t?"18px":"20px"};\n    height: ${"s"===t?"18px":"20px"};\n  `}
  fieldset[data-dimension='s'] & {
    width: ${"18px"};
    height: ${"18px"};
  }
  border: ${"1px"} solid
    ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.basic.tertiary};
  fieldset:disabled & {
    border: ${"1px"} solid ${({theme:e})=>e.color.text.tertiary};
  }
  border-radius: 50%;
  transition: all 0.25s ease-in-out;
`,kt=S.default.input`
  opacity: 0;
  width: 0;
  height: 0;

  &:checked:disabled + ${Et} {
    border: ${"6px"} solid ${({theme:e})=>e.color.text.tertiary};
  }

  &:checked:not(disabled) + ${Et} {
    border: ${"6px"} solid ${({theme:e})=>e.color.basic.primary};
  }

  &:checked:not(:disabled) {
    &:focus + ${Et}, &:hover + ${Et} {
      ${({theme:e})=>`\n          background-color: ${e.color.background.secondary};\n          border: 6px solid ${e.color.basic.hover};\n        `}
    }
  }

  &:not(checked):not(:disabled) {
    &:focus + ${Et}, &:hover + ${Et} {
      ${({theme:e})=>`\n          background-color: ${e.color.background.secondary};\n          border: 1px solid ${e.color.basic.primary};\n        `}
    }
  }
`,Ct=S.default.div`
  position: relative;
  display: inline-block;
  ${({dimension:e})=>`\n    min-width: ${"s"===e?"18px":"20px"};\n    height: ${"s"===e?"18px":"20px"};\n  `}
  fieldset[data-dimension='s'] & {
    min-width: ${"18px"};
    height: ${"18px"};
  }
`,Dt=S.default.span`
  display: inline-flex;
  flex-direction: column;
  box-sizing: content-box;
  ${({dimension:e})=>"s"===e?W["Additional/XS"]:W["Additional/M"]}
  padding-left: ${({dimension:e})=>"s"===e?"9px":"10px"};

  fieldset[data-dimension='s'] & {
    ${W["Additional/XS"]}
    padding-left: ${"9px"};
  }

  color: ${({disabled:e,theme:t})=>e?t.color.text.tertiary:t.color.basic.secondary};
  fieldset:disabled & {
    color: ${({theme:e})=>e.color.text.tertiary};
  }
  margin: auto 0;
  user-select: none;
`,St=S.default.span`
  margin-top: 6px;
  ${({dimension:e})=>"s"===e?W["Caption/XS"]:W["Additional/XS"]}

  fieldset[data-dimension='s'] & {
    ${W["Caption/XS"]}
  }

  color: ${({disabled:e,theme:t})=>e?t.color.text.tertiary:t.color.text.secondary};
  fieldset:disabled & {
    color: ${({theme:e})=>e.color.text.tertiary};
  }
`,Ft=S.default.label`
  margin: 0;
  padding: 0;
  display: inline-flex;
  align-items: flex-start;
  box-sizing: content-box;
  cursor: ${({disabled:e})=>e?"default":"pointer"};
  fieldset:disabled & {
    cursor: default;
  }
`,Mt=e=>e?"6px":"1px";u.css`
  border: ${({theme:e,checked:t,disabled:r})=>`${Mt(t)} solid ${((e,t,r)=>r?e.color.text.tertiary:t?e.color.basic.primary:e.color.basic.tertiary)(e,t,r)}`};
`,u.css`
  border: ${({theme:e,checked:t})=>`${Mt(t)} solid ${((e,t)=>t?e.color.basic.hover:e.color.basic.primary)(e,t)}`};
`,u.css`
  font-family: ${H.fontFamily};
  font-size: ${({dimension:e})=>"m"===e?"16px":"14px"};
  line-height: ${({dimension:e})=>"m"===e?"20px":"16px"};
  font-weight: ${400};
  color: ${({disabled:e,theme:t})=>e?t.color.text.tertiary:t.color.basic.secondary};
`;const _t=D.forwardRef(((e,t)=>{var r=e,{children:n,disabled:o,dimension:i="m",extraText:l,className:s}=r,d=c(r,["children","disabled","dimension","extraText","className"]);return D.createElement(Ft,{disabled:o,className:s},D.createElement(Ct,{dimension:i},D.createElement(kt,a({ref:t,type:"radio",disabled:o},d)),D.createElement(Et,{disabled:o,dimension:i})),D.createElement(Dt,{disabled:o,dimension:i},n,l&&D.createElement(St,{disabled:o,dimension:i},l)))})),Rt=u.css`
  color: ${({theme:e})=>e.color.text.tertiary};
  fill: ${({theme:e})=>e.color.text.tertiary};
  user-select: none;
`,At=u.css`
  color: ${({theme:e,appearance:t})=>"secondary"===t?e.color.text.primary:e.color.basic.primary};
  fill: ${({theme:e,appearance:t})=>"secondary"===t?e.color.basic.tertiary:e.color.basic.primary};
`,Ot=u.css`
  &:hover {
    color: ${({theme:e})=>e.color.basic.hover};
    fill: ${({theme:e})=>e.color.basic.hover};
  }
  &:focus {
    color: ${({theme:e})=>e.color.basic.hover};
    fill: ${({theme:e})=>e.color.basic.hover};
    outline-color: ${({theme:e})=>e.color.basic.hover};
  }
  &:active {
    color: ${({theme:e})=>e.color.basic.press};
    fill: ${({theme:e})=>e.color.basic.press};
  }
`,Bt=u.css`
  font-size: ${({dimension:e})=>"m"===e?16:14}px;
  font-weight: normal;
  line-height: ${({dimension:e})=>"m"===e?24:20}px;
  font-family: ${H.fontFamily};
`,Tt=S.default.a`
  cursor: ${e=>e.disabled?"default":"pointer"};
  display: flex;
  pointer-events: ${e=>e.disabled?"none":"auto"};
  align-items: center;
  width: fit-content;
  text-decoration: none;
  position: relative;
  ${Ot};
  ${At};
  ${Bt};
  ${e=>e.disabled?Rt:""}
`;Tt.defaultProps={theme:N};const zt=D.forwardRef(((e,t)=>{var r=e,{disabled:n=!1,appearance:o="primary",dimension:i="m"}=r,l=c(r,["disabled","appearance","dimension"]);return D.createElement(Tt,a({dimension:i,ref:t,tabIndex:n?-1:l.tabIndex,disabled:n,appearance:o},l))})),It=u.css`
  background-color: ${({theme:e})=>e.color.basic.primary};
  color: ${({theme:e})=>e.color.text.staticWhite};
  border: 2px solid ${({theme:e})=>e.color.basic.primary};
  svg {
    fill: ${({theme:e})=>e.color.text.staticWhite};
  }

  :focus,
  :hover {
    background-color: ${({theme:e})=>e.color.basic.hover};
    border-color: ${({theme:e})=>e.color.basic.hover};
  }

  :active {
    background-color: ${({theme:e})=>e.color.basic.press};
    border-color: ${({theme:e})=>e.color.basic.press};
  }

  &[data-appearance~='disabled'],
  :disabled {
    background-color: ${({theme:e})=>e.color.basic.disable};
    color: ${({theme:e})=>e.color.text.inversion};
    border-color: ${({theme:e})=>e.color.basic.disable};
    svg {
      fill: ${({theme:e})=>e.color.text.inversion};
    }
  }
`,Lt=u.css`
  background-color: transparent;
  color: ${({theme:e})=>e.color.basic.primary};
  border: 2px solid ${({theme:e})=>e.color.basic.primary};

  & svg {
    fill: ${({theme:e})=>e.color.basic.primary};
  }

  &:focus,
  &:hover {
    background-color: ${({theme:e})=>e.color.background.secondary};
    color: ${({theme:e})=>e.color.basic.hover};
    border-color: ${({theme:e})=>e.color.basic.hover};
    & svg {
      fill: ${({theme:e})=>e.color.basic.hover};
    }
  }

  &:active {
    background-color: ${({theme:e})=>e.color.background.secondary};
    color: ${({theme:e})=>e.color.basic.press};
    border-color: ${({theme:e})=>e.color.basic.press};
    & svg {
      fill: ${({theme:e})=>e.color.basic.press};
    }
  }

  &[data-appearance~='disabled'],
  &:disabled {
    background-color: ${({theme:e})=>e.color.background.primary};
    color: ${({theme:e})=>e.color.text.tertiary};
    border-color: ${({theme:e})=>e.color.text.tertiary};
    & svg {
      fill: ${({theme:e})=>e.color.text.tertiary};
    }
  }
`,Pt=u.css`
  background-color: transparent;
  color: ${({theme:e})=>e.color.basic.primary};
  border: 2px solid transparent;

  & svg {
    fill: ${({theme:e})=>e.color.basic.primary};
  }

  &:focus,
  &:hover {
    background-color: ${({theme:e})=>e.color.background.secondary};
    color: ${({theme:e})=>e.color.basic.hover};
    border-color: ${({theme:e})=>e.color.background.secondary};
    & svg {
      fill: ${({theme:e})=>e.color.basic.hover};
    }
  }

  &:active {
    background-color: ${({theme:e})=>e.color.background.secondary};
    color: ${({theme:e})=>e.color.basic.press};
    border-color: ${({theme:e})=>e.color.background.secondary};
    & svg {
      fill: ${({theme:e})=>e.color.basic.press};
    }
  }

  &[data-appearance~='disabled'],
  &:disabled {
    background-color: transparent;
    color: ${({theme:e})=>e.color.text.tertiary};
    border-color: transparent;
    & svg {
      fill: ${({theme:e})=>e.color.text.tertiary};
    }
  }
`,Nt=u.css`
  background-color: transparent;
  color: ${({theme:e})=>e.color.text.staticWhite};
  border: 2px solid ${({theme:e})=>e.color.text.staticWhite};
  & svg {
    fill: ${({theme:e})=>e.color.text.staticWhite};
  }

  &[data-appearance~='disabled'],
  &:disabled {
    border-color: ${({theme:e})=>e.color.basic.tertiary};
    color: ${({theme:e})=>e.color.basic.tertiary};
    & svg {
      fill: ${({theme:e})=>e.color.basic.tertiary};
    }
  }
`,Vt=u.css`
  background-color: ${({theme:e})=>e.color.status.danger};
  color: ${({theme:e})=>e.color.text.staticWhite};
  border: 2px solid ${({theme:e})=>e.color.status.danger};
  & svg {
    fill: ${({theme:e})=>e.color.text.staticWhite};
  }

  &:focus,
  &:hover {
    background-color: ${({theme:e})=>e.color.status.hoverDanger};
    border-color: ${({theme:e})=>e.color.status.hoverDanger};
  }

  &:active {
    background-color: ${({theme:e})=>e.color.status.pressDanger};
    border-color: ${({theme:e})=>e.color.status.pressDanger};
  }

  &[data-appearance~='disabled'],
  &:disabled {
    background-color: ${({theme:e})=>e.color.basic.disable};
    color: ${({theme:e})=>e.color.text.inversion};
    border-color: ${({theme:e})=>e.color.basic.disable};
    & svg {
      fill: ${({theme:e})=>e.color.text.inversion};
    }
  }
`,jt=u.css`
  background-color: ${({theme:e})=>e.color.status.success};
  color: ${({theme:e})=>e.color.text.staticWhite};
  border: 2px solid ${({theme:e})=>e.color.status.success};

  & svg {
    fill: ${({theme:e})=>e.color.text.staticWhite};
  }

  &:focus,
  &:hover {
    background-color: ${({theme:e})=>e.color.status.hoverSuccess};
    border-color: ${({theme:e})=>e.color.status.hoverSuccess};
  }

  &:active {
    background-color: ${({theme:e})=>e.color.status.pressSuccess};
    border-color: ${({theme:e})=>e.color.status.pressSuccess};
  }

  &[data-appearance~='disabled'],
  :disabled {
    background-color: ${({theme:e})=>e.color.basic.disable};
    color: ${({theme:e})=>e.color.text.inversion};
    border-color: ${({theme:e})=>e.color.basic.disable};
    svg {
      fill: ${({theme:e})=>e.color.text.inversion};
    }
  }
`,Ht=u.css`
  &[data-appearance~='primary'] {
    ${It}
  }
  &[data-appearance~='secondary'] {
    ${Lt}
  }
  &[data-appearance~='ghost'] {
    ${Pt}
  }
  &[data-appearance~='white'] {
    ${Nt}
  }
  &[data-appearance~='danger'] {
    ${Vt}
  }
  &[data-appearance~='success'] {
    ${jt}
  }
`,Wt=u.css`
  padding: 0;
  height: 56px;
  ${e=>e.displayAsSquare?"width: 56px;":"padding: 0 30px;"}

  ${W["Button/M"]}
`,Ut=u.css`
  ${Wt}

  &[data-dimension='xl'] {
    ${Wt}
  }

  &[data-dimension='l'] {
    height: 48px;
    ${e=>e.displayAsSquare?"width: 48px;":"padding: 0 22px;"}
  }

  &[data-dimension='m'] {
    height: 40px;
    ${e=>e.displayAsSquare?"width: 40px;":"padding: 0 18px;"}
  }

  &[data-dimension='s'] {
    height: 32px;
    ${e=>e.displayAsSquare?"width: 32px;":"padding: 0 14px;"}
    ${W["Button/S"]}
  }
`;function Yt(){return(performance.now().toString(36)+Math.random().toString(36)).replace(/\./g,"")}const qt=S.default.button.attrs((e=>({"data-dimension":e.dimension,"data-appearance":[e.appearance,e.displayAsDisabled?"disabled":void 0].filter((e=>void 0!==e)).join(" ")})))`
  box-sizing: border-box;
  display: inline-block;
  border: none;
  border-radius: 4px;
  appearance: none;
  vertical-align: center;

  ${Ht};

  ${Ut}
  &:hover {
    cursor: pointer;
  }

  &:disabled {
    cursor: not-allowed;
  }
`,Gt=S.default.span`
  vertical-align: top;

  display: inline-flex;
  flex-direction: row;
  overflow: hidden;
  flex-wrap: nowrap;
  justify-content: center;
  align-items: center;
  height: 24px;

  > * {
    display: inline-block;
    flex: 0 1 auto;
    white-space: nowrap;
  }

  > *:not(:first-child) {
    margin-left: 8px;
  }

  & > svg {
    width: 24px;
    height: 24px;
  }

  [data-dimension='s'] & {
    height: 20px;

    & > svg {
      width: 20px;
      height: 20px;
    }
  }
`,Kt=D.forwardRef(((e,t)=>{var r=e,{appearance:n="primary",dimension:o="xl",type:i="button",children:l}=r,s=c(r,["appearance","dimension","type","children"]);return D.createElement(qt,a({ref:t,appearance:n,dimension:o,type:i},s),D.createElement(Gt,null,D.Children.toArray(l).map((e=>"string"==typeof e?D.createElement("span",{key:Yt()},e):e))))})),Xt=u.css`
  color: ${({active:e,disabled:t,theme:r})=>e?r.color.text.primary:t?r.color.text.tertiary:r.color.text.secondary};
`,Qt=u.css`
  font-weight: 500;
  font-family: ${H.fontFamily};
  white-space: nowrap;
`,Zt=S.default.button`
  border-radius: ${22}px;
  box-sizing: border-box;
  cursor: ${({disabled:e})=>e?"default":"pointer"};
  display: flex;
  align-items: center;
  position: relative;
  outline: 0;
  border: 2px solid transparent;
  background: ${({theme:e,disabled:t,active:r})=>r&&!t?e.color.background.primary:"transparent"};
  ${Qt};
  ${Xt};

  &:hover {
    background: ${({active:e,disabled:t,theme:r})=>e||t?"":r.color.background.tertiaryHover};
  }

  &:focus-visible {
    border: 2px solid ${({theme:e})=>e.color.basic.hover};
    border-radius: ${22}px;
  }

  & svg {
    fill: ${({active:e,disabled:t,theme:r})=>e?r.color.text.primary:t?r.color.text.tertiary:r.color.text.secondary};
  }
`;Zt.defaultProps={type:"button"};const Jt={l:40,m:32,s:24},er={l:24,m:20,s:16},tr={l:24,m:20,s:16},rr=S.default.div`
  background: ${({theme:e})=>e.color.background.tertiary};
  border-radius: ${({dimension:e})=>er[e]}px;
  display: flex;
  height: ${({dimension:e})=>Jt[e]}px;
  padding: 4px;
  width: fit-content;

  ${Zt} {
    height: 100%;
    padding: ${({dimension:e})=>`0 ${tr[e]}px`};
    font-size: ${({dimension:e})=>"l"===e?16:14}px;
    line-height: ${({dimension:e})=>"l"===e?24:20}px;
  }
`,nr=(e,t,r)=>{let n=!1,o=r(e,t);for(;o;){if(o===(null==e?void 0:e.firstChild)){if(n)return;n=!0}if(!o.hasAttribute("disabled"))return void o.focus();o=r(e,o)}},or=(e,t)=>(null==t?void 0:t.nextElementSibling)?t.nextElementSibling:null==e?void 0:e.firstChild,ir=(e,t)=>(null==t?void 0:t.previousElementSibling)?t.previousElementSibling:null==e?void 0:e.lastChild;function lr(e){return null!==e&&!Array.isArray(e)&&"object"==typeof e}const ar={3:"Cancel",6:"Help",8:"Backspace",9:"Tab",12:"Clear",13:"Enter",16:"Shift",17:"Control",18:"Alt",19:"Pause",20:"CapsLock",27:"Escape",28:"Convert",29:"NonConvert",30:"Accept",31:"ModeChange",32:" ",33:"PageUp",34:"PageDown",35:"End",36:"Home",37:"ArrowLeft",38:"ArrowUp",39:"ArrowRight",40:"ArrowDown",41:"Select",42:"Print",43:"Execute",44:"PrintScreen",45:"Insert",46:"Delete",48:["0",")"],49:["1","!"],50:["2","@"],51:["3","#"],52:["4","$"],53:["5","%"],54:["6","^"],55:["7","&"],56:["8","*"],57:["9","("],91:"OS",93:"ContextMenu",144:"NumLock",145:"ScrollLock",181:"VolumeMute",182:"VolumeDown",183:"VolumeUp",186:[";",":"],187:["=","+"],188:[",","<"],189:["-","_"],190:[".",">"],191:["/","?"],192:["`","~"],219:["[","{"],220:["\\","|"],221:["]","}"],222:["'",'"'],224:"Meta",225:"AltGraph",246:"Attn",247:"CrSel",248:"ExSel",249:"EraseEof",250:"Play",251:"ZoomOut"};for(let ka=0;ka<24;ka+=1)ar[112+ka]="F"+(ka+1);for(let ka=0;ka<26;ka+=1){let e=ka+65;ar[e]=[String.fromCharCode(e+32),String.fromCharCode(e)]}const sr={codes:ar,getCode(e){return lr(e)?e.keyCode||e.which||this[e.key]:this[e]},getKey(e){const t=lr(e);if(t&&e.key)return e.key;let r=ar[t?e.keyCode||e.which:e];return Array.isArray(r)&&(r=t?r[e.shiftKey?1:0]:r[0]),r},Cancel:3,Help:6,Backspace:8,Tab:9,Clear:12,Enter:13,Shift:16,Control:17,Alt:18,Pause:19,CapsLock:20,Escape:27,Convert:28,NonConvert:29,Accept:30,ModeChange:31," ":32,PageUp:33,PageDown:34,End:35,Home:36,ArrowLeft:37,ArrowUp:38,ArrowRight:39,ArrowDown:40,Select:41,Print:42,Execute:43,PrintScreen:44,Insert:45,Delete:46,0:48,")":48,1:49,"!":49,2:50,"@":50,3:51,"#":51,4:52,$:52,5:53,"%":53,6:54,"^":54,7:55,"&":55,8:56,"*":56,9:57,"(":57,a:65,A:65,b:66,B:66,c:67,C:67,d:68,D:68,e:69,E:69,f:70,F:70,g:71,G:71,h:72,H:72,i:73,I:73,j:74,J:74,k:75,K:75,l:76,L:76,m:77,M:77,n:78,N:78,o:79,O:79,p:80,P:80,q:81,Q:81,r:82,R:82,s:83,S:83,t:84,T:84,u:85,U:85,v:86,V:86,w:87,W:87,x:88,X:88,y:89,Y:89,z:90,Z:90,OS:91,ContextMenu:93,F1:112,F2:113,F3:114,F4:115,F5:116,F6:117,F7:118,F8:119,F9:120,F10:121,F11:122,F12:123,F13:124,F14:125,F15:126,F16:127,F17:128,F18:129,F19:130,F20:131,F21:132,F22:133,F23:134,F24:135,NumLock:144,ScrollLock:145,VolumeMute:181,VolumeDown:182,VolumeUp:183,";":186,":":186,"=":187,"+":187,",":188,"<":188,"-":189,_:189,".":190,">":190,"/":191,"?":191,"`":192,"~":192,"[":219,"{":219,"\\":220,"|":220,"]":221,"}":221,"'":222,'"':222,Meta:224,AltGraph:225,Attn:246,CrSel:247,ExSel:248,EraseEof:249,Play:250,ZoomOut:251};sr.Spacebar=sr[" "],sr.Digit0=sr[0],sr.Digit1=sr[1],sr.Digit2=sr[2],sr.Digit3=sr[3],sr.Digit4=sr[4],sr.Digit5=sr[5],sr.Digit6=sr[6],sr.Digit7=sr[7],sr.Digit8=sr[8],sr.Digit9=sr[9],sr.Tilde=sr["~"],sr.GraveAccent=sr["`"],sr.ExclamationPoint=sr["!"],sr.AtSign=sr["@"],sr.PoundSign=sr["#"],sr.PercentSign=sr["%"],sr.Caret=sr["^"],sr.Ampersand=sr["&"],sr.PlusSign=sr["+"],sr.MinusSign=sr["-"],sr.EqualsSign=sr["="],sr.DivisionSign=sr["/"],sr.MultiplicationSign=sr["*"],sr.Comma=sr[","],sr.Decimal=sr["."],sr.Colon=sr[":"],sr.Semicolon=sr[";"],sr.Pipe=sr["|"],sr.BackSlash=sr["\\"],sr.QuestionMark=sr["?"],sr.SingleQuote=sr["'"],sr.DoubleQuote=sr['"'],sr.LeftCurlyBrace=sr["{"],sr.RightCurlyBrace=sr["}"],sr.LeftParenthesis=sr["("],sr.RightParenthesis=sr[")"],sr.LeftAngleBracket=sr["<"],sr.RightAngleBracket=sr[">"],sr.LeftSquareBracket=sr["["],sr.RightSquareBracket=sr["]"];var cr,dr;function ur(){return(ur=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function pr(){return(pr=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}const fr=u.css`
  width: ${({dimension:e})=>{switch(e){case"m":return"20px";case"s":return"16px";default:return"20px"}}};

  fieldset[data-dimension='s'] & {
    width: 16px;
  }
`,hr=u.css`
  height: ${({dimension:e})=>{switch(e){case"m":return"20px";case"s":return"16px";default:return"20px"}}};
  fieldset[data-dimension='s'] & {
    height: 16px;
  }
`,mr=S.default((function(e){return D.createElement("svg",ur({viewBox:"0 0 16 16",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),cr||(cr=D.createElement("path",{d:"M13.471 5.805a.667.667 0 10-.942-.943L7 10.39 4.138 7.53a.667.667 0 00-.943.942l3.334 3.334c.26.26.682.26.942 0l6-6z",fill:"#fff"})))}))`
  pointer-events: none;
  & > * {
    fill: ${({theme:e})=>e.color.text.staticWhite};
  }
`,br=S.default((function(e){return D.createElement("svg",pr({viewBox:"-5 -9 20 20",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),dr||(dr=D.createElement("rect",{width:10,height:2,rx:1,fill:"#fff"})))}))`
  pointer-events: none;
  & > * {
    fill: ${({theme:e})=>e.color.text.staticWhite};
  }
`,gr=S.default.div`
  position: relative;
  box-sizing: border-box;

  overflow: visible;
  ${fr};
  ${hr};
`,xr=S.default.div`
  pointer-events: none;
  display: inline-block;
  position: absolute;
  margin: 0;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 4px;
  outline: 0;
  transition: background 0.1s ease-in;

  background-color: ${({theme:e})=>e.color.background.primary};
  border: 1px solid ${({error:e,theme:t})=>e?t.color.status.danger:t.color.basic.tertiary};
`,vr=S.default.input`
  appearance: none;
  ::-ms-check {
    display: none;
  }
  width: calc(100% + 2px);
  height: calc(100% + 2px);

  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

  box-sizing: border-box;
  cursor: pointer;
  border-radius: 5px;
  margin: 0;
  padding: 0;

  :disabled {
    pointer-events: none;
  }

  &:checked + ${xr} {
    background-color: ${({theme:e})=>e.color.basic.primary};
    border: none;
  }
  &:not(:checked) + ${xr} {
    > * {
      display: none;
    }
  }

  &:not(:checked):focus + ${xr}, &:not(:disabled):not(:checked):hover + ${xr} {
    background-color: ${({theme:e})=>e.color.background.secondary};
    border: 1px solid ${({theme:e})=>e.color.basic.hover};
  }

  &:checked:focus + ${xr}, &:not(:disabled):checked:hover + ${xr} {
    background-color: ${({theme:e})=>e.color.basic.hover};
    border: none;
  }

  &:not(:checked):disabled + ${xr} {
    background-color: ${({theme:e})=>e.color.background.primary};
    border: 1px solid ${({theme:e})=>e.color.text.tertiary};
  }

  &:disabled + ${xr} {
    background-color: ${({theme:e})=>e.color.text.tertiary};
    border: none;
  }
`,yr=D.forwardRef(((e,t)=>{var r=e,{className:n,dimension:o="m"}=r,i=c(r,["className","dimension"]);return D.createElement(gr,{dimension:o,className:n},D.createElement(vr,s(a({ref:t},i),{type:"checkbox"})),D.createElement(xr,{error:i.error},i.indeterminate?D.createElement(br,{"aria-hidden":"true",focusable:"false"}):D.createElement(mr,{"aria-hidden":"true",focusable:"false"})))}));function $r(...e){return t=>{e.forEach((e=>{e&&("function"==typeof e?e(t):e.current=t)}))}}function wr(e,t){var r;const{value:n=e.value,selectionStart:o,selectionEnd:i}=t;if(a={value:n,selectionStart:o,selectionEnd:i},(l=e).value!==a.value||l.selectionStart!==a.selectionStart||l.selectionEnd!==a.selectionEnd){const t=null==(r=Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype,"value"))?void 0:r.set;let l;null==t||t.call(e,n),e.setSelectionRange(o||0,i||0),"function"==typeof Event?l=new Event("input",{bubbles:!0}):(l=document.createEvent("Event"),l.initEvent("input",!0,!0)),e.dispatchEvent(l)}var l,a}var Er,kr;function Cr(){return(Cr=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Dr(){return(Dr=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}const Sr=S.default((function(e){return D.createElement("svg",Cr({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Er||(Er=D.createElement("path",{fillRule:"evenodd",clipRule:"evenodd",d:"M3 12a9 9 0 1118 0 9 9 0 01-18 0zm9-4.75a.75.75 0 01.75.75v5a.75.75 0 01-1.5 0V8a.75.75 0 01.75-.75zm0 9.45a.85.85 0 100-1.7.85.85 0 000 1.7z"})))}))`
  fill: ${e=>e.theme.color.status.danger};
`,Fr=S.default((function(e){return D.createElement("svg",Dr({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),kr||(kr=D.createElement("path",{fillRule:"evenodd",clipRule:"evenodd",d:"M3 12a9 9 0 1118 0 9 9 0 01-18 0zm7.79 3.29c.28.28.74.27 1.01-.02l4.42-4.8c.24-.27.22-.68-.04-.92a.65.65 0 00-.92.04l-3.99 4.34-2.4-2.4a.658.658 0 00-.92 0c-.25.25-.25.67 0 .92l2.84 2.84z"})))}))`
  fill: ${e=>e.theme.color.status.success};
`;function Mr(e){var t=e,{status:r}=t,n=c(t,["status"]);switch(r){case"error":return D.createElement(Sr,a({},n));default:return D.createElement(Fr,a({},n))}}var _r,Rr,Ar,Or;function Br(){return(Br=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Tr(e){return D.createElement("svg",Br({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),_r||(_r=D.createElement("path",{d:"M6.41 5.49a.65.65 0 10-.92.92L11.08 12l-5.49 5.49a.65.65 0 00.92.92L12 12.92l5.49 5.49a.65.65 0 00.92-.92L12.92 12l5.59-5.59a.65.65 0 00-.92-.92L12 11.08 6.41 5.49z"})))}function zr(){return(zr=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Ir(){return(Ir=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}const Lr=S.default((function(e){return D.createElement("svg",zr({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Rr||(Rr=D.createElement("path",{d:"M21.7 10.046c.3.198.381.601.182.9-.5.754-1.141 1.573-1.913 2.359l1.835 1.835a.65.65 0 11-.919.92l-1.867-1.868c-.935.8-2.015 1.516-3.226 2.023l1.034 2.49a.65.65 0 11-1.2.498l-1.064-2.56a9.581 9.581 0 01-2.562.349 9.819 9.819 0 01-2.566-.338l-1.06 2.55a.65.65 0 01-1.2-.5l1.027-2.47c-1.244-.517-2.316-1.25-3.221-2.04L3.115 16.06a.65.65 0 01-.92-.919l1.844-1.843a19.143 19.143 0 01-1.903-2.338.65.65 0 111.066-.744c1.463 2.097 4.34 5.477 8.798 5.477 4.034 0 7.15-2.979 8.8-5.465a.65.65 0 01.9-.181z"})))}))`
  > * {
    fill: ${e=>e.theme.color.basic.tertiary};
  }

  &:hover {
    cursor: pointer;
  }

  &:hover > * {
    fill: ${e=>e.theme.color.basic.hover};
  }
`,Pr=S.default((function(e){return D.createElement("svg",Ir({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Ar||(Ar=D.createElement("path",{d:"M11.306 9.08c.376-.088.694.233.694.62v1a1.3 1.3 0 001.3 1.3h1c.387 0 .708.318.62.694a3.001 3.001 0 11-3.614-3.613z"})),Or||(Or=D.createElement("path",{fillRule:"evenodd",clipRule:"evenodd",d:"M12 5c-4.88 0-8.134 3.655-9.563 5.682a2.267 2.267 0 000 2.636C3.867 15.345 7.119 19 12 19c4.88 0 8.134-3.655 9.563-5.682a2.267 2.267 0 000-2.636C20.133 8.655 16.881 5 12 5zM5.709 8.927a15.062 15.062 0 00-2.21 2.505.967.967 0 000 1.136c.522.74 1.263 1.654 2.21 2.505A6.972 6.972 0 015 12c0-1.102.255-2.145.709-3.073zM12 17.7a5.7 5.7 0 110-11.4 5.7 5.7 0 010 11.4zm6.291-8.773a15.06 15.06 0 012.21 2.505.967.967 0 010 1.136 15.06 15.06 0 01-2.21 2.505A6.97 6.97 0 0019 12a6.972 6.972 0 00-.709-3.073z"})))}))`
  > * {
    fill: ${e=>e.theme.color.basic.tertiary};
  }

  &:hover {
    cursor: pointer;
  }

  &:hover > * {
    fill: ${e=>e.theme.color.basic.hover};
  }
`,Nr=S.default(Tr)`
  > * {
    fill: ${e=>e.theme.color.basic.tertiary};
  }

  &:hover {
    cursor: pointer;
  }

  &:hover > * {
    fill: ${e=>e.theme.color.basic.hover};
  }
`,Vr=u.css`
  height: ${({dimension:e})=>{switch(e){case"xl":return"56px";case"s":return"32px";default:return"40px"}}};
`,jr=e=>{switch(e.dimension){case"xl":return 24;case"s":return 20;default:return 24}},Hr=e=>{switch(e.dimension){case"xl":return 16;case"s":return 12;default:return 16}},Wr=u.css`
  padding-right: ${e=>{var t;return Hr(e)+(jr(e)+8)*(null!=(t=e.iconCount)?t:0)-8*(e.iconCount?1:0)}}px;
`,Ur=u.css`
  background-color: ${e=>e.theme.color.background.tertiary};
  border-color: ${e=>e.theme.color.background.tertiary};
`,Yr=u.css`
  background-color: ${e=>e.theme.color.background.primary};
  border: 1px solid ${e=>e.theme.color.basic.tertiary};
  border-radius: 4px;

  ${({disabled:e})=>e?Ur:""}
  &[data-focus-within] {
    border-color: ${e=>e.theme.color.basic.press};
    outline: -webkit-focus-ring-color auto 1px;
  }

  &:hover {
    border-color: ${e=>e.theme.color.basic.hover};
  }
`,qr=u.css`
  ::-ms-clear,
  ::-ms-reveal {
    display: none;
  }
`,Gr=S.default.input`
  outline: none;
  appearance: none;

  box-sizing: border-box;
  flex: 1 1 auto;
  min-width: 10px;
  border: none;
  background: transparent;
  text-overflow: ellipsis;
  padding: 0 ${Hr}px;

  color: ${e=>e.theme.color.text.primary};

  ${e=>"s"===e.dimension?W["Additional/S"]:W["Additional/L"]}
  &::placeholder {
    color: ${e=>e.theme.color.text.secondary};
  }

  &:disabled::placeholder {
    color: ${e=>e.theme.color.text.tertiary};
  }

  [data-read-only] & {
    user-select: none;
    pointer-events: none;
    ${Ur}
  }

  ${Yr}
  ${Wr}
  ${qr}
  [data-status='error'] &,
  &:invalid {
    border-color: ${e=>e.theme.color.status.danger};
  }

  [data-status='success'] & {
    border-color: ${e=>e.theme.color.status.success};
  }
`,Kr=S.default.div`
  position: relative;
  display: flex;
  align-items: stretch;
  border: none;

  ${Vr}
  &[data-read-only] {
    user-select: none;
    pointer-events: none;
  }
`,Xr=S.default.div`
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;

  display: flex;
  align-items: center;

  padding-right: ${Hr}px;

  & > * {
    display: block;
    width: ${jr}px;
  }

  & > *:not(:first-child) {
    margin-left: 8px;
  }
`;function Qr(e){return e}const Zr=e=>e.preventDefault(),Jr=D.forwardRef(((e,t)=>{var r=e,{type:n,value:o,displayStatusIcon:i,displayClearIcon:l,status:u,handleInput:p=Qr,containerRef:f,icons:h,children:m,className:b}=r,g=c(r,["type","value","displayStatusIcon","displayClearIcon","status","handleInput","containerRef","icons","children","className"]);const x=d.useRef(null),v=D.Children.toArray(h),[y,$]=D.useState(!1);if("password"===n){const e=y?Pr:Lr;v.push(D.createElement(e,{key:"eye-icon","aria-hidden":!0,onClick:()=>{$(!y)}}))}i&&v.push(D.createElement(Mr,{key:"status-icon",status:u,"aria-hidden":!0})),l&&v.unshift(D.createElement(Nr,{key:"clear-icon",onClick:()=>{x.current&&wr(x.current,{value:""})},"aria-hidden":!0}));const w=v.length,E=o?p({value:String(o)}):{};return d.useLayoutEffect((()=>{function e(){const{value:e,selectionStart:t,selectionEnd:r}=this;wr(this,p({value:e,selectionStart:t,selectionEnd:r}))}if(x.current){const t=x.current;t.addEventListener("input",e);const{value:r,selectionStart:n,selectionEnd:o}=t,i=p({value:r,selectionStart:n,selectionEnd:o});return wr(t,i),()=>{t.removeEventListener("input",e)}}}),[x.current,p]),D.createElement(Kr,a({className:b,disabled:g.disabled,dimension:g.dimension,ref:f,"data-read-only":g.readOnly},g.disableCopying&&{onMouseDown:Zr}),D.createElement(Gr,s(a({ref:$r(t,x)},g),{iconCount:w,value:E.value,type:"password"===n&&y?"text":n})),w>0&&D.createElement(Xr,{disabled:g.disabled,dimension:g.dimension},v),m)}));var en;function tn(){return(tn=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}Jr.defaultProps={dimension:"m"},Jr.displayName="TextInput";const rn=["bottom","height","left","right","top","width"],nn=new Map;let on;const ln=()=>{const e=[];nn.forEach(((t,r)=>{const{bottom:n,height:o,left:i,right:l,top:a,width:s,x:c,y:d}=r.getBoundingClientRect(),u={bottom:n,height:o,left:i,right:l,top:a,width:s,x:c||i,y:d||a};((e={},t={})=>rn.some((r=>e[r]!==t[r])))(u,t.rect)&&(t.rect=u,e.push(t))})),e.forEach((e=>{e.callbacks.forEach((t=>t(e.rect)))})),on=window.requestAnimationFrame(ln)};function an(e,t){return{observe(){var r;const n=0===nn.size;nn.has(e)?null==(r=nn.get(e))||r.callbacks.push(t):nn.set(e,{rect:void 0,hasRectChanged:!1,callbacks:[t]}),n&&ln()},unobserve(){const r=nn.get(e);if(r){const n=r.callbacks.indexOf(t);n>=0&&r.callbacks.splice(n,1),r.callbacks.length||nn.delete(e),nn.size||cancelAnimationFrame(on)}}}}function sn(e,t,...r){const n=d.useRef();d.useEffect((()=>{n.current=e}),[e]),d.useEffect((()=>{if(null!=t){const e=setInterval((function(){n.current&&n.current(...r)}),t);return()=>clearInterval(e)}}),[t,...r])}function cn(e,t,r,n){return e.slice(0,t)+n+e.slice(t+Math.abs(r))}function dn(e){if(null===e)return{value:"__.__.____",selectionStart:0,selectionEnd:0};const t=e.selectionStart||0;let r=e.value||"";const n=r.length-"__.__.____".length;if("__.__.____"===r.replace(/\d/g,"_"))return e;if(n<0)return r=cn(r,t,0,"__.__.____".substr(t,-n)),s(a({},e),{value:r});let o=0,i=0;"."===r.charAt(t-1)&&(o=-1,i=0),"."===r.charAt(t)&&(i=1,/\d/.test(r.charAt(t-1))?o=1:(o=-1,i=0));const l=cn(r,t+o,n,"").replace(/\./g,"");r=`${l.substr(0,2).replace(/\D/g,"_")}.${l.substr(2,2).replace(/\D/g,"_")}.${l.substr(4,4).replace(/\D/g,"_")}`;return s(a({},e),{value:r,selectionStart:t+i,selectionEnd:t+i})}function un(e){return e instanceof Date&&!isNaN(e.getTime())}function pn(e=""){const[t,r,n]=e.split("."),o=new Date(`${n}-${r}-${t}`);return un(o)?o:null}function fn(e="",t,r=" - "){const n=e.split(r).map(pn);return t?n:n[0]}const hn="__.__.____ - __.__.____";function mn(e){if(!e||!e.value)return{value:hn,selectionStart:0,selectionEnd:0};const t=e.value||hn;if(t.replace(/\d/g,"_")===hn)return e;const r=e.selectionStart||0,n=t.length-hn.length;if("__.__.____".length<r&&"__.__.____".length+" - ".length>r){const e=t.length<hn.length?"__.__.____".length:"__.__.____".length+" - ".length;return{value:t.substr(0,10)+" - "+t.slice(-10),selectionStart:e,selectionEnd:e}}if(r<="__.__.____".length){const r=dn(s(a({},e),{value:t.substr(0,10+n)}));return{value:r.value+" - "+t.slice(-10),selectionStart:r.selectionStart,selectionEnd:r.selectionEnd}}const o=dn({value:t.slice(-10-n),selectionStart:r-("__.__.____".length+" - ".length),selectionEnd:r-("__.__.____".length+" - ".length)});return{value:t.substr(0,10)+" - "+o.value,selectionStart:(o.selectionStart||0)+"__.__.____".length+" - ".length,selectionEnd:(o.selectionEnd||0)+"__.__.____".length+" - ".length}}const bn=S.default(wt)`
  border-radius: 4px;
  ${e=>e.theme.shadow.NonClickable}
`,gn=S.default(Jr)`
  input {
    letter-spacing: 1px;
  }
`,xn=u.css`
  bottom: 0;

  ${bn} {
    position: absolute;
    right: 0;
    top: 8px;
  }
`,vn=u.css`
  top: 0;

  ${bn} {
    position: absolute;
    right: 0;
    bottom: 8px;
  }
`,yn=S.default.div`
  position: absolute;
  left: 0;
  right: 0;
  height: 0;
  ${({position:e})=>"top"===e?vn:xn}
`,$n=S.default((function(e){return D.createElement("svg",tn({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),en||(en=D.createElement("path",{fillRule:"evenodd",clipRule:"evenodd",d:"M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2h-2V4a1 1 0 00-2 0v1H9V4a1 1 0 00-2 0v1H5a2 2 0 00-2 2v12a2 2 0 002 2zm13-10H6v7h12v-7z",fill:"#8A96A8"})))}))`
  > * {
    fill: ${e=>e.theme.color.basic.tertiary};
  }

  &:hover {
    cursor: pointer;
  }

  &:hover > * {
    fill: ${e=>e.theme.color.basic.hover};
  }
`,wn=S.default.div`
  position: fixed;
  overflow: visible;
  z-index: ${({theme:e})=>e.zIndex.dropdown};
`,En=({targetRef:e,children:t})=>{const r=d.useRef(null);return d.useEffect((()=>{const t=r.current;if(t&&e.current){const r=an(e.current,(e=>{if(e){const{x:r,y:n,height:o,width:i}=e,{style:l}=t;l.top=`${n}px`,l.left=`${r}px`,l.height=`${o}px`,l.width=`${i}px`}}));return r.observe(),()=>{r.unobserve()}}}),[e.current,r.current]),p.createPortal(D.createElement(wn,{ref:r,children:t}),document.body)};function kn(e,t=" - "){return e.map((e=>new Date(e))).map((e=>un(e)?e.toLocaleDateString("ru").replace(/[^ -~]/g,""):"__.__.____")).join(t)}function Cn(e,t,r){if(e.current&&r){const n=e.current.getBoundingClientRect(),o=window.innerHeight;"bottom"===t&&o-n.bottom<0?r("top"):"top"===t&&o-n.bottom-n.height-100>0&&r("bottom")}}const Dn=D.forwardRef(((e,t)=>{var r=e,{type:n="date",defaultIsCalendarOpen:o=!1,formatter:i=kn,parser:l=fn,minDate:u,maxDate:p,validator:f,filterDate:h,localeName:m}=r,b=c(r,["type","defaultIsCalendarOpen","formatter","parser","minDate","maxDate","validator","filterDate","localeName"]);const g={minDate:u,maxDate:p,validator:f,filterDate:h,localeName:m},x="date-range"===n,v=b.handleInput||(x?mn:dn),[y,$]=d.useState(null),w=d.useRef(null),E=d.useRef(null),k=d.useRef(null),[C,S]=d.useState(o),[F,M]=d.useState("bottom"),[_,R,A]=Array.isArray(y)?y:[void 0,void 0,y];sn(Cn,500,k,F,M);const O=d.useCallback((e=>{C&&e&&(e.focus(),M("bottom"))}),[C]);return D.createElement(gn,s(a({},b),{ref:$r(t,w),handleInput:v,icons:D.createElement($n,{onClick:()=>{var e;const t=l(null==(e=w.current)?void 0:e.value,x);$(t),S(!0)},tabIndex:0}),containerRef:E}),C&&D.createElement(En,{targetRef:E},D.createElement(yn,{ref:O,tabIndex:-1,onBlur:()=>{S(!1)},position:F},D.createElement(bn,s(a({},g),{ref:k,selected:A,startDate:_,endDate:R,onChange:e=>{if($(e),w.current){if(!e)return void wr(w.current,{value:""});const t=(Array.isArray(e)?e:[e]).map((e=>e?new Date(Date.UTC(e.getFullYear(),e.getMonth(),e.getDate())):e)).map((e=>e?e.toISOString().substring(0,10):"")),r=i(t);wr(w.current,{value:r})}},range:x})))))}));Dn.defaultProps={dimension:"m"},Dn.displayName="DateInput";const Sn=S.default.label`
  text-align: left;
  font-family: ${()=>H.fontFamily};
  font-style: normal;
  font-weight: normal;
  font-size: 14px;
  line-height: 16px;
  color: ${e=>e.theme.color.text.secondary};

  [data-disabled] & {
    color: ${e=>e.theme.color.text.tertiary};
  }

  [data-focus-within] & {
    color: ${e=>e.theme.color.basic.press};
  }
  [data-required-within] &:before {
    content: '* ';
    color: ${e=>e.theme.color.status.danger};
  }
  margin-bottom: 8px;
`,Fn=u.css`
  flex-direction: row;
  justify-content: space-between;
  align-items: center;

  > * {
    flex: 1 1 auto;
  }

  ${Sn} {
    margin: 0 8px 0 0;
  }
`,Mn=S.default.div`
  display: flex;
  ${e=>e.displayInline?Fn:"flex-direction: column;"}
`,_n=S.default.div`
  padding-top: 8px;

  text-align: left;
  font-family: ${()=>H.fontFamily};
  font-style: normal;
  font-weight: normal;
  font-size: 14px;
  line-height: 16px;
  color: ${e=>e.theme.color.text.secondary};

  [data-disabled] & {
    color: ${e=>e.theme.color.text.tertiary};
  }

  [data-status='error'] & {
    color: ${e=>e.theme.color.status.danger};
  }

  [data-status='success'] & {
    color: ${e=>e.theme.color.status.success};
  }
`,Rn=e=>{var t=e,{children:r}=t,n=c(t,["children"]);const o=D.useRef(null),[i,l]=D.useState(!1),{className:d,style:u,displayInline:p,status:f,extraText:h,label:m,required:b,disabled:g}=n,{id:x=Yt()}=n,v={className:d,style:u,displayInline:p},y={htmlFor:x,children:m,required:b};return D.useEffect((()=>{var e;document.hasFocus()&&(null==(e=o.current)?void 0:e.contains(document.activeElement))&&l(!0)}),[o.current,document.activeElement]),D.createElement(Mn,s(a({},v),{"data-status":f,ref:o,onFocusCapture:()=>l(!0),onBlurCapture:()=>l(!1),"data-focus-within":i?"":void 0,"data-required-within":b?"":void 0,"data-disabled":g?"":void 0}),D.createElement(Sn,a({},y)),D.createElement("div",null,r,h&&D.createElement(_n,null,h)))},An=D.forwardRef(((e,t)=>{const r=e,{className:n,displayInline:o,status:i,required:l,extraText:s,label:d,id:u=Yt(),disabled:p}=r,f=c(r,["className","displayInline","status","required","extraText","label","id","disabled"]),h={className:n,extraText:s,status:i,required:l,label:d,id:u,displayInline:o,disabled:p},m=a({ref:t,id:u,"aria-required":l,status:i,disabled:p},f);return D.createElement(Rn,a({},h),D.createElement(Jr,a({},m)))}));An.displayName="InputField";const On=u.css`
  width: ${({dimension:e})=>{switch(e){case"m":return"20px";case"s":return"16px";default:return"20px"}}};

  fieldset[data-dimension='s'] & {
    width: 16px;
  }
`,Bn=u.css`
  height: ${({dimension:e})=>{switch(e){case"m":return"20px";case"s":return"16px";default:return"20px"}}};

  fieldset[data-dimension='s'] & {
    height: 16px;
  }
`,Tn=S.default(yr)`
  position: absolute;
  top: 2px;
  left: 2px;
`,zn=S.default.label`
  display: inline-block;
  position: relative;
  padding-top: 2px;
  padding-bottom: 2px;

  padding-left: ${e=>"s"===e.dimension?28:32}px;

  ${e=>"s"===e.dimension?W["Additional/XS"]:W["Additional/M"]}
  fieldset[data-dimension='s'] & {
    padding-left: 28px;
    ${W["Additional/XS"]}
  }

  color: ${e=>e.disabled?e.theme.color.text.tertiary:e.theme.color.text.primary};
  cursor: ${e=>e.disabled?"default":"pointer"};

  fieldset:disabled & {
    color: ${e=>e.theme.color.text.tertiary};
    cursor: default;
  }
`,In=S.default.div`
  padding-top: 4px;

  ${e=>"s"===e.dimension?W["Caption/XS"]:W["Additional/XS"]}
  fieldset[data-dimension='s'] & {
    ${W["Caption/XS"]}
  }

  color: ${e=>e.theme.color.text.secondary};
`,Ln=D.forwardRef(((e,t)=>{var r=e,{extraText:n,className:o,children:i,dimension:l="m"}=r,s=c(r,["extraText","className","children","dimension"]);return D.createElement(zn,{className:o,dimension:l,disabled:s.disabled},D.createElement(Tn,a({dimension:l,ref:t},s)),i,n&&D.createElement(In,{dimension:l,children:n}))})),Pn=D.forwardRef(((e,t)=>{const r=e,{className:n,displayInline:o,status:i,required:l,extraText:s,label:d,id:u=Yt(),disabled:p}=r,f=c(r,["className","displayInline","status","required","extraText","label","id","disabled"]),h={className:n,extraText:s,status:i,required:l,label:d,id:u,displayInline:o,disabled:p},m=a({ref:t,id:u,"aria-required":l,status:i,disabled:p},f);return D.createElement(Rn,a({},h),D.createElement(Dn,a({},m)))}));var Nn,Vn;function jn(){return(jn=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Hn(e){return D.createElement("svg",jn({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Nn||(Nn=D.createElement("path",{d:"M5.12 8.56c0-.18.07-.35.21-.48.27-.24.68-.22.92.04l5.74 6.37 5.55-6.41a.65.65 0 01.92-.04c.26.24.28.65.04.92l-5.99 6.9c-.28.31-.76.31-1.04 0L5.29 9a.627.627 0 01-.17-.44z"})))}function Wn(){return(Wn=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Un(e){return D.createElement("svg",Wn({viewBox:"0 0 24 24",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Vn||(Vn=D.createElement("path",{d:"M5.12 15.4c0 .18.07.35.21.48.27.24.68.22.92-.04l5.74-6.37 5.55 6.41c.24.27.65.28.92.04.26-.24.28-.65.04-.92l-5.99-6.9a.698.698 0 00-1.04 0l-6.18 6.86c-.12.13-.17.28-.17.44z"})))}Pn.displayName="InputField";const Yn=(e,t,r,n)=>{let o=!1,i=r(e,t,n);for(;i;){if(i===(null==e?void 0:e.firstChild)){if(o)return;o=!0}if(i.hasAttribute("tabindex")&&!i.hasAttribute("disabled"))return void i.focus();i=r(e,i,n)}},qn=(e,t,r)=>(null==t?void 0:t.nextElementSibling)?t.nextElementSibling:r?null:null==e?void 0:e.firstChild,Gn=(e,t,r)=>(null==t?void 0:t.previousElementSibling)?t.previousElementSibling:r?null:null==e?void 0:e.lastChild,Kn=u.css`
  max-height: ${({dimension:e})=>{switch(e){case"l":return"288px";case"m":return"240px";case"s":return"192px";default:return"288px"}}};
`,Xn=S.default.ul`
  ${e=>e.theme.shadow.NonClickable}
  border-radius: 4px;
  background: ${({theme:e})=>e.color.background.primary};
  ${Kn}
  overflow-x: hidden;
  overflow-y: auto;
  cursor: pointer;
  position: absolute;
  padding: 8px 0;
  width: 300px;
  outline: none;
  margin: 0;
`,Qn=D.forwardRef(((e,t)=>{var r=e,{children:n,dimension:o="l"}=r,i=c(r,["children","dimension"]);const l=d.useRef(null),s=d.useCallback((e=>{var t;const r=(l.current&&l.current.ownerDocument||document).activeElement,n=sr.getCode(e);n===sr.ArrowDown?(Yn(l.current,r,qn),e.preventDefault()):n===sr.ArrowUp?(Yn(l.current,r,Gn),e.preventDefault()):32===e.keyCode?e.preventDefault():n===sr.Home?(e.preventDefault(),Yn(l.current,null,qn)):n===sr.End&&(e.preventDefault(),Yn(l.current,null,Gn)),null==(t=null==i?void 0:i.onKeyDown)||t.call(i,e)}),[l.current,Gn,qn]);return d.useLayoutEffect((()=>{var e;l.current!==document.activeElement&&(null==(e=null==l?void 0:l.current)||e.focus());const t=(l.current&&l.current.ownerDocument||document).activeElement;Yn(l.current,t,qn)}),[l,qn,Yn]),d.useEffect((()=>{var e;return null==(e=l.current)||e.addEventListener("keydown",s),()=>{var e;null==(e=l.current)||e.removeEventListener("keydown",s)}}),[l.current]),D.createElement(Xn,a({ref:$r(t,l),dimension:o},i),n)})),Zn=S.default.div`
  position: fixed;
  overflow: visible;
  z-index: ${({theme:e})=>e.zIndex.dropdown};
`,Jn=({targetRef:e,children:t})=>{const r=d.useRef(null);return d.useEffect((()=>{const t=r.current;if(t&&e.current){const r=an(e.current,(e=>{if(e){const{x:r,y:n,height:o,width:i}=e,{style:l}=t;l.top=`${n}px`,l.left=`${r}px`,l.height=`${o}px`,l.width=`${i}px`}}));return r.observe(),()=>{r.unobserve()}}}),[e.current,r.current]),F.default.createPortal(C.default.createElement(Zn,{ref:r,children:t}),document.body)},eo=u.css`
  height: ${({dimension:e})=>{switch(e){case"xl":return"56px";case"m":return"40px";case"s":return"32px";default:return"56px"}}};
`,to=S.default.div`
  position: absolute;
  left: 0;
  right: 0;
  height: 0;
  ${({position:e})=>"top"===e?ro:no}
`,ro=u.css`
  top: 0;

  & ul {
    position: absolute;
    right: 0;
    bottom: 8px;
  }
`,no=u.css`
  bottom: 0;

  & ul {
    position: absolute;
    right: 0;
    top: 8px;
  }
`,oo=u.css`
  > * {
    fill: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.basic.tertiary};
  }
`,io=u.css`
  font-size: ${({dimension:e})=>"s"===e?14:16}px;
  line-height: ${({dimension:e})=>"s"===e?20:24}px;
  font-family: ${H.fontFamily};
`,lo=u.css`
  > * {
    fill: ${({focused:e,theme:t,disabled:r})=>e?t.color.basic.press:r?t.color.text.tertiary:t.color.basic.tertiary};
  }

  &:hover {
    cursor: ${({disabled:e})=>e?"default":"pointer"};
  }

  &:hover > * {
    fill: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.basic.hover};
  }
`,ao=S.default.div`
  position: relative;
  color: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.text.primary};
  display: flex;
  user-select: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  border: 0;
  align-items: center;
  justify-content: space-between;
  ${io}
  box-sizing: border-box;
  border-radius: 4px;
  height: inherit;
  cursor: ${({disabled:e})=>e?"default":"pointer"};
  ${oo}
`,so=S.default.div`
  color: ${({theme:e,focused:t,disabled:r})=>t?e.color.basic.disable:r?e.color.text.tertiary:e.color.text.secondary};
  pointer-events: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  ${io}
`,co=S.default(Qn)`
  width: 100%;
  position: relative;
  ${oo}
`,uo=S.default.div`
  width: ${({dimension:e})=>"s"===e?20:24}px;
  display: flex;
  align-items: center;
  ${lo}
`,po=S.default.div`
  box-sizing: border-box;
  padding: 0 ${({dimension:e})=>"s"===e?12:16}px;
  cursor: ${({disabled:e})=>e?"default":"pointer"};
  background: ${({theme:e,disabled:t})=>t?e.color.background.tertiary:e.color.background.primary};
  position: relative;
  border-radius: 4px;
  ${eo}

  border-color: ${({theme:e,disabled:t,focused:r})=>r?e.color.basic.press:t?"transparent":e.color.basic.tertiary};
  border-width: 1px;
  border-style: solid;

  &:hover {
    border-color: ${({theme:e,disabled:t})=>t?"transparent":e.color.basic.hover};
  }

  &:focus {
    outline: none;
  }

  &:active {
    outline: none;
  }
`,fo=S.default.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`,ho=D.forwardRef(((e,t)=>{var r=e,{children:n,renderInputValue:o,className:i,disabled:l=!1,dimension:u="m",value:p,onChange:f,onOpen:h,onClose:m}=r,b=c(r,["children","renderInputValue","className","disabled","dimension","value","onChange","onOpen","onClose"]);const[g,x]=d.useState(!1),[v,y]=d.useState(!1),$=d.useRef(null),[w,E]=d.useState("bottom"),[k,C]=d.useState(null),S=d.useCallback((e=>{C(e)}),[g]),F=e=>{var t;const r=sr.getCode(e);e.preventDefault();const n=null!=(t=(null==e?void 0:e.currentTarget).getAttribute("value"))?t:"";r===sr.Enter||r===sr[" "]?(x(!1),null==m||m(),f&&f(n),$.current.focus()):r!==sr.Escape&&r!==sr.Tab||(null==m||m(),x(!1),$.current.focus())},M=(e,t)=>{var r;const n=null!=(r=null==e?void 0:e.currentTarget.getAttribute("value"))?r:"";!t&&f&&f(n),y(!0)};let _=!1;const R=()=>{x(!1),y(!1),g&&(null==m||m())},A=(e,t)=>{if(e&&t.current&&E){const t=e.getBoundingClientRect(),r=window.innerHeight;"bottom"===w&&r-t.bottom<0?E("top"):"top"===w&&r-t.bottom-t.height-100>0&&E("bottom")}};d.useLayoutEffect((()=>{g&&A(k,$)}),[g,k,$.current,A]),sn(A,500,k,$),d.useEffect((()=>{const e=e=>{var t;(null==(t=null==$?void 0:$.current)?void 0:t.contains(e.target))||R()};return document.addEventListener("click",e),()=>{document.removeEventListener("click",e)}}),[null==$?void 0:$.current,R]);const O=g?D.createElement(Un,null):D.createElement(Hn,null);return D.createElement(po,s(a({ref:$r(t,$),className:i,disabled:l,dimension:u,focused:v,role:"button"},b),{onKeyDown:e=>{var t;const r=sr.getCode(e);r!==sr[" "]||l||g?r===sr.Tab&&(g||y(!1)):(x(!0),y(!0),e.preventDefault(),g?null==m||m():null==h||h()),null==(t=b.onKeyDown)||t.call(b,e)},tabIndex:l?-1:0,onClick:e=>{var t;l||(x(!g),y(!0),g?null==m||m():null==h||h()),null==(t=b.onClick)||t.call(b,e)}}),D.createElement(ao,{focused:v,dimension:u,disabled:l},b.placeholder&&!p?D.createElement(so,{disabled:l,focused:g,dimension:u},b.placeholder):null!=(B=null==o?void 0:o())?B:D.createElement(fo,null,p),D.createElement(uo,{onClick:()=>{l||(x(!g),y(!0))},disabled:l,focused:v,dimension:u},O)),g&&D.createElement(Jn,{targetRef:$},D.createElement(to,{position:w},D.createElement(co,{ref:S,disabled:l,role:"listbox",dimension:"xl"===u?"l":u,tabIndex:-1},d.Children.map(n,(e=>{var t,r,n;return d.isValidElement(e)?(_=null==(r=null==(t=e.props.children)?void 0:t.some)?void 0:r.call(t,(e=>Array.isArray(e))),_?e.props.children.flat().map((e=>{var t,r;return d.cloneElement(e,a({disabled:l||(null==(t=null==e?void 0:e.props)?void 0:t.disabled),onClick:t=>M(t,e.props.disabled),onKeyDown:F,dimension:"xl"===u?"l":u,selected:(null==p?void 0:p.toString())===(null==(r=e.props.value)?void 0:r.toString())},e.props))})):d.cloneElement(e,a({disabled:l||e.props.disabled,onClick:t=>M(t,e.props.disabled),onKeyDown:F,dimension:"xl"===u?"l":u,selected:(null==p?void 0:p.toString())===(null==(n=e.props.value)?void 0:n.toString())},e.props))):null}))))));var B})),mo=S.default(ho)`
  [data-status='error'] & {
    border-color: ${e=>e.theme.color.status.danger};
  }
`,bo=D.forwardRef(((e,t)=>{const r=e,{className:n,status:o,extraText:i,label:l,id:s=Yt()}=r,d=c(r,["className","status","extraText","label","id"]),u={className:n,extraText:i,status:o,label:l,id:s},p=a({ref:t,id:s},d);return D.createElement(Rn,a({},u),D.createElement(mo,a({},p)))}));bo.displayName="SelectField";const go=S.default.legend`
  ${e=>"s"===e.dimension?W["Additional/S"]:W["Additional/L"]}
  color: ${e=>e.theme.color.text.primary};
  fieldset:disabled & {
    color: ${e=>e.theme.color.text.tertiary};
    cursor: default;
  }
`,xo=S.default.fieldset`
  border: none;
  padding: 0;

  display: flex;
  flex-direction: column;

  > * {
    margin-top: ${e=>"s"===e.dimension?12:16}px;
  }
  ${go} {
    margin-top: 0;
  }
`;u.css`
  width: ${({dimension:e})=>{switch(e){case"m":return"20px";case"s":return"16px";default:return"20px"}}};
`,u.css`
  height: ${({dimension:e})=>{switch(e){case"m":return"20px";case"s":return"16px";default:return"20px"}}};
`;const vo=u.css`
  font-size: ${({dimension:e})=>"m"===e?16:14}px;
  line-height: ${({dimension:e})=>"m"===e?20:16}px;
  font-weight: normal;
  font-family: ${H.fontFamily};
  color: ${({disabled:e,theme:t})=>e?t.color.text.tertiary:t.color.basic.secondary};
  user-select: none;
`;u.css`
  &::before {
    content: '';
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    border: ${({theme:e})=>`2px solid ${e.color.basic.hover}`};
    width: ${({dimension:e})=>"m"===e?22:18}px;
    height: ${({dimension:e})=>"m"===e?22:18}px;
    border-radius: 4px;
  }
`,u.css`
  background: ${({theme:e,checked:t})=>t?e.color.basic.hover:e.color.background.secondary};
`,u.css`
  border: ${({theme:e,disabled:t})=>`1px solid ${t?e.color.basic.disable:e.color.basic.hover}`};
`,u.css`
  background: ${({theme:e})=>e.color.basic.hover};
`,u.css`
  background-color: ${({checked:e,theme:t})=>e?t.color.basic.primary:t.color.background.primary};
`,u.css`
  border: ${({theme:e,disabled:t})=>`1px solid ${t?e.color.basic.disable:e.color.basic.tertiary}`};
`,u.css`
  background: ${({theme:e})=>e.color.basic.disable};
`;const yo=S.default.div`
  display: flex;
  flex-direction: column;
  display: inline-flex;
  box-sizing: border-box;
  ${vo};

  & > div {
    margin-top: ${({dimension:e})=>"m"===e?16:12}px;
  }
`,$o=S.default.div`
  ${({dimension:e})=>u.css`
    position: relative;
    width: fit-content;
    display: flex;
    flex-flow: column;
    & > ${yo} {
      padding-left: ${"m"===e?30:25}px;
    }
    ${vo}
  `}
`,wo=u.css`
  height: ${({dimension:e})=>{switch(e){case"m":return"28px";case"s":return"24px";default:return"28px"}}};
`,Eo=u.css`
  height: ${({dimension:e})=>{switch(e){case"m":return"18px";case"s":return"12px";default:return"16px"}}};
`,ko=u.css`
  width: ${({dimension:e})=>{switch(e){case"m":return"18px";case"s":return"12px";default:return"16px"}}};
`,Co=u.css`
  padding: ${({dimension:e})=>{switch(e){case"m":return"6px 12px";case"s":return"4px 8px";default:return"6px 12px"}}};
`,Do=u.css`
  font-family: ${H.fontFamily};
  font-style: normal;
  font-weight: normal;
  font-size: ${({dimension:e})=>"s"===e?"12px":"14px"};
  line-height: ${({dimension:e})=>"s"===e?"16px":"20px"};
  font-feature-settings: 'tnum' on, 'lnum' on;
  color: ${({theme:e,appearance:t,disabled:r,selected:n})=>r&&!n?e.color.text.tertiary:n||n&&r?e.color.text.staticWhite:"filled"===t?e.color.text.primary:e.color.basic.primary};
  &:hover {
    color: ${({theme:e,appearance:t,selected:r})=>r?e.color.text.staticWhite:"filled"===t?e.color.text.primary:void e.color.basic.hover}
`,So=u.css`
  background-color: ${({theme:e,appearance:t,selected:r,disabled:n})=>r&&!n?e.color.basic.primary:r&&n?e.color.basic.disable:"filled"===t?e.color.background.tertiary:e.color.background.primary};

  border: 1px solid
    ${({theme:e,appearance:t,disabled:r})=>r?e.color.background.tertiary:"filled"===t?"transparent":e.color.basic.primary};

  border-radius: 16px;
  &:hover {
    background-color: ${({theme:e,appearance:t,selected:r})=>r?e.color.basic.hover:"filled"===t?e.color.background.tertiaryHover:e.color.background.secondary};
  }

  &:active {
    color: ${({theme:e,appearance:t})=>"filled"===t?e.color.text.primary:e.color.basic.press};
    background-color: ${({theme:e,appearance:t,selected:r})=>r?e.color.basic.primary:"filled"===t?e.color.background.tertiaryHover:e.color.background.secondary};
  }

  &:focus-visible {
    outline: 0;
    &:before {
      border: 2px solid ${({theme:e})=>e.color.basic.hover};
      border-radius: 20px;
      content: '';
      display: block;
      position: absolute;
      top: -5px;
      left: -5px;
      right: -5px;
      bottom: -5px;
      pointer-events: none;
    }
  }
`,Fo=S.default.div`
  display: inline-flex;
  align-items: center;
  box-sizing: border-box;
  position: relative;
  max-width: 190px;
  user-select: none;
  pointer-events: ${({disabled:e})=>e?"none":"auto"};
  fill: ${({theme:e,appearance:t,disabled:r})=>"filled"===t||r?e.color.basic.tertiary:e.color.basic.primary};
  cursor: ${({defaultChip:e,disabled:t})=>e&&!t?"pointer":"default"};
  ${So}
  ${wo}
  ${Co}
  ${Do}
`,Mo=S.default(Tr)`
  margin-left: 12px;
  cursor: ${({disabled:e})=>e?"default":"pointer"};
  ${Eo}
  ${ko}

  &:hover {
    outline: none;
    cursor: pointer;
    fill: ${({theme:e})=>e.color.basic.hover};
  }
  &:active {
    outline: none;
    border: none;
    fill: ${({theme:e})=>e.color.basic.press};
  }
`,_o=S.default.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  & svg {
    ${Eo}
    ${ko}
    fill: ${({theme:e,appearance:t,disabled:r,selected:n})=>n?e.color.background.primary:"filled"===t||r?e.color.basic.tertiary:e.color.basic.primary};
  }
`,Ro=e=>{var t,r=e,{dimension:n="m",disabled:o,appearance:i="outlined",selected:l,onClose:u,children:p,renderContentTooltip:f=(()=>"")}=r,h=c(r,["dimension","disabled","appearance","selected","onClose","children","renderContentTooltip"]);const m=void 0!==l,[b,g]=d.useState(!1),x=d.useRef(null);d.useEffect((()=>{var e,t;(null==x?void 0:x.current)&&(null==(e=null==x?void 0:x.current)?void 0:e.offsetWidth)<=(null==(t=null==x?void 0:x.current)?void 0:t.scrollWidth)&&g((null==x?void 0:x.current.offsetWidth)<(null==x?void 0:x.current.scrollWidth))}),[g,x.current]);const v=C.default.createElement(Fo,s(a({dimension:n,disabled:o,appearance:i,selected:l,defaultChip:m},h),{tabIndex:null!=(t=h.tabIndex)?t:0}),C.default.createElement(_o,{dimension:n,disabled:o,appearance:i,selected:l,ref:x},p),u&&C.default.createElement(Mo,{dimension:n,disabled:o,onClick:o?void 0:()=>{o||null==u||u()}}));return b?C.default.createElement(vt,{renderContent:f},v):v},Ao=S.default.div`
  display: inline-block;
  position: relative;
  cursor: pointer;
`,Oo=S.default.button`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  cursor: pointer;
  outline: none;
  border: 0;
  margin-left: ${"16px"};
  padding: 0;
  width: ${"20px"};
  height: ${"20px"};
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  ${({theme:e})=>`\n    & svg {\n      fill: ${e.color.basic.tertiary};\n    }\n\n    &:hover,\n    &:focus {\n      & svg {\n        fill: ${e.color.basic.hover};\n      }\n    }\n\n    &:active {\n      & svg {\n        fill: ${e.color.basic.press};\n      }\n    }\n  `}
`,Bo=u.css`
  width: ${({dimension:e})=>"s"===e?"280px":"m"===e?"384px":"488px"};
`,To=S.default.div`
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  justify-content: space-between;
  z-index: ${({theme:e})=>e.zIndex.hint};
  ${({isMobile:e})=>!e&&Bo}
  ${({theme:e})=>`\n    background-color: ${e.color.background.primary};\n    color: ${e.color.text.primary};\n  `}
  font-family: ${H.fontFamily};
  font-size: ${"14px"};
  font-weight: ${400};
  line-height: ${"20px"};
  border-radius: ${"4px"};
  box-shadow: 0 4px 8px rgba(138, 150, 168, 0.4), 0 -2px 4px rgba(138, 150, 168, 0.08);
  box-sizing: border-box;
`,zo=S.default.div`
  display: flex;
  padding: ${"16px"};
`,Io={"bottom-right":{check:(e,t)=>{const r=window.innerHeight-e.bottom>8+t.height,n=window.innerWidth-e.right>=t.width-e.width;return r&&n},getStyles:e=>`\n      translate3d(${Math.floor(e.left)}px, ${Math.floor(e.bottom+8+window.pageYOffset)}px, 0px)\n    `},"bottom-left":{check:(e,t)=>{const r=window.innerHeight-e.bottom>8+t.height,n=e.left>t.width-e.width;return r&&n},getStyles:(e,t)=>`\n      translate3d(${Math.floor(e.left-(t.width-e.width))}px, ${Math.floor(e.bottom+8+window.pageYOffset)}px, 0px)\n    `},"top-right":{check:(e,t)=>{const r=e.top>8+t.height;return window.innerWidth-e.right>=t.width-e.width&&r},getStyles:(e,t)=>`\n      translate3d(${Math.floor(e.left)}px, ${Math.floor(e.top-t.height-8+window.pageYOffset)}px, 0px)\n    `},"top-left":{check:(e,t)=>{const r=e.top>8+t.height;return e.left>t.width-e.width&&r},getStyles:(e,t)=>`\n    translate3d(${Math.floor(e.left-(t.width-e.width))}px, ${Math.floor(e.top-t.height-8+window.pageYOffset)}px, 0px)\n    `}};const Lo=(e,t)=>{const r=d.useCallback((r=>{const n=r.target;e.every((e=>e&&!e.contains(n)))&&t(r)}),[t,e]);d.useEffect((()=>(document.addEventListener("mousedown",r),document.addEventListener("touchstart",r),()=>{document.removeEventListener("mousedown",r),document.removeEventListener("touchstart",r)})),[r])};function Po(e=document){return[...e.querySelectorAll('a, button, input, textarea, select, details,[tabindex]:not([tabindex="-1"])')].filter((e=>!e.hasAttribute("disabled")))}const No=D.forwardRef(((e,t)=>{var r=e,{dimension:n,isMobile:o,content:i,visibilityTrigger:l,portal:u,scrollableParents:p,anchorElementRef:f,anchorId:h,dispatch:m}=r,b=c(r,["dimension","isMobile","content","visibilityTrigger","portal","scrollableParents","anchorElementRef","anchorId","dispatch"]);const g=()=>m({type:"setInvisible"}),x="hover"===l,v=D.useRef(null),y=D.useRef(null),[$,w]=D.useState(),[E,k]=D.useState();"click"===l&&Lo([v.current,null==f?void 0:f.current],g),d.useEffect((()=>{y.current=document.activeElement}),[]),d.useLayoutEffect((()=>{if(v.current&&"click"===l){const e=Po(v.current);w(e[0]),k(e[e.length-1]),null==$||$.focus()}}),[v.current,l,i]),d.useImperativeHandle(t,(()=>v.current)),d.useLayoutEffect((()=>{const[e,t]=x?[g,()=>{}]:((e,t)=>{let r,n=!1,o=!1;return[(...i)=>{if(o)return;if(n)return;const l=e(...i);return n=!0,r=window.setTimeout((()=>{n=!1}),t),l},()=>{o=!0,clearTimeout(r)}]})((()=>{m({type:"triggerRecalculation"})}),150);return window.addEventListener("resize",e),window.addEventListener("scroll",e),p.forEach((t=>t.addEventListener("scroll",e))),()=>{window.removeEventListener("resize",e),window.removeEventListener("scroll",e),p.forEach((t=>t.removeEventListener("scroll",e))),t()}}));return F.default.createPortal(D.createElement(To,s(a({},b),{role:"tooltip",ref:v,dimension:n,isMobile:o,onKeyDown:e=>{e.stopPropagation();const t=sr.getCode(e);t===sr.Escape?(e.preventDefault(),y.current.focus(),g()):t===sr.Tab&&"click"===l&&(t===sr.Shift?document.activeElement===$&&(E.focus(),e.preventDefault()):document.activeElement===E&&($.focus(),e.preventDefault()))}}),D.createElement(zo,{role:"dialog","aria-labelledby":h},i,"click"===l&&D.createElement(Oo,{"aria-label":"Закрыть подсказку",onClick:e=>{null==e||e.stopPropagation(),y.current.focus(),g()}},D.createElement(Tr,{width:20,height:20,"aria-hidden":!0})))),u)})),Vo={recalculation:null,visible:!1,hintWidth:null},jo=(e,t)=>{switch(t.type){case"setVisible":return s(a({},e),{visible:!0});case"setInvisible":return s(a({},e),{visible:!1});case"setHintWidth":return s(a({},e),{hintWidth:t.payload});case"triggerRecalculation":return s(a({},e),{recalculation:{}});default:throw new Error}};var Ho,Wo,Uo,Yo;function qo(){return(qo=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Go(e){return D.createElement("svg",qo({viewBox:"0 0 64 64",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Ho||(Ho=D.createElement("path",{clipRule:"evenodd",d:"M58.607 49.778a32 32 0 005.252-14.782C64.016 33.346 62.657 32 61 32s-2.982 1.349-3.17 2.995C56.343 47.944 45.346 58 32 58 17.64 58 6 46.36 6 32 6 18.654 16.056 7.656 29.005 6.17 30.651 5.983 32 4.658 32 3S30.654-.015 29.004.14a32 32 0 1029.603 49.638z"})))}function Ko(){return(Ko=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Xo(e){return D.createElement("svg",Ko({viewBox:"0 0 48 48",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Wo||(Wo=D.createElement("path",{clipRule:"evenodd",d:"M43.955 37.334a24.001 24.001 0 003.915-10.838C48.014 25.122 46.88 24 45.5 24c-1.38 0-2.483 1.125-2.662 2.494C41.616 35.81 33.648 43 24 43 13.507 43 5 34.493 5 24c0-9.648 7.191-17.616 16.506-18.838C22.875 4.982 24 3.881 24 2.5c0-1.38-1.122-2.513-2.496-2.37a24 24 0 1022.451 37.204z"})))}function Qo(){return(Qo=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function Zo(e){return D.createElement("svg",Qo({viewBox:"0 0 24 24",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Uo||(Uo=D.createElement("path",{clipRule:"evenodd",d:"M21.978 18.667a12 12 0 001.928-5.17C24.01 12.673 23.328 12 22.5 12s-1.487.677-1.623 1.494C20.165 17.754 16.462 21 12 21a9 9 0 01-1.494-17.877C11.323 2.987 12 2.328 12 1.5S11.326-.01 10.504.094a12 12 0 1011.474 18.573z"})))}function Jo(){return(Jo=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function ei(e){return D.createElement("svg",Jo({viewBox:"0 0 16 16",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),Yo||(Yo=D.createElement("path",{clipRule:"evenodd",d:"M14.652 12.445a8 8 0 001.286-3.448C16.006 8.45 15.552 8 15 8c-.552 0-.991.451-1.082.996A6.002 6.002 0 012 8a6.002 6.002 0 015.004-5.918C7.55 1.992 8 1.552 8 1S7.55-.006 7.003.062a8 8 0 107.649 12.383z"})))}const ti=u.css`
  width: ${({dimension:e})=>{switch(e){case"s":return"16px";case"m":return"24px";case"xl":return"64px";case"l":default:return"48px"}}};
  height: ${({dimension:e})=>{switch(e){case"s":return"16px";case"m":return"24px";case"xl":return"64px";case"l":default:return"48px"}}};
`,ri=u.keyframes`
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
`,ni=S.default.div`
  display: flex;
  ${ti}
  animation: ${ri} 1s linear infinite;
  z-index: ${({theme:e})=>e.zIndex.spinner};
  & svg {
    path {
      fill: ${({inverse:e,theme:t})=>e?t.color.background.primary:t.color.basic.primary};
    }
  }
`;ni.defaultProps={theme:Ne};const oi=e=>{var t=e,{dimension:r="l",inverse:n=!1}=t,o=c(t,["dimension","inverse"]);const i=(e=>{switch(e){case"s":return ei;case"m":return Zo;case"xl":return Go;case"l":default:return Xo}})(r);return C.default.createElement(ni,a({inverse:n,dimension:r,role:"alert","aria-live":"assertive"},o),C.default.createElement(i,null))},ii=u.css`
  ${({dimension:e})=>"s"===e?W["Additional/S"]:W["Additional/L"]}
  ul[data-dimension='s'] & {
    ${W["Additional/S"]}
  }
  text-overflow: ellipsis;
`,li=u.css`
  color: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.text.primary};
`,ai=u.css`
  background: ${({theme:e,disabled:t})=>t?e.color.background.primary:e.color.background.secondary};
  ${({id:e,disabled:t,theme:r})=>e&&`ul[data-selectedid="${e}"] && {background: ${t?r.color.background.primary:r.color.background.secondary};}`}}
`,si=u.css`
  background: ${({theme:e,selected:t})=>t?e.color.background.tertiary:e.color.background.primary};
  ${({id:e,theme:t})=>e&&`ul[data-selectedid="${e}"] && {background: ${t.color.background.tertiary};}`}}
`,ci=u.css`
  padding: ${({dimension:e})=>{switch(e){case"l":return"12px 16px";case"m":return"8px 16px";case"s":return"6px 12px";default:return"12px 16px"}}};
  ul[data-dimension='m'] && {
    padding: 8px 16px;
  }
  ul[data-dimension='s'] && {
    padding: 6px 12px;
  }
`,di=S.default.li`
  display: flex;
  align-items: center;
  user-select: none;
  flex-flow: wrap;
  position: relative;
  justify-content: space-between;
  outline: none;
  cursor: pointer;
  white-space: pre;
  margin: 0;
  ${ci}
  ${ii}
  ${li}
  ${si}
  ${({disabled:e})=>e&&"cursor: default;"};

  &:hover {
    ${ai}
  }

  &:focus {
    background: ${({theme:e,disabled:t})=>t?e.color.background.primary:e.color.background.secondary};
    ${({id:e,disabled:t,theme:r})=>e&&`ul[data-selectedid="${e}"] && {background: ${t?r.color.background.primary:r.color.background.secondary};}`}}
  }
`,ui=D.forwardRef(((e,t)=>{var r=e,{children:n,disabled:o=!1,dimension:i="l",selected:l=!1,id:s}=r,d=c(r,["children","disabled","dimension","selected","id"]);return D.createElement(di,a({ref:t,disabled:o,dimension:i,tabIndex:d.tabIndex?d.tabIndex:-1,selected:l,id:s},d),n)})),pi=u.css`
  background: ${({kind:e,theme:t})=>{switch(e){case"green":return t.color.status.success;case"blue":return t.color.basic.primary;case"red":return t.color.status.danger;case"orange":default:return t.color.status.warn}}};
`,fi=S.default.button`
  position: relative;
  box-sizing: border-box;
  height: ${"24px"};
  ${({width:e})=>e&&`width: ${"number"==typeof e?`${e}px`:e};`}
  padding: ${"4px 8px"};
  margin-top: ${"8px"};
  border-radius: ${"4px"};
  background: ${({theme:e})=>e.color.background.tertiary};

  display: inline-flex;
  align-items: center;
  border: none;
  cursor: ${({clickable:e})=>e?"pointer":"default"};
  ${({clickable:e})=>!e&&"pointer-events: none;"}

  &:hover,
  &:active {
    background: ${({theme:e})=>e.color.background.tertiaryHover};
  }

  &:focus {
    outline: none;
    &:before {
      content: '';
      position: absolute;
      top: ${"-4px"};
      bottom: ${"-4px"};
      left: ${"-4px"};
      right: ${"-4px"};
      border-radius: ${"6px"};
      border: ${"2px"} solid ${({theme:e})=>e.color.basic.hover};
    }
  }
`,hi=S.default.span`
  color: ${({theme:e})=>e.color.text.primary};
  user-select: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  ${W["Caption/XS"]}
`,mi=S.default.div`
  display: flex;
  flex-shrink: 0;
  width: ${"8px"};
  height: ${"8px"};
  border-radius: 50%;
  margin-right: ${"6px"};
  ${pi}
`,bi=S.default.div`
  display: flex;
  flex-flow: wrap;
  & > * {
    margin-right: ${"8px"};
  }

  & > *:last-child {
    margin-right: 0;
  }
`,gi=u.css`
  width: ${({dimension:e})=>"s"===e?"28px":"36px"};
  height: ${({dimension:e})=>"s"===e?"16px":"20px"};
`,xi=u.css`
  width: ${({dimension:e})=>"s"===e?"12px":"16px"};
  height: ${({dimension:e})=>"s"===e?"12px":"16px"};
  border-radius: ${({dimension:e})=>"s"===e?"12px":"16px"};
`,vi=S.default.input`
  position: absolute;
  width: 0;
  height: 0;
  opacity: 0;

  &:checked + span {
    &:before {
      ${({dimension:e})=>`left: calc(100% - 2px - ${"m"===e?"16px":"12px"});`}
    }

    background: ${({theme:e})=>e.color.basic.primary};
  }

  &:disabled + span {
    background: ${({theme:e})=>e.color.text.tertiary};
  }

  &:checked:not(:disabled) {
    &:hover + span,
    &:focus + span {
      background: ${({theme:e})=>e.color.basic.hover};
    }
  }

  &:not(:checked):not(:disabled) {
    &:hover + span,
    &:focus + span {
      background: ${({theme:e})=>e.color.text.secondary};
    }
  }
`,yi=S.default.div`
  display: flex;
  flex-direction: column;
  ${({position:e})=>"right"===e?"margin-left: 8px;":"margin-right: 8px;"}
  ${({dimension:e})=>"s"===e?W["Additional/XS"]:W["Additional/M"]}
  color: ${({disabled:e,theme:t})=>e?t.color.text.tertiary:t.color.text.primary};
`,$i=S.default.div`
  margin-top: 4px;
  ${({dimension:e})=>"s"===e?W["Caption/XS"]:W["Additional/XS"]}
  color: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.text.secondary};
`,wi=S.default.span`
  position: relative;
  ${gi}
  border-radius: ${"10px"};
  flex-shrink: 0;

  &:before {
    content: '';
    position: absolute;
    top: ${"2px"};
    left: ${"2px"};
    ${xi}
    background: ${({theme:e})=>e.color.background.primary};
    transition: all 0.3s;
  }

  background: ${({theme:e})=>e.color.basic.tertiary};
`,Ei=S.default.label`
  display: flex;
  flex-direction: ${({labelPosition:e})=>"right"===e?"row":"row-reverse"};
  align-items: flex-start;
  justify-content: space-between;
  position: relative;
  width: ${({width:e})=>e?"number"==typeof e?`${e}px`:e:"fit-content"};
  cursor: ${({disabled:e})=>e?"default":"pointer"};
  -webkit-tap-highlight-color: transparent;
  user-select: none;
`,ki=D.forwardRef(((e,t)=>{var r=e,{dimension:n="m",labelPosition:o="right",disabled:i=!1,width:l,extraText:s,className:d,children:u}=r,p=c(r,["dimension","labelPosition","disabled","width","extraText","className","children"]);return D.createElement(Ei,{className:d,width:"left"===o?l:void 0,labelPosition:o,disabled:i,role:"switch","aria-checked":p.checked||p["aria-checked"]},D.createElement(vi,a({ref:t,type:"checkbox",dimension:n,disabled:i},p)),D.createElement(wi,{dimension:n,disabled:i,"aria-hidden":!0}),u&&D.createElement(yi,{dimension:n,disabled:i,position:o},u,s&&D.createElement($i,{dimension:n,disabled:i},s)))})),Ci=S.default.div`
  position: fixed;
  overflow: visible;
  z-index: ${({theme:e})=>e.zIndex.dropdown};
`,Di=S.default.div`
  position: absolute;
  left: 0;
  right: 0;
  height: 0;
  ${({position:e})=>"top"===e?Fi:Mi}
`,Si=S.default.div`
  width: 100%;
`,Fi=u.css`
  top: 0;

  ${Si} {
    position: absolute;
    right: 0;
    bottom: 8px;
  }
`,Mi=u.css`
  bottom: 0;

  ${Si} {
    position: absolute;
    right: 0;
    top: 8px;
  }
`,_i=({targetRef:e,parentRef:t,children:r,open:n})=>{const o=d.useRef(null),[i,l]=d.useState("bottom"),a=d.useRef(null),s=d.useCallback((e=>{n&&e&&l("bottom")}),[n]),c=d.useCallback(((e,t,r)=>{if((null==e?void 0:e.current)&&r){const n=e.current.getBoundingClientRect(),o=window.innerHeight;"bottom"===t&&o-n.bottom<0?r("top"):"top"===t&&o-n.bottom-n.height-100>0&&r("bottom")}}),[]);return d.useEffect((()=>{var r;const n=o.current,l=(null==(r=t.current)?void 0:r.getBoundingClientRect().height)||0;if(n&&e.current){const t=an(e.current,(e=>{if(e){const{x:t,y:r,height:o,width:a}=e,{style:s}=n;s.top=`${"top"===i?r-l:r}px`,s.left=t-1+"px",s.height=`${o}px`,s.width=`${a+2}px`}}));return t.observe(),()=>{t.unobserve()}}}),[e.current,o.current,i]),sn(c,500,a,i,l),p.createPortal(C.default.createElement(Ci,{ref:o},C.default.createElement(Di,{ref:s,position:i},C.default.createElement(Si,{ref:a},r))),document.body)},Ri=u.css`
  height: ${({dimension:e})=>{switch(e){case"xl":return"54px";case"m":return"38px";case"s":return"30px";default:return"54px"}}};
`,Ai=u.css`
  font-size: ${({dimension:e})=>"s"===e?14:16}px;
  line-height: ${({dimension:e})=>"s"===e?20:24}px;
  font-family: ${H.fontFamily};
`,Oi=u.css`
  > * {
    fill: ${({focused:e,theme:t,disabled:r})=>e?t.color.basic.press:r?t.color.text.tertiary:t.color.basic.tertiary};
  }

  &:hover {
    cursor: ${({disabled:e})=>e?"default":"pointer"};
  }

  &:hover > * {
    fill: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.basic.hover};
  }
`,Bi=S.default.div`
  white-space: nowrap;
  position: relative;
  color: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.text.primary};
  display: flex;
  user-select: none;
  flex-flow: nowrap;
  cursor: pointer;
  border: 0;
  pointer-events: none;
  align-items: center;
  padding-left: 16px;
  justify-content: space-between;
  ${Ai}
  border-sizing: border-box;
  border-radius: 4px;
  height: inherit;
  cursor: ${({disabled:e})=>e?"default":"pointer"};

  > * {
    fill: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.basic.tertiary};
  }
`,Ti=S.default.div`
  color: ${({theme:e,focused:t,disabled:r})=>t?e.color.basic.disable:r?e.color.text.tertiary:e.color.text.secondary};
  pointer-events: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  ${Ai}
`,zi=S.default(Qn)`
  width: 100%;
  position: relative;

  > * {
    fill: ${({theme:e,disabled:t})=>t?e.color.text.tertiary:e.color.basic.tertiary};
  }
`,Ii=S.default.div`
  padding-right: ${({dimension:e})=>"s"===e?16:20}px;
  display: flex;
  align-items: center;
  ${Oi}
`,Li=S.default.div`
  cursor: ${({disabled:e})=>e?"default":"pointer"};
  background: ${({theme:e,disabled:t})=>t?e.color.background.tertiary:e.color.background.primary};
  position: relative;
  border-radius: 4px;
  ${Ri}
  border-color: ${({theme:e,disabled:t,focused:r})=>r?e.color.basic.press:t?"transparent":e.color.basic.tertiary};
  border-width: 1px;
  border-style: solid;

  &:hover {
    border-color: ${({theme:e,disabled:t})=>t?"transparent":e.color.basic.hover};
  }

  &:focus {
    outline: 1;
  }

  &:active {
    outline: 0;
  }
`,Pi=S.default.div`
  align-items: center;
  box-sizing: border-box;
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  overflow: hidden;
  position: relative;
  height: 32px;
  padding-right: 46px;
  pointer-events: none;
`,Ni=S.default(Ro)`
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex: 0 0 auto;
  margin: ${e=>"s"===e.dimension?"4px 8px 4px 0":"2px 8px 4px 0"};
  margin-right: ${e=>"s"===e.dimension?-34:-45}px;
`,Vi=S.default.div`
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex: 0 0 auto;
  margin: ${e=>"s"===e.dimension?"4px 8px 4px 0":"2px 8px 4px 0"};
`,ji=D.forwardRef(((e,t)=>{var r=e,{children:n,renderInputValue:o,className:i,disabled:l=!1,dimension:u="xl",value:p,onChange:f,dropDownRef:h,onOpen:m,onClose:b}=r,g=c(r,["children","renderInputValue","className","disabled","dimension","value","onChange","dropDownRef","onOpen","onClose"]);const[x,v]=d.useState(!1),[y,$]=d.useState(!1),w=d.useRef(null),E=d.useRef(null),k=d.useRef(null),C=(e,t)=>t.indexOf(e)===t.lastIndexOf(e)?t:t.filter((t=>t!==e)),S=e=>{var t;const r=sr.getCode(e);if(e.preventDefault(),r===sr.Enter||r===sr[" "]){const r=null!=(t=(null==e?void 0:e.currentTarget).getAttribute("value"))?t:"";r&&f&&f(C(r,[...p,r]))}else r!==sr.Escape&&r!==sr.Tab||(null==b||b(),v(!1),k.current.focus())},F=(e,t)=>{var r,n,o;const i=null!=(o=null==(n=(r=null==e?void 0:e.currentTarget).getAttribute)?void 0:n.call(r,"value"))?o:"";i&&!t&&f&&f(C(i,[...p,i]))};let M=!1;const _=()=>{v(!1),$(!1),x&&(null==b||b())};d.useEffect((()=>{const e=e=>{var t,r,n;(null==(t=null==E?void 0:E.current)?void 0:t.contains(e.target))||(null==(r=null==k?void 0:k.current)?void 0:r.contains(e.target))||(null==(n=null==w?void 0:w.current)?void 0:n.contains(e.target))||_()};return document.addEventListener("mousedown",e),document.addEventListener("touchstart",e),()=>{document.removeEventListener("mousedown",e),document.removeEventListener("touchstart",e)}}),[null==E?void 0:E.current,null==k?void 0:k.current,null==w?void 0:w.current,_]);const R=x?D.createElement(Un,{width:24}):D.createElement(Hn,{width:24}),A=()=>{const e="s"===u?"s":"m";return D.createElement(Pi,null,p.map(((t,r)=>D.createElement(d.Fragment,{key:r},D.createElement(Ni,{tabIndex:-1,disabled:l,dimension:e,appearance:"filled"},(e=>" + "+e)(p.length-r)),D.createElement(Vi,{dimension:u},D.createElement(Ro,{tabIndex:-1,disabled:l,dimension:e,appearance:"filled",onClose:()=>{return e=r,void f(p.filter(((t,r)=>r!==e)));var e}},t))))))};return D.createElement(Li,s(a({ref:$r(t,k),className:i,disabled:l,dimension:u,focused:y},g),{onKeyDown:e=>{var t;const r=sr.getCode(e);r!==sr[" "]||l||e.target!==e.currentTarget?r===sr.Escape?v(!1):r===sr.Tab&&(x||$(!1)):(v(!x),$(!0),e.preventDefault(),x?null==b||b():null==m||m()),null==(t=g.onKeyDown)||t.call(g,e)},onClick:e=>{var t;e.currentTarget===e.target&&(e.preventDefault(),e.stopPropagation(),l||(x?null==b||b():null==m||m(),v(!x),$(!0)),null==(t=g.onClick)||t.call(g,e))},tabIndex:l?-1:0}),D.createElement(Bi,{focused:y,dimension:u,disabled:l},g.placeholder&&0===p.length?D.createElement(Ti,{disabled:l,focused:x,dimension:u},g.placeholder):null!=(O=null==o?void 0:o())?O:A(),D.createElement(Ii,{onMouseDown:e=>{l||(x?null==b||b():null==m||m(),e.preventDefault(),v(!x),$(!0),e.stopPropagation())},disabled:l,focused:y,dimension:u},R)),x&&D.createElement("div",{ref:E,onClick:e=>e.preventDefault()},D.createElement(_i,{open:x,parentRef:k,targetRef:E},D.createElement(zi,{disabled:l,ref:$r(h,w),dimension:"xl"===u?"l":u,role:"listbox",tabIndex:-1},d.Children.map(n,(e=>{var t,r,n,o,i;return d.isValidElement(e)?(M=null==(r=null==(t=e.props.children)?void 0:t.some)?void 0:r.call(t,(e=>Array.isArray(e))),M?e.props.children.flat().map((e=>{var t,r,n,o;return d.cloneElement(e,a({disabled:l||(null==(t=null==e?void 0:e.props)?void 0:t.disabled),onMouseDown:t=>F(t,e.props.disabled),onKeyDown:S,dimension:"xl"===u?"l":u,selected:0!==p.length&&p.includes(null==(o=null==(n=null==(r=null==e?void 0:e.props)?void 0:r.value)?void 0:n.toString)?void 0:o.call(n))},e.props))})):d.cloneElement(e,a({disabled:l||e.props.disabled,onMouseDown:t=>F(t,e.props.disabled),onKeyDown:S,dimension:"xl"===u?"l":u,selected:0!==p.length&&p.includes(null==(i=null==(o=null==(n=null==e?void 0:e.props)?void 0:n.value)?void 0:o.toString)?void 0:i.call(o))},e.props))):null}))))));var O})),Hi=u.css`
  background: ${({theme:e,appearance:t})=>"secondary"===t||"ghost"===t?e.color.background.tertiary:"white"===t?e.color.basic.tertiary:e.color.background.secondary};
`,Wi=S.default.div`
  border-radius: ${4}px;
  width: ${({dimension:e})=>"s"===e?65:74}px;
  height: ${({dimension:e})=>"s"===e?8:12}px;
  ${Hi}
`;Wi.defaultProps={theme:N};const Ui=S.default.div`
  border-radius: 50%;
  width: ${({dimension:e})=>"s"===e?20:24}px;
  height: ${({dimension:e})=>"s"===e?20:24}px;
  ${Hi}
`;Ui.defaultProps={theme:N};const Yi=(e,t,r,n)=>{if(e&&t.current&&n){const o=e.getBoundingClientRect(),i=t.current.getBoundingClientRect(),l=window.innerHeight;let a=r;r.indexOf("bottom")>-1&&l-o.bottom<0&&(a=a.replace("bottom","top")),r.indexOf("top")>-1&&l-i.bottom-8>=o.height&&(a=a.replace("top","bottom")),r.indexOf("right")>-1&&o.right-o.width<0&&(a=a.replace("right","left")),r.indexOf("left")>-1&&i.right-o.width>=0&&(a=a.replace("left","right")),n(a)}},qi=u.css`
  bottom: 0;
  & ul {
    right: 0;
    top: 8px;
  }
`,Gi=u.css`
  bottom: 0;
  & ul {
    left: 0;
    top: 8px;
  }
`,Ki=u.css`
  top: 0;
  & ul {
    right: 0;
    bottom: 8px;
  }
`,Xi=u.css`
  top: 0;
  & ul {
    left: 0;
    bottom: 8px;
  }
`,Qi=S.default.div`
  position: fixed;
  overflow: visible;
  z-index: ${({theme:e})=>e.zIndex.dropdown};
`,Zi=({targetRef:e,position:t,children:r})=>{const n=d.useRef(null),o="bottom-left"===t||"bottom-right"===t;return d.useEffect((()=>{const t=n.current;if(t&&e.current){const r=an(e.current,(e=>{if(e){const{x:r,y:n,height:i,width:l}=e,{style:a}=t;a.top=o?`${n+i}px`:`${n}px`,a.left=`${r}px`,a.width=`${l}px`}}));return r.observe(),()=>{r.unobserve()}}}),[e.current,n.current,t]),F.default.createPortal(C.default.createElement(Qi,{ref:n,children:r}),document.body)},Ji=S.default.div`
  position: absolute;
  left: 0;
  right: 0;
  height: 0;
  ${({position:e})=>{switch(e){case"bottom-left":return Gi;case"top-right":return Ki;case"top-left":return Xi;case"bottom-right":default:return qi}}}
`,el=S.default(Hn)`
  transition: all 0.3s;
  ${({$menuOpened:e})=>e&&"transform: rotate(180deg);"}
`,tl=S.default(Wi)`
  margin-right: 8px;
`,rl=D.forwardRef(((e,t)=>{var r=e,{children:n,dimension:o="l",appearance:i="primary",disabled:l=!1,loading:u=!1,skeleton:p=!1,onClose:f,onOpen:h,options:m,selected:b,onChange:g}=r,x=c(r,["children","dimension","appearance","disabled","loading","skeleton","onClose","onOpen","options","selected","onChange"]);const[v,y]=d.useState(!1),[$,w]=d.useState("bottom-right"),E=d.useRef(null),[k,C]=d.useState(null),S=d.useCallback((e=>{C(e)}),[v]),F="xl"===o?"l":o,M="s"===o?"s":"m",_=()=>{var e;y(!1),null==f||f(),null==(e=E.current)||e.focus()};d.useLayoutEffect((()=>{v&&Yi(k,E,$,w)}),[v,k,E.current,$,w]),sn(Yi,500,k,E,$,w),Lo([E.current,k],_);return D.createElement(D.Fragment,null,D.createElement(Kt,s(a({},x),{ref:$r(t,E),dimension:o,appearance:i,disabled:!(!u&&!p)||l,onKeyDown:e=>{const t=sr.getCode(e);t!==sr.ArrowDown&&t!==sr.Enter&&t!==sr[" "]||(y(!0),null==h||h(),e.preventDefault())},onClick:()=>{y((e=>(e?null==f||f():null==h||h(),!e)))},"aria-expanded":v}),u?D.createElement(oi,{dimension:M}):p?D.createElement(D.Fragment,null,D.createElement(tl,{dimension:o,appearance:i}),D.createElement(Ui,{dimension:o,appearance:i})):D.createElement(D.Fragment,null,D.Children.toArray(n).map((e=>"string"==typeof e?D.createElement("span",{key:Yt()},e):e)),D.createElement(el,{$menuOpened:v,"aria-hidden":!0}))),v&&!u&&!p&&D.createElement(Zi,{targetRef:E,position:$},D.createElement(Ji,{position:$},D.createElement(Qn,{ref:S,dimension:F,onKeyDown:e=>{sr.getCode(e)===sr.Escape&&_()},role:"listbox"},m.map((e=>{var t=e,{display:r,disabled:n,id:o}=t,i=c(t,["display","disabled","id"]);return D.createElement(ui,s(a({},i),{key:o,id:o,dimension:F,disabled:l||n,selected:b===o,role:"option",onClick:e=>{g(e.currentTarget.id),_()},onKeyDown:e=>{const t=sr.getCode(e);t!==sr.Enter&&t!==sr[" "]||(g(e.currentTarget.id),_(),e.preventDefault())}}),r)}))))))}));var nl,ol,il,ll,al,sl,cl,dl,ul;function pl(){return(pl=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function fl(e){return D.createElement("svg",pl({viewBox:"0 0 24 24",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),nl||(nl=D.createElement("circle",{cx:4.8,cy:12,r:1.8,fill:"#083F91"})),ol||(ol=D.createElement("circle",{cx:12,cy:12,r:1.8,fill:"#083F91"})),il||(il=D.createElement("circle",{cx:19.2,cy:12,r:1.8,fill:"#083F91"})))}function hl(){return(hl=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function ml(e){return D.createElement("svg",hl({viewBox:"0 0 20 20",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),ll||(ll=D.createElement("circle",{cx:4,cy:10,fill:"#083F91",r:1.5})),al||(al=D.createElement("circle",{cx:10,cy:10,fill:"#083F91",r:1.5})),sl||(sl=D.createElement("circle",{cx:16,cy:10,fill:"#083F91",r:1.5})))}function bl(){return(bl=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e}).apply(this,arguments)}function gl(e){return D.createElement("svg",bl({viewBox:"0 0 16 16",fill:"none",xmlns:"http://www.w3.org/2000/svg",focusable:!1},e),cl||(cl=D.createElement("circle",{cx:3.2,cy:8,fill:"#083F91",r:1.2})),dl||(dl=D.createElement("circle",{cx:8,cy:8,fill:"#083F91",r:1.2})),ul||(ul=D.createElement("circle",{cx:12.8,cy:8,fill:"#083F91",r:1.2})))}const xl=u.css`
  width: ${({dimension:e})=>{switch(e){case"s":return"16px";case"m":return"20px";case"l":default:return"24px"}}};
  height: ${({dimension:e})=>{switch(e){case"s":return"16px";case"m":return"20px";case"l":default:return"24px"}}};
`,vl=u.css`
  svg > * {
    fill: ${({theme:e})=>e.color.basic.hover};
  }

  &:before {
    content: '';
    position: absolute;
    top: -4px;
    left: -4px;
    bottom: -4px;
    right: -4px;
    border: 2px solid ${({theme:e})=>e.color.basic.hover};
    border-radius: 4px;
  }
`,yl=S.default.button`
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  position: relative;
  ${xl}
  cursor: pointer;
  border: 0;
  outline: none;
  background: transparent;
  -webkit-tap-highlight-color: transparent;

  & svg {
    ${xl}
    flex-shrink: 0;
  }

  & svg > * {
    fill: ${({theme:e})=>e.color.basic.tertiary};
  }

  &:hover svg > * {
    fill: ${({theme:e})=>e.color.basic.hover};
  }

  ${({menuOpened:e})=>e&&vl}
  &:focus {
    ${vl}
  }

  &:active svg > * {
    fill: ${({theme:e})=>e.color.basic.press};
  }

  &:disabled {
    cursor: default;

    & svg > * {
      fill: ${({theme:e})=>e.color.text.tertiary};
    }
  }
`,$l=D.forwardRef(((e,t)=>{var r=e,{dimension:n="l",menuOpened:o,type:i="button"}=r,l=c(r,["dimension","menuOpened","type"]);return D.createElement(yl,a({ref:t,dimension:n,menuOpened:o,type:i},l),"l"===n?D.createElement(fl,{width:18,"aria-hidden":!0}):"m"===n?D.createElement(ml,{width:14,"aria-hidden":!0}):D.createElement(gl,{width:12,"aria-hidden":!0}))})),wl=(e,t,r,n)=>{if(e&&t.current&&n){const o=e.getBoundingClientRect(),i=t.current.getBoundingClientRect(),l=window.innerHeight;let a=r;r.indexOf("bottom")>-1&&l-o.bottom<0&&(a=a.replace("bottom","top")),r.indexOf("top")>-1&&l-i.bottom-8>=o.height&&(a=a.replace("top","bottom")),r.indexOf("right")>-1&&o.right-o.width<0&&(a=a.replace("right","left")),r.indexOf("left")>-1&&i.right-o.width>=0&&(a=a.replace("left","right")),n(a)}},El=u.css`
  bottom: 0;
  & ul {
    right: -4px;
    top: 8px;
  }
`,kl=u.css`
  bottom: 0;
  & ul {
    left: -4px;
    top: 8px;
  }
`,Cl=u.css`
  top: 0;
  & ul {
    right: -4px;
    bottom: 8px;
  }
`,Dl=u.css`
  top: 0;
  & ul {
    left: -4px;
    bottom: 8px;
  }
`,Sl=S.default.div`
  position: fixed;
  overflow: visible;
  z-index: ${({theme:e})=>e.zIndex.dropdown};
`,Fl=({targetRef:e,position:t,children:r})=>{const n=d.useRef(null),o="bottom-left"===t||"bottom-right"===t;return d.useEffect((()=>{const t=n.current;if(t&&e.current){const r=an(e.current,(e=>{if(e){const{x:r,y:n,height:i,width:l}=e,{style:a}=t;a.top=o?`${n+i}px`:`${n}px`,a.left=`${r}px`,a.width=`${l}px`}}));return r.observe(),()=>{r.unobserve()}}}),[e.current,n.current,t]),F.default.createPortal(C.default.createElement(Sl,{ref:n,children:r}),document.body)},Ml=S.default.div`
  position: absolute;
  left: 0;
  right: 0;
  height: 0;
  ${({position:e})=>{switch(e){case"bottom-left":return kl;case"top-right":return Cl;case"top-left":return Dl;case"bottom-right":default:return El}}}
`,_l=D.forwardRef(((e,t)=>{var r=e,{children:n,dimension:o="l",disabled:i=!1,onClose:l,onOpen:u,selected:p,onChange:f}=r,h=c(r,["children","dimension","disabled","onClose","onOpen","selected","onChange"]);const[m,b]=d.useState(!1),[g,x]=d.useState("bottom-right"),v=d.useRef(null),[y,$]=d.useState(null),w=d.useCallback((e=>{$(e)}),[m]),E="s"===o?"s":"m",k=()=>{var e;b(!1),null==l||l(),null==(e=v.current)||e.focus()};d.useLayoutEffect((()=>{m&&wl(y,v,g,x)}),[m,y,v.current,g,x]),sn(wl,500,y,v,g,x),Lo([v.current,y],k);return D.createElement(D.Fragment,null,D.createElement($l,s(a({},h),{ref:$r(t,v),dimension:o,disabled:i,menuOpened:m,onKeyDown:e=>{const t=sr.getCode(e);t!==sr.ArrowDown&&t!==sr.Enter&&t!==sr[" "]||(b(!0),null==u||u(),e.preventDefault())},onClick:()=>{b((e=>(e?null==l||l():null==u||u(),!e)))},"aria-expanded":m,"aria-haspopup":m})),m&&D.createElement(Fl,{targetRef:v,position:g},D.createElement(Ml,{position:g},D.createElement(Qn,{ref:w,dimension:E,onKeyDown:e=>{const t=sr.getCode(e);t!==sr.Escape&&t!==sr.Tab||k()},role:"listbox","data-dimension":E,"data-selectedid":p,tabIndex:-1},D.Children.map(n,(e=>{if(!D.isValidElement(e))return null;return D.cloneElement(e,a({onClick:t=>{e.props.disabled||(f(t.currentTarget.id),k())},onKeyDown:t=>{const r=sr.getCode(t);r!==sr.Enter&&r!==sr[" "]||e.props.disabled||(f(t.currentTarget.id),k(),t.preventDefault())},"aria-selected":p===e.props.id},e.props))}))))))})),Rl=S.default.div`
  background: ${({theme:e})=>e.color.background.tertiary};
  border-radius: 2px;
  min-width: 140px;
  min-height: 4px;
  display: flex;
`,Al=S.default.div`
  background: ${({theme:e,appearance:t})=>"error"===t?e.color.status.danger:e.color.basic.primary};
  width: ${({percent:e})=>e}%;
  border-radius: 2px;
  transition: all 0.3s linear;
`,Ol=S.default.div`
  display: flex;
  padding-bottom: 8px;
  font-size: 12px;
  display: flex;
  justify-content: space-between;
  width: 100%;
  line-height: 16px;
  font-style: normal;
  font-weight: normal;
  font-family: ${H.fontFamily};
  color: ${({theme:e,appearance:t})=>"error"===t?e.color.status.danger:e.color.text.primary};
`,Bl=S.default.div``,Tl=S.default.div`
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
`,zl=S.default.div`
  background: ${({theme:e})=>e.color.background.tertiary};
  min-width: 140px;
  min-height: 4px;
  display: flex;
`,Il=S.default.div`
  background: ${({theme:e,appearance:t})=>"error"===t?e.color.status.danger:e.color.basic.primary};
  width: ${({percent:e})=>e}%;
  transition: all 0.3s linear;
`,Ll=S.default.div`
  display: flex;
  align-items: center;
  justify-content: center;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  transition: opacity 0.3s ease 0s;
  background-color: ${({theme:e})=>((e,t)=>{const r=e.replace("#",""),n=r.match(new RegExp(`(.{${r.length/3}})`,"g")),o=[];if(!n)throw new Error("Incorrect color srting");for(let i=0;i<n.length;i++)o[i]=parseInt(1==n[i].length?n[i]+n[i]:n[i],16);return void 0!==t&&o.push(t),`rgba(${o.join(",")})`})(e.color.basic.secondary,.6)};
  backdrop-filter: blur(8px);
  z-index: ${({theme:e})=>e.zIndex.modal};
`,Pl=u.css`
  width: ${({dimension:e,mobile:t})=>{if(t)return"calc(100% - 32px)";switch(e){case"s":return"384px";case"m":return"488px";case"xl":return"800px";case"l":default:return"592px"}}};
`,Nl=S.default.h5`
  ${W["Main/S"]}
  margin: 0 32px 16px 0;
`,Vl=S.default.div`
  overflow-x: hidden;
  overflow-y: auto;
  ${({$overflow:e})=>e&&"padding-right: 24px;"}
`,jl=S.default.div`
  display: flex;
  flex-direction: row-reverse;
  margin-top: 24px;
  & > button {
    margin-right: 16px;
  }
  & > button:first-child {
    margin-right: 0;
  }
`,Hl=S.default.div`
  position: absolute;
  box-sizing: border-box;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  display: flex;
  flex-direction: column;
  overflow: hidden;
  padding: ${({mobile:e})=>e?"20px 16px 24px 16px":"20px 24px 24px 24px"};
  ${Pl}
  max-height: ${({mobile:e})=>e?"84vh":"90vh"};
  background-color: ${({theme:e})=>e.color.background.primary};
  ${({theme:e})=>e.shadow.ClickableHover}
  border-radius: 8px;
  ${({mobile:e})=>e?W["Additional/S"]:W["Additional/L"]}
  color: ${({theme:e})=>e.color.text.primary};

  ${({mobile:e})=>e&&`\n    & > ${Nl} {\n      ${W["Main/XS"]}\n      margin: 0 30px 16px 0;\n    }\n    & > ${jl} {\n      flex-direction: column-reverse;\n      & > button {\n        width: 100%;\n        margin-bottom: 16px;\n      }\n      & > button:first-child {\n          margin-bottom: 0;\n      }\n    }\n  `}
`,Wl=S.default.button`
  position: absolute;
  top: 20px;
  right: ${({mobile:e})=>e?16:24}px;
  flex-shrink: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  cursor: pointer;
  outline: none;
  border: 0;
  padding: 0;
  width: 24px;
  height: 24px;
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  ${({theme:e})=>`\n    & svg {\n      fill: ${e.color.basic.tertiary};\n    }\n\n    &:hover,\n    &:focus {\n      & svg {\n        fill: ${e.color.basic.hover};\n      }\n    }\n\n    &:active {\n      & svg {\n        fill: ${e.color.basic.press};\n      }\n    }\n  `}
`,Ul=d.forwardRef(((e,t)=>{var r=e,{dimension:n="l",container:o,mobile:i,onClose:l,closeOnEscapeKeyDown:s,closeOnOutsideClick:u,children:p}=r,f=c(r,["dimension","container","mobile","onClose","closeOnEscapeKeyDown","closeOnOutsideClick","children"]);const h=d.useRef(null),m=d.useRef(null),b=d.useRef(null),[g,x]=d.useState(),[v,y]=d.useState();d.useEffect((()=>(b.current=document.activeElement,()=>{var e;null==(e=b.current)||e.focus()})),[]),d.useLayoutEffect((()=>{if(h.current){const e=Po(h.current);x(e[0]),y(e[e.length-1]),null==g||g.focus()}}),[h.current,p]);return F.default.createPortal(C.default.createElement(Ll,{ref:m,onClick:e=>{u&&e.target===m.current&&(null==l||l())}},C.default.createElement(Hl,a({ref:$r(t,h),role:"dialog","aria-modal":!0,dimension:n,mobile:i,onKeyDown:e=>{e.stopPropagation();const t=sr.getCode(e);t===sr.Escape&&s?(e.preventDefault(),null==l||l()):t===sr.Tab&&(t===sr.Shift?document.activeElement===g&&(v.focus(),e.preventDefault()):document.activeElement===v&&(g.focus(),e.preventDefault()))}},f),p,l&&C.default.createElement(Wl,{"aria-label":"Закрыть модальное окно",mobile:i,onClick:e=>{e.stopPropagation(),null==l||l()}},C.default.createElement(Tr,{width:24,height:24,"aria-hidden":!0})))),o||document.body)})),Yl=S.default.div`
  position: fixed;
  overflow: visible;
  z-index: ${({theme:e})=>e.zIndex.dropdown};
`,ql=({targetRef:e,position:t,container:r,children:n})=>{const o=d.useRef(null),i="bottom-left"===t||"bottom-right"===t;return d.useEffect((()=>{const t=o.current;if(t&&e.current){const r=an(e.current,(e=>{if(e){const{x:r,y:n,height:o,width:l}=e,{style:a}=t;a.top=i?`${n+o}px`:`${n}px`,a.left=`${r}px`,a.width=`${l}px`}}));return r.observe(),()=>{r.unobserve()}}}),[e.current,o.current,t]),F.default.createPortal(C.default.createElement(Yl,{ref:o,children:n,role:"none"}),r||document.body)},Gl=S.default.li`
  position: relative;
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  width: 100%;
  height: ${({dimension:e})=>"m"===e?40:48}px;
  background: ${({theme:e,selected:t})=>t?e.color.background.tertiary:e.color.background.primary};
  color: ${({theme:e})=>e.color.text.primary};
  ${W["Additional/L"]}
  user-select: none;
  cursor: pointer;

  & > div {
    overflow: hidden;
    display: flex;
  }

  & svg {
    display: inline-block;
    flex-shrink: 0;
    fill: ${({theme:e})=>e.color.basic.tertiary};
    width: ${({dimension:e})=>"m"===e?20:24}px;
    height: ${({dimension:e})=>"m"===e?20:24}px;
    margin-right: ${"8px"};
  }
  & [data-badge] {
    margin-left: ${"8px"};
  }

  &:hover,
  &:focus {
    outline: none;
    ${({disabled:e,theme:t})=>!e&&`background: ${t.color.background.secondary};`}
  }

  ${({disabled:e,theme:t})=>e&&`\n  color: ${t.color.text.tertiary};\n  background: ${t.color.background.primary};\n  cursor: default;\n  & svg {\n    fill: ${t.color.text.tertiary};\n  }\n  `}
`,Kl=S.default.div`
  display: flex;
  overflow: hidden;
  text-overflow: ellipsis;
  align-items: center;
  height: ${({dimension:e})=>"m"===e?40:48}px;
  padding: ${({dimension:e})=>"m"===e?"0 12px":"0 16px"};
`,Xl=S.default.div`
  display: inline-block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  & > div {
    display: flex;
    overflow: hidden;
    text-overflow: ellipsis;
    align-items: center;
  }
`,Ql=e=>{const t=d.useContext(u.ThemeContext);return C.default.createElement(vt,s(a({},e),{theme:"dark"==t.name?N:V}))},Zl=e=>{var t=e,{id:r,dimension:n,disabled:o,selected:i,icon:l,badge:u,content:p,onClick:f,onKeyDown:h}=t,m=c(t,["id","dimension","disabled","selected","icon","badge","content","onClick","onKeyDown"]);const b=d.useRef(null),[g,x]=d.useState(!1),v=e=>e.offsetWidth<e.scrollWidth;d.useLayoutEffect((()=>{b.current&&v(b.current)!==g&&x(v(b.current))}),[p,n]);const y=()=>C.default.createElement(Kl,{dimension:n},l&&l,C.default.createElement(Xl,{ref:b},p),void 0!==u&&C.default.createElement(ue,{"data-badge":!0,dimension:"s",appearance:r===i?"info":o?"lightDisable":"light"},u));return C.default.createElement(Gl,s(a({},m),{id:r,dimension:n,disabled:o,selected:i===r,"aria-selected":i===r,role:"tab",tabIndex:i===r?0:-1,onClick:f,onKeyDown:h}),g?C.default.createElement(Ql,{renderContent:()=>p},y()):y())},Jl=(e,t,r,n)=>{var o;if(e&&t.current&&n){const i=e.getBoundingClientRect(),l=null==(o=t.current)?void 0:o.getBoundingClientRect(),a=window.innerHeight;let s=r;r.indexOf("bottom")>-1&&a-i.bottom<0&&(s=s.replace("bottom","top")),r.indexOf("top")>-1&&a-l.bottom-8>=i.height&&(s=s.replace("top","bottom")),r.indexOf("right")>-1&&i.right-i.width<0&&(s=s.replace("right","left")),r.indexOf("left")>-1&&l.right-i.width>=0&&(s=s.replace("left","right")),n(s)}},ea=u.css`
  bottom: 0;
  & ul {
    right: 0px;
    top: 8px;
  }
`,ta=u.css`
  bottom: 0;
  & ul {
    left: 0px;
    top: 8px;
  }
`,ra=u.css`
  top: 0;
  & ul {
    right: 0px;
    bottom: 8px;
  }
`,na=u.css`
  top: 0;
  & ul {
    left: 0px;
    bottom: 8px;
  }
`,oa=(e,t,r,n,o,i)=>{let l=!1,a=r(e,t,n,o,i);for(;a;){if(a===(null==e?void 0:e.firstChild)){if(l)return;l=!0}if(a.hasAttribute("tabindex")&&!a.hasAttribute("disabled"))return void a.focus();a=r(e,a,n,o,i)}},ia=(e,t,r,n,o)=>(null==t?void 0:t.nextElementSibling)?t.nextElementSibling:(null==o||o(),r?null:null==e?void 0:e.firstChild),la=(e,t,r,n,o)=>(null==t?void 0:t.previousElementSibling)?t.previousElementSibling:(null==n||n(),r?null:null==e?void 0:e.lastChild),aa=u.css`
  max-height: ${({dimension:e})=>{switch(e){case"l":return"288px";case"m":return"240px";default:return"288px"}}};
`,sa=S.default.ul`
  border-radius: 4px;
  ${({theme:e})=>e.shadow.NonClickable};
  background: ${({theme:e})=>e.color.background.primary};
  ${aa}
  overflow-x: hidden;
  overflow-y: auto;
  cursor: pointer;
  position: absolute;
  padding: 8px 0;
  width: 300px;
  outline: none;
  margin: 0;
`,ca=d.forwardRef(((e,t)=>{var r=e,{children:n,dimension:o="l",onMenuReachBottom:i,onMenuReachTop:l,menuFocus:u,setMenuFocus:p}=r,f=c(r,["children","dimension","onMenuReachBottom","onMenuReachTop","menuFocus","setMenuFocus"]);const h=d.useRef(),m=()=>{var e;let t=null==(e=h.current)?void 0:e.firstElementChild;for(;t.hasAttribute("disabled");)t=t.nextElementSibling;return t};return d.useLayoutEffect((()=>{var e,t;h.current!==document.activeElement&&(null==(e=h.current)||e.focus());const r=null==(t=h.current)?void 0:t.querySelector('[aria-selected="true"]');let n;"firstOption"===u&&(n=m()),"lastOption"===u&&(n=(()=>{var e;let t=null==(e=h.current)?void 0:e.lastElementChild;for(;t.hasAttribute("disabled");)t=t.previousElementSibling;return t})()),"activeOption"===u&&(n=r||m()),null==n||n.focus()}),[h,ia,oa]),C.default.createElement(sa,s(a({ref:$r(t,h),dimension:o},f),{onKeyDown:e=>{var t,r;const n=((null==(t=h.current)?void 0:t.ownerDocument)||document).activeElement,o=sr.getCode(e);o===sr.ArrowRight?(oa(h.current,n,ia,!0,l,i),e.preventDefault()):o===sr.ArrowLeft?(oa(h.current,n,la,!0,l,i),e.preventDefault()):32===e.keyCode?e.preventDefault():o===sr.Home?(e.preventDefault(),oa(h.current,null,ia)):o===sr.End&&(e.preventDefault(),oa(h.current,null,la)),null==(r=null==f?void 0:f.onKeyDown)||r.call(f,e)}}),n)})),da=u.css`
  width: ${({dimension:e})=>{switch(e){case"m":return"20px";case"l":default:return"24px"}}};
  height: ${({dimension:e})=>{switch(e){case"m":return"20px";case"l":default:return"24px"}}};
`,ua=u.css`
  svg > * {
    fill: ${({theme:e})=>e.color.basic.hover};
  }
  border: 2px solid ${({theme:e})=>e.color.basic.hover};
`,pa=S.default.button`
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
  border: 0;
  outline: none;
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  margin-left: 4px;
  padding: ${({dimension:e})=>"l"===e?"10px":"8px"};
  border: 2px solid transparent;

  & svg {
    ${da}
    flex-shrink: 0;
  }

  & svg > * {
    fill: ${({theme:e})=>e.color.basic.tertiary};
  }

  &:hover svg > * {
    fill: ${({theme:e})=>e.color.basic.hover};
  }

  ${({menuOpened:e})=>e&&ua}
  &:focus {
    ${ua}
  }

  &:active svg > * {
    fill: ${({theme:e})=>e.color.basic.press};
  }

  &:disabled {
    cursor: default;
    & svg > * {
      fill: ${({theme:e})=>e.color.text.tertiary};
    }
  }
`,fa=d.forwardRef(((e,t)=>{var r=e,{dimension:n="l",menuOpened:o,type:i="button"}=r,l=c(r,["dimension","menuOpened","type"]);return C.default.createElement(pa,a({ref:t,dimension:n,menuOpened:o,type:i},l),"l"===n?C.default.createElement(fl,{width:18}):C.default.createElement(ml,{width:14}))})),ha=S.default.div`
  position: absolute;
  left: 0;
  right: 0;
  height: 0;
  ${({position:e})=>{switch(e){case"bottom-left":return ta;case"top-right":return ra;case"top-left":return na;case"bottom-right":default:return ea}}}
`,ma=S.default.div`
  height: 0;
  width: 0;
  opacity: 0;
`,ba=d.forwardRef(((e,t)=>{var r,n=e,{dimension:o="l",disabled:i=!1,options:l,selected:u,onMenuReachTop:p,menuFocus:f,onChange:h,onMenuReachBottom:m,setMenuFocus:b}=n,g=c(n,["dimension","disabled","options","selected","onMenuReachTop","menuFocus","onChange","onMenuReachBottom","setMenuFocus"]);const[x,v]=d.useState(!1),[y,$]=d.useState("bottom-right"),w=d.useRef(null),[E,k]=d.useState(null),D=d.useCallback((e=>{k(e)}),[x]),S=()=>{var e;null==(e=w.current)||e.focus(),v(!1),b("activeOption")};d.useLayoutEffect((()=>{x&&Jl(E,w,y,$)}),[x]),sn(Jl,500,E,w,y,$),Lo([w.current,E],S);return C.default.createElement(C.default.Fragment,null,!x&&l.map((({id:e,content:t})=>C.default.createElement(ma,{role:"tab",tabIndex:-1,key:e,id:e},t))),C.default.createElement(fa,s(a({},g),{ref:$r(t,w),dimension:o,disabled:i,menuOpened:x,onMouseDown:e=>{e.preventDefault(),v((e=>!e))},onFocus:e=>{v(!0)},role:"none","aria-hidden":!0})),x&&C.default.createElement(ql,{targetRef:w,position:y,container:null==(r=w.current)?void 0:r.parentElement},C.default.createElement(ha,{position:y,role:"none"},C.default.createElement(ca,{ref:D,dimension:o,onKeyDown:e=>{sr.getCode(e)===sr.Tab&&S()},onMenuReachTop:()=>{S(),null==p||p()},onMenuReachBottom:()=>{S(),null==m||m()},menuFocus:f,setMenuFocus:b,role:"none"},l.map((e=>{var t=e,{disabled:r,id:n}=t,l=c(t,["disabled","id"]);const s=i||r;return C.default.createElement(Zl,a({key:n,id:n,disabled:s,dimension:o,selected:u,onClick:e=>{s||(h(e.currentTarget.id),S())},onKeyDown:e=>{const t=sr.getCode(e);t!==sr.Enter&&t!==sr[" "]||s||(h(e.currentTarget.id),e.preventDefault())}},l))}))))))})),ga=S.default.div`
  position: relative;
  display: flex;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  box-shadow: inset 0 -${"2px"} 0 0 ${({theme:e,underline:t})=>t?e.color.background.tertiary:"transparent"};
  overflow-x: ${({mobile:e})=>e?"scroll":"visible"};

  &::-webkit-scrollbar {
    width: 0 !important;
    height: 0 !important;
  }

  overflow: -moz-scrollbars-none;
  -ms-overflow-style: none;
  scrollbar-width: none;
  scrollbar-height: none;
`,xa=S.default.div`
  display: flex;
`,va=S.default.div`
  position: absolute;
  bottom: 0;
  ${({theme:e,left:t,width:r})=>`\n    display: ${r?"block":"none"};\n    left: ${t}px;\n    width: ${r}px;\n    background-color: ${e.color.basic.primary};\n  `}
  height: ${"2px"};
  transition: all 0.2s ease-out;
`,ya=S.default.div`
  height: 100%;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
`,$a=S.default.div`
  display: inline-flex;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: ${({dimension:e})=>"l"===e?"0 16px":"0 12px"};
`,wa=S.default.button`
  position: relative;
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  height: ${({dimension:e})=>"m"===e?40:48}px;
  padding: 0;
  background: transparent;
  appearance: none;
  border: none;
  color: ${({theme:e})=>e.color.text.primary};
  ${({dimension:e})=>"m"===e?W["Additional/S"]:W["Additional/L"]}
  user-select: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;

  &:focus {
    &:before {
      position: absolute;
      content: '';
      border: ${"2px"} solid ${({theme:e})=>e.color.basic.hover};
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
    }
  }
  &:focus,
  & > ${$a} {
    outline: none;
  }

  & svg {
    display: inline-block;
    flex: 1 0 auto;
    fill: ${({theme:e})=>e.color.basic.tertiary};
    width: ${({dimension:e})=>"m"===e?20:24}px;
    height: ${({dimension:e})=>"m"===e?20:24}px;
    margin-right: ${"8px"};
  }
  & [data-badge] {
    margin-left: ${"8px"};
  }

  &:hover:not(:disabled) {
    color: ${({theme:e})=>e.color.basic.hover};
    & svg {
      fill: ${({theme:e})=>e.color.basic.hover};
    }
    [data-badge] {
      background: ${({theme:e})=>e.color.basic.hover};
      color: ${({theme:e})=>e.color.text.staticWhite};
    }
  }
  &:active:not(:disabled) {
    color: ${({theme:e})=>e.color.basic.press};
    & svg {
      fill: ${({theme:e})=>e.color.basic.press};
    }
    [data-badge] {
      background: ${({theme:e})=>e.color.basic.press};
      color: ${({theme:e})=>e.color.text.staticWhite};
    }
  }
  &:disabled {
    color: ${({theme:e})=>e.color.text.tertiary};
    cursor: default;
    & svg {
      fill: ${({theme:e})=>e.color.text.tertiary};
    }
  }
`,Ea=`\n    @font-face {\n        font-family: 'VTB Group UI';\n        src: url('${M.default}');\n        src: url('${M.default}?#iefix') format('embedded-opentype'),\n          url('${_.default}') format('woff2'),\n          url('${R.default}') format('woff'),\n          url('${A.default}') format('truetype'),\n          url('${O.default}#VTBGroupUIWeb-Book') format('svg');\n        font-display: swap;\n        font-weight: 500;\n        font-style: normal;\n        font-feature-settings: 'tnum' on, 'lnum' on, 'cv03' on, 'cv04' on;\n    }\n\n    @font-face {\n        font-family: 'VTB Group UI';\n        src: url('${B.default}');\n        src: url('${B.default}?#iefix') format('embedded-opentype'),\n          url('${T.default}') format('woff2'),\n          url('${z.default}') format('woff'),\n          url('${I.default}') format('truetype'),\n          url('${L.default}#VTBGroupUIWeb-Book') format('svg');\n        font-display: swap;\n        font-weight: normal;\n        font-style: normal;\n        font-feature-settings: 'tnum' on, 'lnum' on, 'cv03' on, 'cv04' on;\n    }\n`;exports.ALL_FIELDSET_DIMENSIONS_VALUES=["m","s"],exports.Additional_L_bold=J,exports.Additional_L_normal=ee,exports.Additional_M_normal=te,exports.Additional_S_bold=re,exports.Additional_S_normal=ne,exports.Additional_XS_normal=oe,exports.Background=xr,exports.Badge=ue,exports.Button=Kt,exports.ButtonGroup=e=>{var t=e,{dimension:r="l",tabIndex:n=-1}=t,o=c(t,["dimension","tabIndex"]);const i=d.useRef(null);return C.default.createElement(rr,a({ref:i,onKeyDown:e=>{var t;const r=(i.current&&i.current.ownerDocument||document).activeElement,n=sr.getCode(e);n===sr.ArrowRight?nr(i.current,r,or):n===sr.ArrowLeft&&nr(i.current,r,ir),null==(t=null==o?void 0:o.onKeyDown)||t.call(o,e)},dimension:r,tabIndex:n},o))},exports.ButtonGroupItem=Zt,exports.Button_M=ie,exports.Button_S=le,exports.Calendar=wt,exports.Caption_XS=ae,exports.Checkbox=yr,exports.CheckboxCompositeGroup=e=>{var t=e,{children:r,disabled:n=!1,dimension:o="m"}=t,i=c(t,["children","disabled","dimension"]);return C.default.createElement($o,a({dimension:o,disabled:n},i),d.Children.map(r,(e=>d.isValidElement(e)?d.cloneElement(e,a({disabled:n||e.props.disabled,dimension:o},e.props)):null)))},exports.CheckboxField=Ln,exports.CheckboxGroup=e=>{var t=e,{children:r,label:n,disabled:o=!1,dimension:i="m"}=t,l=c(t,["children","label","disabled","dimension"]);return C.default.createElement(yo,a({disabled:o,dimension:i},l),n&&n,d.Children.map(r,(e=>d.isValidElement(e)?d.cloneElement(e,a({disabled:o||e.props.disabled,dimension:i},e.props)):null)))},exports.ChipComponent=Fo,exports.Chips=Ro,exports.CloseIcon=Mo,exports.Container=Mn,exports.DARK_THEME=V,exports.DateField=Pn,exports.DateInput=Dn,exports.DropDownItem=ui,exports.DropDownMenu=Qn,exports.Field=Rn,exports.FieldSet=e=>{const t=e,{legend:r,children:n}=t,o=c(t,["legend","children"]);return D.createElement(xo,s(a({},o),{"data-dimension":o.dimension}),r&&D.createElement(go,{dimension:e.dimension,children:r}),n)},exports.FontsVTBGroup=function(e){return C.default.createElement("style",a({type:"text/css",children:Ea},e))},exports.Hint=e=>{var t=e,{renderContent:r,container:n,target:o,visibilityTrigger:i="hover",isMobile:l=!1,dimension:s="l",hintRef:u,children:p,className:f,id:h}=t,m=c(t,["renderContent","container","target","visibilityTrigger","isMobile","dimension","hintRef","children","className","id"]);const b=o,g=d.useRef(null),x=d.useRef(null),v=n||document.body,y=r(),$=h||Yt(),[w,E]=d.useReducer(jo,Vo),k=()=>E({type:"setInvisible"});d.useLayoutEffect((()=>{const e=(null==b?void 0:b.current)||g.current;if(w.visible&&e&&x.current){const t=e.getBoundingClientRect(),r=x.current.getBoundingClientRect();if(l){const e=8,r=t.right>window.innerWidth-t.left?t.right:window.innerWidth-t.left;x.current.style.maxWidth=r-e+"px",E({type:"setHintWidth",payload:r})}const n=function(e,t){const r=e.getBoundingClientRect(),n=t.getBoundingClientRect(),o=Object.entries(Io).filter((e=>e[1].check(r,n)));return o.length?o[0][1]:Io["bottom-left"]}(e,x.current).getStyles(t,r);x.current.style.transform=n}}),[null==b?void 0:b.current,g.current,x.current,w.visible,w.recalculation,w.hintWidth,s,y,l]);const D=d.useMemo((()=>{var e;return null!=(e=mt(g.current))?e:[]}),[g.current]),S=()=>{E({type:"setVisible"})};return C.default.createElement(Ao,{onMouseEnter:"click"===i?void 0:S,onMouseLeave:"click"===i?void 0:k,onFocus:"click"===i?void 0:S,onBlur:"click"===i?void 0:k,onClick:"click"===i?S:void 0,onKeyDown:"click"===i?e=>{const t=sr.getCode(e);t!==sr.Enter&&t!==sr[" "]||(e.preventDefault(),E({type:"setVisible"}))}:void 0,ref:g,className:f,id:$},p,w.visible&&C.default.createElement(No,a({ref:e=>ht(e,u,x),dimension:s,isMobile:l,content:y,visibilityTrigger:i,portal:v,scrollableParents:D,dispatch:E,anchorElementRef:g,anchorId:$},m)))},exports.INPUT_DIMENSIONS_VALUES=["xl","m","s"],exports.INPUT_STATUS_VALUES=["error","success"],exports.InputField=An,exports.LIGHT_THEME=N,exports.Label=Sn,exports.Link=zt,exports.Main_L=G,exports.Main_M=K,exports.Main_S=X,exports.Main_XL=q,exports.Main_XS_bold=Q,exports.Main_XS_normal=Z,exports.Main_XXL=Y,exports.MenuButton=rl,exports.Modal=Ul,exports.ModalButtonPanel=e=>{var t=e,{children:r}=t,n=c(t,["children"]);return C.default.createElement(jl,a({},n),r)},exports.ModalContent=e=>{var t=e,{children:r}=t,n=c(t,["children"]);const o=d.useRef(null),[i,l]=d.useState(!1),s=e=>e.clientHeight<e.scrollHeight;return d.useLayoutEffect((()=>{o.current&&s(o.current)!==i&&l(s(o.current))}),[r]),C.default.createElement(Vl,a({ref:o,$overflow:i},n),r)},exports.ModalTitle=e=>{var t=e,{children:r}=t,n=c(t,["children"]);return C.default.createElement(Nl,a({},n),r)},exports.MultiSelect=ji,exports.OverflowMenu=_l,exports.ProgressHeader=e=>{var t=e,{percent:r=0,appearance:n="primary"}=t,o=c(t,["percent","appearance"]);return C.default.createElement(Tl,a({},o),C.default.createElement(zl,{appearance:n},C.default.createElement(Il,{appearance:n,percent:r})))},exports.ProgressPage=e=>{var t=e,{percent:r=0,children:n,label:o,appearance:i="primary"}=t,l=c(t,["percent","children","label","appearance"]);return C.default.createElement(Bl,a({},l),C.default.createElement(Ol,{appearance:i},o),C.default.createElement(Rl,{appearance:i},C.default.createElement(Al,{appearance:i,percent:r})))},exports.RadioButton=_t,exports.Select=ho,exports.SelectField=bo,exports.Spinner=oi,exports.TYPOGRAPHY=H,exports.TabMenu=e=>{var t=e,{tabs:r,dimension:n="l",underline:o=!1,mobile:i=!1,activeTab:l,onChange:p}=t,f=c(t,["tabs","dimension","underline","mobile","activeTab","onChange"]);const h=C.default.useMemo((()=>r.map((e=>s(a({},e),{ref:d.createRef()})))),[r]),m=d.useRef(null),b=d.useRef(null),g=d.useRef(null),x=d.useRef(null),[v,y]=d.useState(h.length),[$,w]=d.useState({width:0,left:0}),[E,k]=d.useState("activeOption"),{left:D,width:S}=$,M=i?h:h.slice(0,v),_=i?[]:h.slice(v),R=()=>{var e,t;const r=null==(t=null==(e=h.filter((e=>e.id===l)))?void 0:e[0])?void 0:t.ref.current;!r||r.offsetLeft===D&&r.clientWidth===S||w({left:r.offsetLeft,width:r.clientWidth}),r&&!_.filter((e=>e.id===l)).length||w({left:0,width:0})};d.useLayoutEffect((()=>{h.forEach((e=>{(({icon:e,badge:t,content:r},n,o)=>{const i=document.createElement("div");i.style.position="absolute",i.style.visibility="hidden",i.style.left="-10000",i.style.top="-10000",document.body.appendChild(i);const l=C.default.createElement(u.ThemeProvider,{theme:N},C.default.createElement(wa,{dimension:n},C.default.createElement($a,{dimension:n,tabIndex:-1},e&&e,C.default.createElement(ya,null,r),void 0!==t&&C.default.createElement(ue,{"data-badge":!0,dimension:"s",appearance:"lightDisable"},t))));F.default.render(l,i,(()=>{var e;const t=i.clientWidth;o(t),F.default.unmountComponentAtNode(i),null==(e=i.parentNode)||e.removeChild(i)}))})(e,n,(t=>{e.width=t}))}))}),[h,n]),d.useLayoutEffect((()=>R()),[r,D,S,h,l,v]),d.useLayoutEffect((()=>{if(b.current){const e=an(b.current,(e=>{const t=(null==e?void 0:e.width)||0;let r="l"===n?52:44,o=0;h.forEach((e=>{r+=e.width||0,r<=t&&o++})),y(o)}));return e.observe(),()=>{e.unobserve()}}}),[b.current]),d.useLayoutEffect((()=>{if(m.current){const e=an(m.current,(()=>R()));return e.observe(),()=>{e.unobserve()}}}),[m.current]);const A=e=>{i&&e.currentTarget.scrollIntoView({behavior:"smooth",inline:"center",block:"nearest"}),p(e.currentTarget.id)},O=e=>{const t=sr.getCode(e);t!==sr.Enter&&t!==sr[" "]||p(e.currentTarget.id)};return C.default.createElement(ga,a({role:"tablist",ref:b,underline:o,mobile:i},f),C.default.createElement(va,{ref:x,left:D,width:S,"aria-hidden":!0}),C.default.createElement(xa,{ref:m,onKeyDown:e=>{const{target:t}=e;let r;const n=sr.getCode(e);switch(n){case sr.ArrowLeft:r=(e=>{var t,r;let n=e.previousElementSibling||g.current||(null==(t=m.current)?void 0:t.lastElementChild);for(;null==n?void 0:n.hasAttribute("disabled");)n=n.previousElementSibling||g.current||(null==(r=m.current)?void 0:r.lastElementChild);return n})(t),e.preventDefault();break;case sr.ArrowRight:r=(e=>{var t,r;let n=e.nextElementSibling||g.current||(null==(t=m.current)?void 0:t.firstElementChild);for(;null==n?void 0:n.hasAttribute("disabled");)n=n.nextElementSibling||g.current||(null==(r=m.current)?void 0:r.firstElementChild);return n})(t),e.preventDefault();break;case sr.Tab:k("activeOption")}r&&(r.focus(),n===sr.ArrowLeft&&k("lastOption"),n===sr.ArrowRight&&k("firstOption"))}},M.map((e=>{const t=e,{disabled:r,content:o,id:i,icon:s,badge:d,ref:u,width:p}=t,f=c(t,["disabled","content","id","icon","badge","ref","width"]);return C.default.createElement(wa,a({ref:u,key:i,id:i,role:"tab","aria-selected":i===l,tabIndex:i===l?0:-1,dimension:n,disabled:r,onClick:A,onKeyUp:O},f),C.default.createElement($a,{dimension:n,tabIndex:-1},s&&s,C.default.createElement(ya,null,o),void 0!==d&&C.default.createElement(ue,{"data-badge":!0,dimension:"s",appearance:i===l?"info":r?"lightDisable":"light"},d)))}))),_.length&&!i?C.default.createElement(ba,{ref:g,options:_.map((e=>{var t=e,{ref:r}=t;return c(t,["ref"])})),selected:l,dimension:n,disabled:_.length===_.filter((e=>e.disabled)).length,onChange:e=>{p(e),w({left:0,width:0})},tabIndex:(null==_?void 0:_.filter((e=>e.id===l)).length)?0:-1,onMenuReachTop:()=>{var e;let t=null==(e=m.current)?void 0:e.lastElementChild;for(;null==t?void 0:t.hasAttribute("disabled");)t=t.previousElementSibling;t?t.focus():g.current.focus()},onMenuReachBottom:()=>{var e;let t=null==(e=m.current)?void 0:e.firstElementChild;for(;null==t?void 0:t.hasAttribute("disabled");)t=t.nextElementSibling;t?t.focus():g.current.focus()},menuFocus:E,setMenuFocus:k}):null)},exports.Tag=e=>{var t=e,{children:r,kind:n="neutral",width:o,onClick:i}=t,l=c(t,["children","kind","width","onClick"]);const s=d.useRef(null),[u,p]=d.useState(!1),f=e=>e.offsetWidth<e.scrollWidth;d.useLayoutEffect((()=>{s.current&&f(s.current)!==u&&p(f(s.current))}),[r,o]);const h=()=>C.default.createElement(fi,a({width:o,onClick:i,clickable:!!i},l),"neutral"!==n&&C.default.createElement(mi,{kind:n}),r&&C.default.createElement(hi,{ref:s},r));return u?C.default.createElement(vt,{renderContent:()=>r},h()):h()},exports.Tags=({children:e,width:t,kind:r,onClick:n})=>C.default.createElement(bi,null,d.Children.map(e,(e=>d.isValidElement(e)?d.cloneElement(e,a({onClick:e.props.onClick||n,width:e.props.width||t,kind:e.props.kind||r},e.props)):null))),exports.TextInput=Jr,exports.Toggle=ki,exports.Tooltip=vt,exports.getTextStyle=U,exports.height=Bn,exports.heights=Kn,exports.typography=W,exports.width=On;
