// @vtb/ui-kit3/src/components/form/DateField/DateField.stories.tsx
import type { DateFieldProps } from '~/components/form/DateField';
import { DateField } from '~/components/form/DateField';
import * as React from 'react';
import { Meta, Story } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import styled from 'styled-components';
import { INPUT_DIMENSIONS_VALUES } from '~/components';

export default {
  title: 'Form Field Components/DateField',
  component: DateField,
  decorators: [withDesign],
  parameters: {
    design: {
      type: 'figma',
      url:
        'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=65740%3A76866',
    },
  },
  argTypes: {
    dimension: {
      options: INPUT_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },
  },
} as Meta;

const DisplayContainer = styled.div`
  > * {
    margin-bottom: 24px;
  }
`;
const Template: Story<DateFieldProps> = (props) => {
  const [localValue, setValue] = React.useState<string>(String(props.value) ?? '');

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.currentTarget.value;
    setValue(inputValue);
    props.onChange?.(e);
  };

  return (
    <DisplayContainer>
      <DateField {...props} value={localValue} onChange={handleChange} />
      <DateField required label="uncontrolled input" />
      <DateField type="date-range" label="uncontrolled date range" />
    </DisplayContainer>
  );
};

export const InputFieldInput = Template.bind({});

InputFieldInput.args = {
  value: 'Привет',
  label: 'Label',
};

InputFieldInput.storyName = 'Input field example';
