// @vtb/ui-kit3/src/components/Select/Select.stories.tsx
import React, { Fragment, useState } from 'react';
import styled from 'styled-components';
import { Meta, Story } from '@storybook/react';
import { Select, SelectProps } from './index';
import { withDesign } from 'storybook-addon-designs';
import { ReactComponent as CardSolid } from '@openvtb/admiral-icons/build/finance/CardSolid.svg';
import { ReactComponent as StarSolid } from '@openvtb/admiral-icons/build/system/StarSolid.svg';
import { DropDownItem } from '~/components/DropDownItem';
import { Additional_S_normal } from '../Typography';
import { Tooltip } from '~/components/Tooltip';

const StyledText = styled(Additional_S_normal)`
  color: ${({ theme }) => theme.color.text.primary};
  pointer-events: none;
`;

const StyledDropDownItem = styled(DropDownItem)`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const StyledSelect = styled(Select)`
  width: 190px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const WrapperInputValue = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

const WrapperContent = styled.div`
  display: flex;
  overflow: hidden;
  align-items: center;
  & svg {
    margin-right: 10px;
  }
`;

const StyledLabel = styled.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export default {
  title: 'Example/Select',
  component: Select,
  decorators: [withDesign],
  parameters: {
    docs: {
      source: {
        type: 'code',
      },
    },
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=65740%3A76866',
    },
  },
  argTypes: {
    dimension: {
      options: ['xl', 'm', 's'],
      control: { type: 'radio' },
      defaultValue: 'xl',
    },
    disabled: {
      options: [undefined, true],
      control: { type: 'radio' },
    },
  },
} as Meta;

const Template: Story<SelectProps> = (props) => {
  const [selectValue, setValue] = useState('');
  const options = [
    { id: 1, label: 'Option one', value: '1' },
    { id: 2, label: 'Option two', value: '2' },
    { id: 3, label: 'Option three', value: '3' },
    { id: 4, label: 'Option four', value: '4' },
    { id: 5, label: 'Option five', value: '5' },
    { id: 6, label: 'Option six', value: '6' },
    { id: 7, label: 'Option seven', value: '7' },
  ];
  return (
    <Select {...props} value={selectValue} placeholder={'Placeholder'} onChange={(value) => setValue(value)}>
      {options.map((item) => {
        return (
          <DropDownItem role="option" value={item.label} key={item.id}>
            {item.label}
          </DropDownItem>
        );
      })}
    </Select>
  );
};

export const SelectDefault = Template.bind({});
SelectDefault.storyName = 'Стандартный селект';

const Template1: Story<SelectProps> = (props) => {
  const options = [
    { id: 1, label: 'Option one', value: '1', content: <StarSolid width={24} /> },
    { id: 2, label: 'Option two', value: '2', content: <CardSolid width={24} /> },
    { id: 3, label: 'Option three', value: '3', content: <CardSolid width={24} /> },
    { id: 4, label: 'Option four', value: '4', content: <CardSolid width={24} /> },
    { id: 5, label: 'Option five', value: '5', content: <CardSolid width={24} /> },
    { id: 6, label: 'Option six', value: '6', content: <StarSolid width={24} /> },
    { id: 7, label: 'Option seven', value: '7', content: <StarSolid width={24} /> },
  ];
  const [selectValue, setValue] = useState('1');
  return (
    <Select
      {...props}
      value={selectValue}
      onChange={(value) => setValue(value)}
      renderInputValue={() => (
        <WrapperContent>
          {options.filter((d) => d.value === selectValue)[0]?.content}
          <StyledLabel>{options.filter((d) => d.value === selectValue)[0]?.label}</StyledLabel>
        </WrapperContent>
      )}
    >
      {options.map((item) => {
        return (
          <DropDownItem aria-selected={selectValue === item.value} key={item.id} role="option" value={item.value}>
            <WrapperContent>
              {item.content} {item.label}
            </WrapperContent>
          </DropDownItem>
        );
      })}
    </Select>
  );
};

export const SelectCustomRender = Template1.bind({});
SelectCustomRender.storyName = 'Селект с иконкой';

const Template2: Story<SelectProps> = (props) => {
  const iconWidth = props.dimension === 's' ? 24 : 30;
  const category = [
    {
      name: 'Категория 1',
      id: '1',
      content: [
        {
          id: '2',
          label: 'Номер Карты /****45',
          value: '1',
          icon: <CardSolid width={iconWidth} height={iconWidth} />,
          subText: 'Дополнительный текст',
        },
        {
          id: '3',
          label: 'Номер Карты /****75',
          value: '2',
          icon: <CardSolid width={iconWidth} height={iconWidth} />,
          subText: 'Дополнительный текст',
        },
        {
          id: '4',
          label: 'Номер Карты /****22',
          value: '3',
          icon: <CardSolid width={iconWidth} height={iconWidth} />,
          subText: 'Дополнительный текст',
        },
        {
          id: '5',
          label: 'Номер Карты /****33',
          value: '4',
          icon: <CardSolid width={iconWidth} height={iconWidth} />,
          subText: 'Дополнительный текст',
        },
      ],
    },
    {
      name: 'Категория 2',
      id: '9',
      content: [
        {
          id: '10',
          label: 'Номер Карты /****21',
          value: '5',
          icon: <StarSolid width={iconWidth} height={iconWidth} />,
          subText: 'Дополнительный текст',
        },
        {
          id: '11',
          label: 'Номер Карты /****35',
          value: '6',
          icon: <StarSolid width={iconWidth} height={iconWidth} />,
          subText: 'Дополнительный текст',
        },
        {
          id: '12',
          label: 'Номер Карты /****39',
          value: '7',
          subText: 'Дополнительный текст',
          icon: <StarSolid width={iconWidth} height={iconWidth} />,
        },
        {
          id: '13',
          label: 'Номер Карты /****41',
          value: '8',
          icon: <StarSolid width={iconWidth} height={iconWidth} />,
          subText: 'Дополнительный текст',
        },
      ],
    },
  ];

  const [selectValue, setValue] = useState('1');

  const renderValue = () => {
    const content = [...category[0].content, ...category[1].content].filter(
      (d) => d.value?.toString() === selectValue,
    )[0];
    return (
      <WrapperInputValue>
        <StyledLabel>{content.label}</StyledLabel>
        {content.icon}
      </WrapperInputValue>
    );
  };

  return (
    <Select {...props} value={selectValue} renderInputValue={() => renderValue()} onChange={(value) => setValue(value)}>
      {category.map((item, index) => {
        return (
          <Fragment key={index}>
            <DropDownItem disabled key={item.id}>
              <StyledText> {item.name}</StyledText>
            </DropDownItem>
            {item.content.map((subCategory) => {
              return (
                <DropDownItem
                  tabIndex={0}
                  role="option"
                  key={subCategory.id}
                  value={subCategory.value}
                  id={subCategory.id}
                >
                  <div style={{ width: '100%' }}>
                    <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between' }}>
                      {subCategory.label} {subCategory.icon}
                    </div>
                    <StyledText>{subCategory.subText}</StyledText>
                  </div>
                </DropDownItem>
              );
            })}
          </Fragment>
        );
      })}
    </Select>
  );
};

export const SelectGroup = Template2.bind({});
SelectGroup.storyName = 'Продуктовый селект';

const Template3: Story<SelectProps> = (props) => {
  const [selectValue, setValue] = useState('');
  const options = [
    { id: 1, label: 'Option one long text long text', value: '1' },
    { id: 2, label: 'Option two long text long text ', value: '2' },
    { id: 3, label: 'Option three long text long text', value: '3' },
    { id: 4, label: 'Option four long text long text', value: '4' },
    { id: 5, label: 'Option five long text long text', value: '5' },
    { id: 6, label: 'Option six long text long text', value: '6' },
    { id: 7, label: 'Option seven long text long text', value: '7' },
  ];
  return (
    <StyledSelect
      {...props}
      renderInputValue={() => {
        return (
          <Tooltip renderContent={() => selectValue}>
            <div style={{ textOverflow: 'ellipsis', overflow: 'hidden', width: '140px' }}>{selectValue}</div>
          </Tooltip>
        );
      }}
      value={selectValue}
      placeholder={'Placeholder'}
      onChange={(value) => setValue(value)}
    >
      {options.map((item) => {
        return (
          <StyledDropDownItem role="option" value={item.label} key={item.id}>
            <Tooltip renderContent={() => item.label}>
              <div style={{ textOverflow: 'ellipsis', overflow: 'hidden', width: '150px' }}>{item.label}</div>
            </Tooltip>
          </StyledDropDownItem>
        );
      })}
    </StyledSelect>
  );
};

export const SelectWithTooltip = Template3.bind({});
SelectWithTooltip.storyName = 'Селект с тултипом';
