// @vtb/ui-kit3/src/components/OverflowMenu/Buton.tsx
import type { ButtonHTMLAttributes, ReactNode } from 'react';
import * as React from 'react';
import styled, { css } from 'styled-components';

import { ReactComponent as ButtonL } from './svgs/ButtonL.svg';
import { ReactComponent as ButtonM } from './svgs/ButtonM.svg';
import { ReactComponent as ButtonS } from './svgs/ButtonS.svg';

export type Dimension = 'l' | 'm' | 's';

const sizes = css<{ dimension: Dimension }>`
  width: ${({ dimension }) => {
    switch (dimension) {
      case 's':
        return '16px';
      case 'm':
        return '20px';
      case 'l':
      default:
        return '24px';
    }
  }};
  height: ${({ dimension }) => {
    switch (dimension) {
      case 's':
        return '16px';
      case 'm':
        return '20px';
      case 'l':
      default:
        return '24px';
    }
  }};
`;

const focusStyle = css`
  svg > * {
    fill: ${({ theme }) => theme.color.basic.hover};
  }

  &:before {
    content: '';
    position: absolute;
    top: -4px;
    left: -4px;
    bottom: -4px;
    right: -4px;
    border: 2px solid ${({ theme }) => theme.color.basic.hover};
    border-radius: 4px;
  }
`;

const ButtonComponent = styled.button<{ dimension: Dimension; menuOpened: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  position: relative;
  ${sizes}
  cursor: pointer;
  border: 0;
  outline: none;
  background: transparent;
  -webkit-tap-highlight-color: transparent;

  & svg {
    ${sizes}
    flex-shrink: 0;
  }

  & svg > * {
    fill: ${({ theme }) => theme.color.basic.tertiary};
  }

  &:hover svg > * {
    fill: ${({ theme }) => theme.color.basic.hover};
  }

  ${({ menuOpened }) => menuOpened && focusStyle}
  &:focus {
    ${focusStyle}
  }

  &:active svg > * {
    fill: ${({ theme }) => theme.color.basic.press};
  }

  &:disabled {
    cursor: default;

    & svg > * {
      fill: ${({ theme }) => theme.color.text.tertiary};
    }
  }
`;

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Элементы содержимого */
  children?: ReactNode;
  /** Размер кнопки */
  dimension?: Dimension;
  /** Состояние меню */
  menuOpened: boolean;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ dimension = 'l', menuOpened, type = 'button', ...props }, ref) => {
    return (
      <ButtonComponent ref={ref} dimension={dimension} menuOpened={menuOpened} type={type} {...props}>
        {dimension === 'l' ? (
          <ButtonL width={18} aria-hidden />
        ) : dimension === 'm' ? (
          <ButtonM width={14} aria-hidden />
        ) : (
          <ButtonS width={12} aria-hidden />
        )}
      </ButtonComponent>
    );
  },
);
