// @vtb-ermo/authorization/lib/Authorization.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthProvider = exports.AuthContext = exports._devAuthHeaders = void 0;
var react_1 = __importStar(require("react"));
var utils_1 = require("./utils");
var jsonwebtoken_1 = require("jsonwebtoken");
var constants_1 = require("./constants");
exports._devAuthHeaders = {};
exports.AuthContext = react_1.createContext({});
var AuthProvider = function (props) {
    var _a = props.endpoint, endpoint = _a === void 0 ? '' : _a, devMode = props.devMode;
    var _b = __read(react_1.useState(false), 2), isAuthorized = _b[0], setAuthorized = _b[1];
    var _c = __read(react_1.useState(false), 2), isInit = _c[0], setIsInit = _c[1];
    var _d = __read(react_1.useState({}), 2), user = _d[0], setUser = _d[1];
    var _e = __read(react_1.useState(localStorage.getItem('userId')), 2), userId = _e[0], setUserId = _e[1];
    var _f = __read(react_1.useState([]), 2), statuses = _f[0], setStatuses = _f[1];
    var permissionCodes = react_1.useMemo(function () {
        var _a;
        var codes = [];
        (_a = user === null || user === void 0 ? void 0 : user.roles) === null || _a === void 0 ? void 0 : _a.forEach(function (r) { var _a; return (_a = r === null || r === void 0 ? void 0 : r.permissions) === null || _a === void 0 ? void 0 : _a.forEach(function (p) { return codes.push(p.code); }); });
        return utils_1.unique(codes);
    }, [user === null || user === void 0 ? void 0 : user.roles]);
    var roleCodes = react_1.useMemo(function () {
        var _a;
        return ((_a = user === null || user === void 0 ? void 0 : user.roles) === null || _a === void 0 ? void 0 : _a.map(function (r) { return r.code; })) || [];
    }, [user === null || user === void 0 ? void 0 : user.roles]);
    var availableStatuses = react_1.useMemo(function () {
        return statuses.map(function (status) {
            var _a;
            return ({
                code: status,
                available: !!((_a = constants_1.userStatusDependencies[user.status]) === null || _a === void 0 ? void 0 : _a.includes(status)),
            });
        });
    }, [statuses, user.status]);
    var checkPermission = react_1.useCallback(function (requiredPermissions) {
        return utils_1.isIntersection(permissionCodes, requiredPermissions);
    }, [permissionCodes]);
    var updateUserStatus = react_1.useCallback(function (userStatus) {
        if (!user || !userStatus)
            return Promise.reject(new Error('User unknown'));
        return fetch(endpoint + "/users/" + user.id + "/status", {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({ status: userStatus }),
        }).then(function (response) {
            if (response.ok)
                setUser(function (prevState) { return prevState && __assign(__assign({}, prevState), { status: userStatus }); });
            else
                return Promise.reject(new Error('Bad request'));
        });
    }, [user]);
    var _changeUser = react_1.useCallback(function (userId) {
        localStorage.setItem('userId', userId.toString());
        setUserId(userId);
    }, []);
    react_1.useLayoutEffect(function () {
        if (userId && devMode) {
            var token = jsonwebtoken_1.sign({ sub: userId }, 'your-256-bit-secret', { algorithm: 'HS256' });
            Object.assign(exports._devAuthHeaders, { Authorization: "Bearer " + token });
        }
        else {
            delete exports._devAuthHeaders['Authorization'];
        }
    }, [userId]);
    react_1.useEffect(function () {
        var requestUser = function () { return __awaiter(void 0, void 0, void 0, function () {
            var response, _a, err_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 5, , 6]);
                        return [4 /*yield*/, fetch(endpoint + "/users/current", { headers: exports._devAuthHeaders })];
                    case 1:
                        response = _b.sent();
                        if (!response.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, response.json()];
                    case 2:
                        _a = _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        _a = Promise.reject(response.status + " (" + response.statusText + ")");
                        _b.label = 4;
                    case 4: return [2 /*return*/, _a];
                    case 5:
                        err_1 = _b.sent();
                        return [2 /*return*/, Promise.reject(err_1)];
                    case 6: return [2 /*return*/];
                }
            });
        }); };
        var requestStatuses = function () { return __awaiter(void 0, void 0, void 0, function () {
            var response, _a, err_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 5, , 6]);
                        return [4 /*yield*/, fetch(endpoint + "/users/statuses", { headers: exports._devAuthHeaders })];
                    case 1:
                        response = _b.sent();
                        if (!response.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, response.json()];
                    case 2:
                        _a = _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        _a = Promise.reject(response.status + " (" + response.statusText + ")");
                        _b.label = 4;
                    case 4: return [2 /*return*/, _a];
                    case 5:
                        err_2 = _b.sent();
                        return [2 /*return*/, Promise.reject(err_2)];
                    case 6: return [2 /*return*/];
                }
            });
        }); };
        Promise.all([requestUser(), requestStatuses()])
            .then(function (response) {
            var _a = __read(response, 2), user = _a[0], statuses = _a[1];
            if (user && utils_1.isValidUser(user) && statuses.length) {
                setUser(user);
                setStatuses(statuses);
                setAuthorized(true);
            }
            else {
                setAuthorized(false);
            }
        })
            .catch(function (err) { return console.error(err); })
            .finally(function () { return setIsInit(true); });
        var timerId = setInterval(function () {
            requestUser().then(function (response) {
                if (user && utils_1.isValidUser(user)) {
                    setUser(response);
                }
            });
        }, constants_1.userUpdateIntervalSec * 1000);
        return function () { return clearInterval(timerId); };
    }, [userId]);
    return (react_1.default.createElement(exports.AuthContext.Provider, { value: {
            checkPermission: checkPermission,
            isAuthorized: isAuthorized,
            isInit: isInit,
            availableStatuses: availableStatuses,
            permissionCodes: permissionCodes,
            roleCodes: roleCodes,
            updateUserStatus: updateUserStatus,
            user: user,
            _changeUser: _changeUser,
            _devAuthHeaders: exports._devAuthHeaders,
        } }, props.children));
};
exports.AuthProvider = AuthProvider;
