// @vtb-ermo/authorization/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserStatusCode = exports.PermissionCode = exports.RoleCode = void 0;
var RoleCode;
(function (RoleCode) {
    RoleCode["OPS_USER_EXECUTOR"] = "OPS_USER_EXECUTOR";
    RoleCode["OPS_USER_SUPERVISOR"] = "OPS_USER_SUPERVISOR";
    RoleCode["OPS_USER_ANALYST_PROCESS"] = "OPS_USER_ANALYST_PROCESS";
    RoleCode["OPS_USER_MANAGER_PROCESS"] = "OPS_USER_MANAGER_PROCESS";
})(RoleCode = exports.RoleCode || (exports.RoleCode = {}));
var PermissionCode;
(function (PermissionCode) {
    PermissionCode["GET_MY_TEAM_INFO"] = "GET_MY_TEAM_INFO";
    PermissionCode["GET_WORKLOAD_AND_CAPACITY_INFO"] = "GET_WORKLOAD_AND_CAPACITY_INFO";
    PermissionCode["GET_PERSONAL_TASKS"] = "GET_PERSONAL_TASKS";
    PermissionCode["GET_COMPLETED_PERSONAL_TASKS"] = "GET_COMPLETED_PERSONAL_TASKS";
    PermissionCode["START_WORK_TASK"] = "START_WORK_TASK";
    PermissionCode["RETURN_TASK_IN_POOL"] = "RETURN_TASK_IN_POOL";
    PermissionCode["OPEN_TASK"] = "OPEN_TASK";
    PermissionCode["COMPLETE_TASK"] = "COMPLETE_TASK";
    PermissionCode["REJECT_TASK"] = "REJECT_TASK";
    PermissionCode["GET_NOTIFICATIONS"] = "GET_NOTIFICATIONS";
    PermissionCode["GET_TASK_IN_POOL"] = "GET_TASK_IN_POOL";
    PermissionCode["GET_EXECUTORS_TRANSFER_TASK"] = "GET_EXECUTORS_TRANSFER_TASK";
    PermissionCode["TRANSFER_TASK_EXECUTION"] = "TRANSFER_TASK_EXECUTION";
    PermissionCode["GET_COMPLETED_TASKS"] = "GET_COMPLETED_TASKS";
    PermissionCode["CHANGE_STATUS"] = "CHANGE_STATUS";
    PermissionCode["GET_TASK_TYPES"] = "GET_TASK_TYPES";
    PermissionCode["CREATE_TASK_TYPE"] = "CREATE_TASK_TYPE";
    PermissionCode["APPROVE_TASK_TYPE"] = "APPROVE_TASK_TYPE";
    PermissionCode["DEACTIVATE_TASK_TYPE"] = "DEACTIVATE_TASK_TYPE";
    PermissionCode["GET_SPECIFICATION_TASK_TYPE"] = "GET_SPECIFICATION_TASK_TYPE";
    PermissionCode["CREATE_SPECIFICATION_TASK_TYPE"] = "CREATE_SPECIFICATION_TASK_TYPE";
    PermissionCode["GET_RELEVANCE_EXECUTORS_TASK_TYPE"] = "GET_RELEVANCE_EXECUTORS_TASK_TYPE";
    PermissionCode["TUNE_RELEVANCE_EXECUTORS_TASK_TYPE"] = "TUNE_RELEVANCE_EXECUTORS_TASK_TYPE";
    PermissionCode["TUNE_NOTIFICATIONS"] = "TUNE_NOTIFICATIONS";
    PermissionCode["GET_CONFIG_NOTIFICATIONS"] = "GET_CONFIG_NOTIFICATIONS";
})(PermissionCode = exports.PermissionCode || (exports.PermissionCode = {}));
var UserStatusCode;
(function (UserStatusCode) {
    UserStatusCode["TECHNICAL_BREAK"] = "TECHNICAL_BREAK";
    UserStatusCode["PERFORMING_THE_TASK_OF_THE_SUPERVISOR"] = "PERFORMING_THE_TASK_OF_THE_SUPERVISOR";
    UserStatusCode["ACTIVE"] = "ACTIVE";
    UserStatusCode["NOT_ACTIVE"] = "NOT_ACTIVE";
})(UserStatusCode = exports.UserStatusCode || (exports.UserStatusCode = {}));
