// @openvtb/react-ui-kit/src/ProductOrCardSelectItem/Image.tsx
import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { IMAGE_MARGIN, IMAGE_PARAMS } from './constants';
import type { ItemType } from './constants';

type PropsType = {
  $type: ItemType;
  $image?: string;
};

export const Image = styled.div<PropsType>`
  ${({ theme, $type, $image }) => css`
    display: inline-block;
    position: relative;
    overflow: visible;
    width: ${IMAGE_PARAMS[$type].width}px;
    min-width: ${IMAGE_PARAMS[$type].width}px;
    height: ${IMAGE_PARAMS[$type].height}px;
    min-height: ${IMAGE_PARAMS[$type].height}px;
    margin: ${IMAGE_MARGIN};
    border-radius: ${IMAGE_PARAMS[$type].radius};
    & > * {
      border-radius: ${IMAGE_PARAMS[$type].radius};
    }
    background-color: ${theme.color.neutral[20]};
    background-size: cover;
    background-position: center center;
    background-repeat: no-repeat;
    ${$image ? `background-image: url(${$image});` : ''}
  `}
  & > * {
    height: 100%;
    width: 100%;
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
    top: 0;
  }
`;

Image.defaultProps = {
  theme: DEFAULT_THEME,
};
