// @vtb-ermo/notifications/src/components/NotificationList/styled.ts
import styled from 'styled-components';
import { NotificationItemContainer } from '../NotificationItem/styled';

export const NotificationListContainer = styled.div`
    position: absolute;
    top: 68px;
    right: 32px;
    width: 488px;
    padding: 4px;
    z-index: 100;
    background-color: ${({ theme }) => theme.color.background.primary};
    box-shadow: ${({ theme }) => {
        return `0 4px 8px ${theme.color.background.tertiaryHover}, 0 -2px 4px ${theme.color.background.tertiaryHover}`;
    }};
    border-radius: 4px;
    :before {
        content: '';
        position: absolute;
        left: 80%;
        top: -10px;
        width: 20px;
        height: 20px;
        background-color: ${({ theme }) => theme.color.background.primary};
        box-shadow: -2px -2px 2px ${({ theme }) => theme.color.background.tertiaryHover};
        transform: rotate(45deg);
    }
    h1 {
        margin: 0 0 24px 0;
        padding: 0;
        font-style: normal;
        font-weight: 500;
        font-size: 22px;
        line-height: 28px;
    }
    h2 {
        margin: 0 0 12px 0;
        padding: 0;
        font-style: normal;
        font-weight: 500;
        font-size: 16px;
        line-height: 24px;
        color: ${({ theme }) => theme.color.text.secondary};
    }
    ${NotificationItemContainer} {
        width: 460px;
        margin-bottom: 24px;
        margin-left: -14px;
        padding: unset;
        border: none;
        box-shadow: none;
        background-color: transparent;
    }
`;

export const ScrollWrapper = styled.div`
    padding: 20px 32px 0 20px;
    max-height: calc(100vh - 128px);
    overflow-y: auto;
    :before {
        position: absolute;
        content: '';
        top: 4px;
        left: 0;
        height: 24px;
        width: calc(100% - 20px);
        background-image: ${({ theme }) => {
            return `linear-gradient(to bottom, ${theme.color.background.primary}, transparent)`;
        }};
        z-index: 100;
    }
    :after {
        position: absolute;
        content: '';
        bottom: 4px;
        left: 0;
        height: 24px;
        width: calc(100% - 20px);
        background-image: ${({ theme }) => {
            return `linear-gradient(to top, ${theme.color.background.primary}, transparent)`;
        }};
        z-index: 100;
    }
`;

export const EmptyNotification = styled.div`
    margin-bottom: 28px;
    color: ${({ theme }) => theme.color.text.tertiary};
`;

export const ViewedWrapper = styled.div`
    ${NotificationItemContainer} {
        opacity: 0.5;
    }
`;
