// @vtb/ui-kit3/src/components/OverflowMenu/Button.tsx
import type { ButtonHTMLAttributes, ReactNode } from 'react';
import * as React from 'react';
import styled, { css } from 'styled-components';

import { ReactComponent as MoreHorizontalOutline } from '@openvtb/admiral-icons/build/system/MoreHorizontalOutline.svg';

export type Dimension = 'l' | 'm' | 's';

const ICON_SIZE = {
  l: 24,
  m: 20,
  s: 16,
};

const focusStyle = css`
  & *[fill^='#'] {
    fill: ${({ theme }) => theme.color.basic.hover};
  }

  &:before {
    content: '';
    position: absolute;
    top: -4px;
    left: -4px;
    bottom: -4px;
    right: -4px;
    border: 2px solid ${({ theme }) => theme.color.basic.hover};
    border-radius: 4px;
  }
`;

const ButtonComponent = styled.button<{ dimension: Dimension; menuOpened: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
  border: 0;
  outline: none;
  padding: 0;
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  overflow: visible;

  & svg {
    flex-shrink: 0;
  }

  &:hover:not(:disabled) *[fill^='#'] {
    fill: ${({ theme }) => theme.color.basic.hover};
  }

  ${({ menuOpened }) => menuOpened && focusStyle}
  &:focus {
    ${focusStyle}
  }

  &:active *[fill^='#'] {
    fill: ${({ theme }) => theme.color.basic.press};
  }

  &:disabled {
    cursor: default;

    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.text.tertiary};
    }
  }
`;

export const ButtonContent = styled.span<{ $isVertical?: boolean; dimension: Dimension }>`
  position: relative;
  display: inline-flex;
  align-items: center;
  height: 100%;

  & > svg {
    transform: rotate(${(p) => (p.$isVertical ? 90 : 0)}deg);
    height: ${({ dimension }) => ICON_SIZE[dimension]}px;
    width: ${({ dimension }) => ICON_SIZE[dimension]}px;

    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.text.secondary};
    }
  }
`;

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Элементы содержимого */
  children?: ReactNode;
  /** Размер кнопки */
  dimension?: Dimension;
  /** Состояние меню */
  menuOpened: boolean;
  /** Ориентация компонента */
  isVertical?: boolean;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ dimension = 'l', menuOpened, type = 'button', isVertical = false, ...props }, ref) => {
    return (
      <ButtonComponent ref={ref} dimension={dimension} menuOpened={menuOpened} type={type} {...props}>
        <ButtonContent dimension={dimension} $isVertical={isVertical}>
          <MoreHorizontalOutline aria-hidden />
        </ButtonContent>
      </ButtonComponent>
    );
  },
);
