// @openvtb/react-ui-kit/src/Select/styled-components/MenuList.ts
import styled from 'styled-components';

import { DEFAULT_THEME } from '../../common';
import { MENU_MAX_HEIGHT, MENU_VERTICAL_PADDING, Size } from '../constants';

type PropsType = {
  size: Size;
};

export const MenuList = styled.div<PropsType>`
  -webkit-overflow-scrolling: touch;
  box-sizing: border-box;
  position: relative;
  max-height: ${({ size }) => `${MENU_MAX_HEIGHT[size]} !important`};
  padding: ${MENU_VERTICAL_PADDING} 0;
  overflow: hidden;
  overflow-y: auto;
  border-radius: ${({ theme }) => theme.borderRadius};

  ::-webkit-scrollbar {
    width: 16px;
  }

  ::-webkit-scrollbar-track {
    background: ${({ theme }) => theme.color.neutral.white};
    background-clip: content-box;
  }

  ::-webkit-scrollbar-thumb {
    background: ${({ theme }) => theme.color.neutral[30]};
    border: 6px solid ${({ theme }) => theme.color.neutral.white};
    border-radius: 10px;
  }

  ::-webkit-scrollbar-thumb:hover {
    border: 5px solid ${({ theme }) => theme.color.neutral.white};
  }

  &:focus {
    outline: none;
  }
`;

MenuList.defaultProps = {
  theme: DEFAULT_THEME,
};
