// @vtb/ui-kit3/src/components/MenuButton/MenuButton.stories.tsx
import React from 'react';
import { Meta, Story } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import { ReactComponent as MinusCircleOutline } from '@openvtb/admiral-icons/build/service/MinusCircleOutline.svg';

import { MenuButton, MenuButtonItem, MenuButtonProps } from '../MenuButton';

export default {
  title: 'Example/MenuButton',
  decorators: [withDesign],
  component: MenuButton,
  parameters: {
    layout: 'centered',
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=67455%3A86885',
    },
  },
  argTypes: {
    dimension: {
      options: ['xl', 'l', 'm', 's'],
      control: { type: 'radio' },
    },
    appearance: {
      options: ['primary', 'secondary'],
      control: { type: 'radio' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    loading: {
      control: { type: 'boolean' },
    },
    skeleton: {
      control: { type: 'boolean' },
    },
  },
} as Meta;

const Template1: Story<MenuButtonProps> = (args) => {
  const items: Array<MenuButtonItem> = [
    {
      id: '1',
      display: 'Option one',
    },
    {
      id: '2',
      display: 'Option two',
    },
    {
      id: '3',
      display: 'Option three',
    },
    {
      id: '4',
      display: 'Option four',
    },
    {
      id: '5',
      display: 'Option five',
    },
    {
      id: '6',
      display: 'Option six',
    },
    {
      id: '7',
      display: 'Option seven',
    },
  ];

  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <MenuButton
      {...args}
      selected={selected}
      onChange={(id) => setSelected(id)}
      options={items}
      onOpen={() => console.log('open menu')}
      onClose={() => console.log('close menu')}
    >
      test
    </MenuButton>
  );
};

const Template2: Story<MenuButtonProps> = (args) => {
  const items: Array<MenuButtonItem> = [
    {
      id: '1',
      display: 'Option one',
    },
    {
      id: '2',
      display: (
        <div style={{ display: 'flex', width: '115px', justifyContent: 'space-between', alignItems: 'center' }}>
          <MinusCircleOutline width={20} height={20} />
          Option two
        </div>
      ),
    },
    {
      id: '3',
      display: 'Option three',
      disabled: true,
    },
    {
      id: '4',
      display: 'Option four',
    },
  ];

  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <MenuButton
      {...args}
      selected={selected}
      onChange={(id) => setSelected(id)}
      options={items}
      onOpen={() => console.log('open menu')}
      onClose={() => console.log('close menu')}
    >
      test
    </MenuButton>
  );
};

export const MenuButtonBase = Template1.bind({});
MenuButtonBase.args = {};
MenuButtonBase.storyName = 'MenuButton. Базовый пример.';
MenuButtonBase.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};

export const MenuButtonOptions = Template2.bind({});
MenuButtonOptions.args = {};
MenuButtonOptions.storyName = 'MenuButton. Задизейбленные и кастомизированные опции.';
MenuButtonOptions.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};
