// @vtb/ui-kit3/src/components/Avatar/Avatar.stories.tsx
import { withDesign } from 'storybook-addon-designs';
import React from 'react';
import { Meta, Story } from '@storybook/react';
import styled from 'styled-components';
import { Avatar } from '~/components/Avatar';
import { typography } from '~/components';

const imageURL = 'https://github.com/DrUNE.png?size=100';

const Desc = styled.div`
  ${typography['Additional/L']}
`;

const StoryContainer = styled.div`
  background-color: #e5e5e5;
  padding: 24px;
  position: relative;
  display: flex;

  > * {
    margin: 8px 40px 0 0;
  }
`;

export default {
  title: 'Example/Avatar',
  decorators: [withDesign],
  component: Avatar,
  parameters: {
    componentSubtitle: <Desc>Avatar - компонент отображающий аватарку пользователя и его статус</Desc>,
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=79337%3A133927',
    },
  },
} as Meta;

const Template: Story = (args) => (
  <StoryContainer>
    <Avatar href={imageURL} width={24} height={24} />
    <Avatar href={imageURL} width={32} height={32} />
    <Avatar href={imageURL} width={40} height={40} />
    <Avatar href={imageURL} width={48} height={48} />
    <Avatar href={imageURL} />
  </StoryContainer>
);
export const Playground = Template.bind({});
Playground.args = {};
Playground.storyName = 'Playground';
