// @vtb-ermo/webpack-config/index.js
'use strict';

const ModuleFederationPlugin = require('webpack/lib/container/ModuleFederationPlugin');
const HTMLWebpackTagsPlugin = require('html-webpack-tags-plugin');

module.exports = {
    createHostFederationConfig: function (hostAppConfig) {
        const remotesApps = {};
        const remotesScripts = [];
        Object.keys(hostAppConfig.remotes).forEach(key => {
            remotesApps[key] = `${key}`;
            remotesScripts.push({
                path: hostAppConfig.remotes[key].entry,
                publicPath: false,
            })
        });
        return {
            module: {
                rules: [
                    {
                        test: /bootstrap\.tsx$/,
                        loader: 'bundle-loader',
                        options: {
                            lazy: true,
                        },
                    },
                ],
            },
            plugins: [
                new ModuleFederationPlugin({
                    name: hostAppConfig.name,
                    library: { type: 'var', name: hostAppConfig.name },
                    remotes: remotesApps,
                    shared: {
                        'react': {
                            eager: true,
                            requiredVersion: '16.14.0',
                            singleton: true,
                        },
                        'react-dom': {
                            eager: true,
                            requiredVersion: '16.14.0',
                            singleton: true,
                        },
                        '@vtb/ui-kit3': {
                            eager: true,
                            requiredVersion: '0.0.35',
                            singleton: true,
                        },
                        '@openvtb/react-ui-kit': {
                            eager: true,
                            requiredVersion: '0.5.1',
                            singleton: true,
                        },
                        '@vtb-ermo/router': {
                            eager: true,
                            requiredVersion: '1.21.0',
                            singleton: true,
                        },
                        '@vtb-ermo/authorization': {
                            eager: true,
                            requiredVersion: '1.0.0',
                            singleton: true,
                        },
                        '@vtb-ermo/notifications': {
                            eager: true,
                            requiredVersion: '0.0.8',
                            singleton: true,
                        },
                        'styled-components': {
                            eager: true,
                            requiredVersion: '5.2.3',
                            singleton: true,
                        },
                    },
                }),
                new HTMLWebpackTagsPlugin({
                    append: false,
                    scripts: remotesScripts,
                }),
            ],
        };
    },
    createRemoteFederationConfig: function (remoteAppConfig) {
        return {
            output: {
                publicPath: 'auto', // Should be 'auto' on the remote host
            },
            devServer: {
                headers: {
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
                    'Access-Control-Allow-Headers': 'X-Requested-With, content-type, Authorization',
                },
            },
            plugins: [
                new ModuleFederationPlugin({
                    name: remoteAppConfig.name,
                    filename: remoteAppConfig.filename,
                    exposes: remoteAppConfig.exposes,
                    shared: {
                        'react': {
                            eager: false,
                            requiredVersion: '16.14.0',
                            singleton: true,
                        },
                        'react-dom': {
                            eager: false,
                            requiredVersion: '16.14.0',
                            singleton: true,
                        },
                        '@openvtb/react-ui-kit': {
                            eager: false,
                            requiredVersion: '0.5.1',
                            singleton: true,
                        },
                        '@vtb/ui-kit3': {
                            eager: false,
                            requiredVersion: '0.0.35',
                            singleton: true,
                        },
                        '@vtb-ermo/router': {
                            eager: false,
                            requiredVersion: '1.21.0',
                            singleton: true,
                        },
                        '@vtb-ermo/authorization': {
                            eager: false,
                            requiredVersion: '1.0.0',
                            singleton: true,
                        },
                        '@vtb-ermo/notifications': {
                            eager: false,
                            requiredVersion: '0.0.8',
                            singleton: true,
                        },
                        'styled-components': {
                            eager: true,
                            requiredVersion: '5.2.3',
                            singleton: true,
                        },
                    },
                }),
            ],
        };
    }
};
