// @vtb/ui-kit3/src/components/form/CheckboxField/CheckboxField.stories.tsx
import React, { useEffect, useState } from 'react';
import { Meta, Story } from '@storybook/react';
import { CheckboxField } from '~/components/form/CheckboxField/index';

import { withDesign } from 'storybook-addon-designs';
import { ALL_DIMENSIONS_VALUES } from '~/components/Checkbox/CheckboxDimension';
import styled from 'styled-components';

export default {
  title: 'Form Field Components/CheckboxField',
  decorators: [withDesign],
  component: CheckboxField,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=64585%3A231',
    },
  },
  argTypes: {
    dimension: {
      options: ALL_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },
    indeterminate: {
      options: [false, true],
      control: { type: 'radio' },
    },
    checked: {
      options: [false, true],
      control: { type: 'radio' },
    },
    disabled: {
      options: [undefined, true],
      control: { type: 'radio' },
    },
  },
} as Meta;

const Container = styled.div`
  display: flex;
  flex-direction: column;

  > * {
    margin-top: 16px;
  }
`;
const CheckboxFieldDemo: Story<React.ComponentProps<typeof CheckboxField>> = (props) => {
  const args = (Object.keys(props) as Array<keyof typeof props>).reduce((acc, key) => {
    if (props[key] !== undefined) acc[key] = props[key];

    return acc;
  }, {} as Record<any, any>);

  const [checked, setChecked] = useState<boolean>(args.checked ?? false);
  useEffect(() => {
    setChecked(Boolean(args.checked));
  }, [args.checked]);
  return (
    <Container>
      <CheckboxField
        {...args}
        checked={checked}
        onChange={(e) => {
          setChecked(e.target.checked);
        }}
      >
        Управляемый чекбокс
      </CheckboxField>
      <CheckboxField dimension="s">Не управляемый маленький чекбокс</CheckboxField>
      <CheckboxField disabled defaultChecked>
        Disabled не управляемый чекбокс
      </CheckboxField>
      <CheckboxField
        error
        extraText={
          <span>
            Вариация с <i>дополнительно</i> декорированным <b>текстом</b>
          </span>
        }
      >
        Error не управляемый чекбокс
      </CheckboxField>
      <CheckboxField dimension="s" indeterminate extraText="Вариация с дополнительным текстом">
        Не управляемый маленький чекбокс indeterminate
      </CheckboxField>
    </Container>
  );
};

export const Example = CheckboxFieldDemo.bind({});
