// @vtb-ermo/router/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.routes = exports.useNavigation = exports.Switch = exports.Redirect = exports.Route = exports.NavLink = exports.Link = exports.CoreRouter = exports.withRouter = exports.useRouteMatch = exports.useParams = exports.useLocation = exports.useHistory = void 0;
var routes_1 = require("./routes");
__exportStar(require("react-router-dom"), exports);
var react_router_dom_1 = require("react-router-dom");
Object.defineProperty(exports, "useHistory", { enumerable: true, get: function () { return react_router_dom_1.useHistory; } });
Object.defineProperty(exports, "useLocation", { enumerable: true, get: function () { return react_router_dom_1.useLocation; } });
Object.defineProperty(exports, "useParams", { enumerable: true, get: function () { return react_router_dom_1.useParams; } });
Object.defineProperty(exports, "useRouteMatch", { enumerable: true, get: function () { return react_router_dom_1.useRouteMatch; } });
Object.defineProperty(exports, "withRouter", { enumerable: true, get: function () { return react_router_dom_1.withRouter; } });
var CoreRouter_1 = require("./CoreRouter");
Object.defineProperty(exports, "CoreRouter", { enumerable: true, get: function () { return CoreRouter_1.CoreRouter; } });
var Link_1 = require("./Link");
Object.defineProperty(exports, "Link", { enumerable: true, get: function () { return Link_1.Link; } });
var NavLink_1 = require("./NavLink");
Object.defineProperty(exports, "NavLink", { enumerable: true, get: function () { return NavLink_1.NavLink; } });
var Route_1 = require("./Route");
Object.defineProperty(exports, "Route", { enumerable: true, get: function () { return Route_1.Route; } });
var Redirect_1 = require("./Redirect");
Object.defineProperty(exports, "Redirect", { enumerable: true, get: function () { return Redirect_1.Redirect; } });
var Switch_1 = require("./Switch");
Object.defineProperty(exports, "Switch", { enumerable: true, get: function () { return Switch_1.Switch; } });
var useNavigation_1 = require("./hooks/useNavigation");
Object.defineProperty(exports, "useNavigation", { enumerable: true, get: function () { return useNavigation_1.useNavigation; } });
exports.routes = routes_1.routesConfig;
