// @vtb/ui-kit3/src/components/Link/Link.stories.tsx
import React from 'react';
import styled from 'styled-components';
import { Meta, Story } from '@storybook/react';
import { Link, LinkProps } from './index';
import { Additional_L_normal } from '../Typography';
import { ReactComponent as ArrowLeftOutline } from '@openvtb/admiral-icons/build/system/ArrowLeftOutline.svg';
import { ReactComponent as ArrowRightOutline } from '@openvtb/admiral-icons/build/system/ArrowRightOutline.svg';
import { withDesign } from 'storybook-addon-designs';

export default {
  title: 'Example/Link',
  decorators: [withDesign],
  component: Link,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=63767%3A72254',
    },
  },
  argTypes: {
    target: {
      control: {
        type: 'select',
        options: ['_parent', '_self', '_top', '_blank'],
      },
    },
    appearance: {
      control: {
        type: 'radio',
        options: ['primary', 'secondary'],
      },
    },
    dimension: {
      control: {
        type: 'radio',
        options: ['m', 's'],
      },
    },
  },
} as Meta;

const Separator = styled.div`
  margin-top: 10px;
`;

const Devider = styled.div`
  width: 10px;
`;

const LinkPrimary: Story = () => {
  return (
    <>
      <Additional_L_normal>Dimension - M </Additional_L_normal>
      <Link appearance={'primary'} href="" onClick={(e) => e.preventDefault()}>
        Link
      </Link>
      <Separator />
      <Additional_L_normal>Dimension - S </Additional_L_normal>
      <Link appearance={'primary'} dimension="s" href="" onClick={(e) => e.preventDefault()}>
        Link
      </Link>
    </>
  );
};

const LinkSecondary: Story = () => (
  <>
    <Additional_L_normal>Dimension - M </Additional_L_normal>
    <Link appearance={'secondary'} href="" onClick={(e) => e.preventDefault()}>
      Link
    </Link>
    <Separator />
    <Additional_L_normal>Dimension - S </Additional_L_normal>
    <Link appearance={'secondary'} dimension="s" href="" onClick={(e) => e.preventDefault()}>
      Link
    </Link>
  </>
);

const LinkWithIcon: Story = () => {
  return (
    <>
      <Additional_L_normal>Link with icon. Dimension - m </Additional_L_normal>
      <Separator />
      <Link appearance={'primary'} href="" onClick={(e) => e.preventDefault()}>
        <ArrowLeftOutline width={24} />
        <Devider />
        Link
      </Link>
      <Separator />
      <Additional_L_normal> Link with icon. Dimension - s </Additional_L_normal>
      <Separator />
      <Link appearance={'secondary'} dimension="s" href="" onClick={(e) => e.preventDefault()}>
        Link
        <Devider />
        <ArrowRightOutline width={20} />
      </Link>
    </>
  );
};

const LinkDefault: Story<LinkProps> = ({ children, ...args }) => <Link {...args}>{children}</Link>;
export const Primary = LinkPrimary.bind({});
export const Secondary = LinkSecondary.bind({});
export const IconLink = LinkWithIcon.bind({});
export const Playground = LinkDefault.bind({});

Playground.args = {
  children: 'Link',
  href: 'http://localhost:6006/?path=/story/example-link--playground',
};
