// @vtb/ui-kit3/src/components/Field/Field.stories.tsx
import React from 'react';
import { Meta, Story } from '@storybook/react';
import { Field, TextInput } from '~/components';
import { withDesign } from 'storybook-addon-designs';
import styled from 'styled-components';

export default {
  title: 'Atoms/Field',
  component: Field,
  decorators: [withDesign],
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=65740%3A76866',
    },
  },
} as Meta;
const Container = styled.div`
  > *:not(:first-child) {
    margin-top: 24px;
  }
`;
const Template: Story<React.ComponentProps<typeof Field>> = (props) => {
  return (
    <Container>
      <Field {...props}>
        <TextInput id={props.id} />
      </Field>
      <Field label="Inline label" id="some_id" displayInline>
        <TextInput id="some_id" />
      </Field>
    </Container>
  );
};

export const FieldStory = Template.bind({});

FieldStory.args = {
  label: 'some label of wrapped component',
  id: 'some_id',
};

FieldStory.storyName = 'Пример обертывания компонента TextInput';
