// @openvtb/react-ui-kit/src/Select/DropdownMenu.tsx
import React, { Children, isValidElement, createRef, cloneElement } from 'react';
import type { FC, MutableRefObject, ReactNode, MouseEvent, KeyboardEvent } from 'react';

import { SelectItem } from '../SelectItem';
import { ProductSelectItem, CardSelectItem } from '../ProductOrCardSelectItem';

import { MenuList, MenuWrapper, NoOptionMessage, OptionWrapper } from './styled-components';
import { areEqual } from './utils';
import type { Size, SelectType } from './constants';
import type { ISelectComponentItem } from './types';

type PropsType = {
  menuRef: MutableRefObject<HTMLDivElement | null>;
  menuListRef: MutableRefObject<HTMLDivElement | null>;
  optionsRefs: { [value: string]: any };
  tooltipContainer?: Element | null;
  size: Size;
  value: any;
  type: SelectType;
  opened?: boolean;
  disabled?: boolean;
  width?: string | number;
  list?: ISelectComponentItem[];
  withoutOptionMessage: string;
  onOptionClick: (value: any) => (event: MouseEvent<HTMLDivElement>) => void;
  onKeyDown: (event: KeyboardEvent<HTMLDivElement>) => void;
};

export const DropdownMenu: FC<PropsType> = ({
  list,
  value,
  size,
  type,
  withoutOptionMessage,
  opened,
  disabled,
  width,
  children,
  menuRef,
  menuListRef,
  optionsRefs,
  tooltipContainer,
  onOptionClick,
  onKeyDown,
}) => {
  const renderChildren = () => {
    return Children.map(children, (child: ReactNode) => {
      if (!isValidElement(child)) {
        return null;
      }

      const selected = areEqual(value, child.props.value);
      const onClick =
        child.props.disabled || disabled ? undefined : onOptionClick(child.props.value);

      const ref = createRef<HTMLDivElement>();
      optionsRefs.current[child.props.value] = ref;
      return (
        <OptionWrapper
          data-value={child.props.value}
          tabIndex={-1}
          ref={ref}
          size={size}
          disabled={disabled || child.props.disabled}
          $type={type}
          selected={selected}
          onClick={onClick}
        >
          {cloneElement(child, {
            value: undefined,
            tooltipPortalContainer: tooltipContainer,
            tooltipTargetContainer: ref,
            size: size,
          })}
        </OptionWrapper>
      );
    });
  };
  const renderListItems = (list: ISelectComponentItem[]) => {
    return list.map((item: any, index: number) => {
      const {
        label,
        value: childValue,
        icon: Icon,
        disabled: childDisabled,
        currency,
        note,
        image,
      } = item;
      const selected = areEqual(value, childValue);
      const onClick = childDisabled || disabled ? undefined : onOptionClick(childValue);

      const ref = createRef<HTMLDivElement>();
      optionsRefs.current[childValue] = ref;

      return (
        <OptionWrapper
          key={`${value}_${index}`}
          data-value={childValue}
          tabIndex={-1}
          ref={ref}
          size={size}
          disabled={disabled || childDisabled}
          $type={type}
          selected={selected}
          onClick={onClick}
        >
          {type === 'default' && (
            <SelectItem
              value={undefined}
              disabled={childDisabled}
              tooltipPortalContainer={tooltipContainer}
              tooltipTargetContainer={ref}
              size={size}
            >
              {Icon && <Icon />}
              {label}
            </SelectItem>
          )}
          {type === 'card' && (
            <CardSelectItem
              value={undefined}
              disabled={childDisabled}
              note={note}
              image={image}
              currency={currency}
              tooltipPortalContainer={tooltipContainer}
              tooltipTargetContainer={ref}
              size={size}
            >
              {label}
            </CardSelectItem>
          )}
          {type === 'product' && (
            <ProductSelectItem
              value={undefined}
              disabled={childDisabled}
              note={note}
              image={image}
              currency={currency}
              tooltipPortalContainer={tooltipContainer}
              tooltipTargetContainer={ref}
              size={size}
            >
              {label}
            </ProductSelectItem>
          )}
        </OptionWrapper>
      );
    });
  };
  return (
    <MenuWrapper ref={menuRef} onKeyDown={onKeyDown} opened={opened} width={width}>
      <MenuList ref={menuListRef} size={size} tabIndex={-1}>
        {children && !list && renderChildren()}
        {list && renderListItems(list)}
        {!children && !list && (
          <NoOptionMessage size={size} $type={type}>
            {withoutOptionMessage}
          </NoOptionMessage>
        )}
      </MenuList>
    </MenuWrapper>
  );
};
