// @openvtb/react-ui-kit/src/Select/Value.tsx
import React, { isValidElement, cloneElement } from 'react';
import type { FC, ReactNode, MutableRefObject } from 'react';

import { ValueWrapper, Value as SingleValue, Placeholder } from './styled-components';
import type { Size } from './constants';

type PropsType = {
  selectedItem: ReactNode;
  placeholder: string;
  size: Size;
  tooltipContainer?: Element | null;
  inputRef: MutableRefObject<HTMLDivElement | null>;
  disabled?: boolean;
  focused?: boolean;
};

export const Value: FC<PropsType> = ({
  selectedItem,
  placeholder,
  tooltipContainer,
  inputRef,
  disabled,
  focused,
  size,
}) => (
  <ValueWrapper>
    {selectedItem ? (
      <SingleValue disabled={disabled}>
        {isValidElement(selectedItem) &&
          cloneElement(selectedItem, {
            tooltipPortalContainer: tooltipContainer,
            tooltipTargetContainer: inputRef,
            size,
          })}
      </SingleValue>
    ) : (
      <Placeholder disabled={disabled} focused={focused}>
        {placeholder}
      </Placeholder>
    )}
  </ValueWrapper>
);
