// @vtb/ui-kit3/src/components/form/SelectField/SelectField.stories.tsx
import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react';
import { SelectField, SelectFieldProps } from './';
import { withDesign } from 'storybook-addon-designs';
import { DropDownItem } from '~/components/DropDownItem';
import styled from 'styled-components';
import { INPUT_DIMENSIONS_VALUES } from '~/components';

export default {
  title: 'Form Field Components/ SelectField',
  component: SelectField,
  decorators: [withDesign],
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=65740%3A76866',
    },
  },
  argTypes: {
    dimension: {
      options: INPUT_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },
  },
} as Meta;

const DisplayContainer = styled.div`
  > * {
    margin-bottom: 24px;
  }
`;
const Template: Story<SelectFieldProps> = (props) => {
  const [selectValue, setValue] = useState('');
  const options = [
    { id: 1, label: 'Option one', value: '1' },
    { id: 2, label: 'Option two', value: '2' },
    { id: 3, label: 'Option three', value: '3' },
    { id: 4, label: 'Option four', value: '4' },
    { id: 5, label: 'Option five', value: '5' },
    { id: 6, label: 'Option six', value: '6' },
    { id: 7, label: 'Option seven', value: '7' },
  ];

  return (
    <DisplayContainer>
      <SelectField dimension={props.dimension} label="Опции" value="123" onChange={() => console.log('hello!')}>
        {options.map((item) => {
          return (
            <DropDownItem role="option" value={item.label} key={item.id}>
              {item.label}
            </DropDownItem>
          );
        })}
      </SelectField>
    </DisplayContainer>
  );
};

export const SelectFieldFieldInput = Template.bind({});

SelectFieldFieldInput.args = {
  value: 'Привет',
  label: 'Label',
};

SelectFieldFieldInput.storyName = 'uncontrolled';
