// @vtb/ui-kit3/src/components/form/InputField/InputField.stories.tsx
import * as React from 'react';
import { Meta, Story } from '@storybook/react';
import { InputField, InputFieldProps } from '~/components/form/InputField';
import { withDesign } from 'storybook-addon-designs';
import { INPUT_DIMENSIONS_VALUES, Spinner } from '~/components';
import styled from 'styled-components';

export default {
  title: 'Form Field Components/InputField',
  component: InputField,
  decorators: [withDesign],
  parameters: {
    design: {
      type: 'figma',
      url:
        'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=65740%3A76866',
    },
  },
  argTypes: {
    dimension: {
      options: INPUT_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },
  },
} as Meta;

const DisplayContainer = styled.div`
  > * {
    margin-bottom: 24px;
  }
`;
const Template: Story<InputFieldProps> = (props) => {
  const [localValue, setValue] = React.useState<string>(String(props.value) ?? '');

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.currentTarget.value;
    setValue(inputValue);
    props.onChange?.(e);
  };

  return (
    <DisplayContainer>
      <InputField {...props} value={localValue} onChange={handleChange} />
      <InputField required label="Поле необходимо заполнить" />
      <InputField readOnly aria-readonly label="Поле с атрибутом readOnly" defaultValue="Этот текст не редактируемый" />
      <InputField disabled aria-disabled label="Поле с атрибутом disabled" placeholder="Placeholder" />
      <InputField
        status="error"
        label="Поле с ошибкой"
        extraText="Поле не прошло валидацию, необходимо ввести корректное значение"
      />
      <InputField
        status="success"
        displayStatusIcon
        label="Поле с индикацией успеха"
        extraText="Поле не прошло валидацию, необходимо ввести корректное значение"
      />
      <InputField
        displayClearIcon
        placeholder="идет поиск ..."
        label="Поле с иконкой загрузки"
        icons={<Spinner dimension="s" />}
      />
      <InputField label="Поле для ввода пароля (type='password')" type="password" />
    </DisplayContainer>
  );
};

export const InputFieldInput = Template.bind({});

InputFieldInput.args = {
  value: 'Привет',
  label: 'Label',
};

InputFieldInput.storyName = 'Input field example';
