// @vtb/ui-kit3/src/components/Calendar/Calendar.stories.tsx
import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';

import { Calendar, CalendarPropType } from './index';

export default {
  title: 'Example/Calendar',
  decorators: [withDesign],
  component: Calendar,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=61882%3A72131',
    },
    docs: {
      source: {
        type: 'code',
      },
    },
  },
  argTypes: {
    localeName: {
      options: ['ru', 'enUS', 'de'],
      control: { type: 'radio' },
    },
    range: {
      control: { type: 'boolean' },
    },
  },
} as Meta;

const Template1: Story<CalendarPropType> = (args: CalendarPropType) => {
  const [selected, setSelected] = useState<Date | null>(null);
  return (
    <Calendar
      {...args}
      selected={selected}
      onChange={(value: any) => {
        setSelected(value);
      }}
    />
  );
};

const Template2: Story<CalendarPropType> = (args: CalendarPropType) => {
  const [startDate, setStartDate] = useState<Date | null>(null);
  const [endDate, setEndDate] = useState<Date | null>(null);
  return (
    <Calendar
      {...args}
      range
      startDate={startDate}
      endDate={endDate}
      onChange={(value: any) => {
        setStartDate(value[0]);
        setEndDate(value[1]);
      }}
    />
  );
};

const Template3: Story<CalendarPropType> = (args: CalendarPropType) => {
  const [selected, setSelected] = useState<Date | null>(null);
  const tomorrow = new Date();
  tomorrow.setDate(new Date().getDate() + 1);
  return (
    <Calendar
      {...args}
      maxDate={tomorrow}
      selected={selected}
      onChange={(value: any) => {
        setSelected(value);
      }}
    />
  );
};

const Template4: Story<CalendarPropType> = (args: CalendarPropType) => {
  const [selected, setSelected] = useState<Date | null>(null);
  const isWeekday = (date: Date) => {
    const day = date.getDay();
    return day !== 0 && day !== 6;
  };
  return (
    <Calendar
      {...args}
      filterDate={isWeekday}
      selected={selected}
      onChange={(value: any) => {
        setSelected(value);
      }}
    />
  );
};

export const CalendarSimple = Template1.bind({});
CalendarSimple.args = {};
CalendarSimple.storyName = 'Calendar. Simple.';

export const Range = Template2.bind({});
Range.args = {};
Range.storyName = 'Calendar. Range.';

export const SimpleWithMaxDate = Template3.bind({});
SimpleWithMaxDate.args = {};
SimpleWithMaxDate.storyName = 'Calendar. Simple with maxDate.';

export const SimpleWithFilterDate = Template4.bind({});
SimpleWithFilterDate.args = {};
SimpleWithFilterDate.storyName = 'Calendar. Simple with filterDate.';
