// @vtb/ui-kit3/src/components/Tooltip/Tooltip.stories.tsx
import React, { useRef } from 'react';
import { ReactComponent as DeleteOutline } from '@openvtb/admiral-icons/build/system/DeleteOutline.svg';
import { Meta, Story } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import styled from 'styled-components';

import { ITooltipProps, Tooltip } from '../Tooltip';
import { Button } from '../Button';

const Separator = styled.div<{ height?: number }>`
  height: ${({ height }) => (height ? height : 20)}px;
`;

const Desc = styled.div`
  font-family: 'VTB Group UI';
  font-size: 16px;
  line-height: 24px;
`;

const Description = () => (
  <Desc>
    Тултип — компонент отвечающий за подсказки, детализацию информации или пояснения. Появляется при наведении на
    элемент к которому он привязан сразу, либо с задержкой в 1.5 секунды (опционально). Позиционируется по центру
    объекта, сверху, снизу, слева или справа с отступом 8px.
    <Separator />
    Рекомендуется максимальная ширина 488px. При большем объеме используйте компонент Hint.
    <Separator />
    Приоритетным является расположение тултипа снизу объекта. Далее по приоритетности идут расположения сверху, слева и
    справа.
  </Desc>
);

export default {
  title: 'Example/Tooltip',
  decorators: [withDesign],
  component: Tooltip,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=63639%3A71167',
    },
    componentSubtitle: <Description />,
    layout: 'centered',
    docs: {
      source: {
        type: 'code',
      },
    },
  },
  argTypes: {
    withDelay: {
      control: { type: 'boolean' },
    },
  },
} as Meta;

const Template1: Story<ITooltipProps> = ({ withDelay }: ITooltipProps) => {
  return (
    <Tooltip renderContent={() => 'Delete file'} withDelay={withDelay} id="test1">
      <Button dimension="m" displayAsSquare aria-label="Delete" aria-describedby="test1">
        <DeleteOutline height={24} width={24} aria-hidden />
      </Button>
    </Tooltip>
  );
};

const Template2: Story<ITooltipProps> = () => {
  return (
    <Tooltip renderContent={() => 'Delete file'} withDelay id="test2">
      <Button dimension="m" displayAsSquare aria-label="Delete" aria-describedby="test2">
        <DeleteOutline height={24} width={24} />
      </Button>
    </Tooltip>
  );
};

const Template3: Story<ITooltipProps> = () => {
  const tooltipRef = useRef(null);
  return (
    <Tooltip renderContent={() => 'Delete file'} tooltipRef={tooltipRef} id="test3">
      <Button dimension="m" displayAsSquare aria-label="Delete" aria-describedby="test3">
        <DeleteOutline height={24} width={24} />
      </Button>
    </Tooltip>
  );
};

export const TooltipBase = Template1.bind({});
TooltipBase.args = {};
TooltipBase.storyName = 'Tooltip. Базовый пример.';

export const TooltipDelay = Template2.bind({});
TooltipDelay.args = {};
TooltipDelay.storyName = 'Tooltip. Пример с задержкой в появлении.';

export const TooltipRef = Template3.bind({});
TooltipRef.args = {};
TooltipRef.storyName = 'Tooltip. Пример с получением ref тултипа.';
