// @vtb-ermo/router/src/routes/rmvk/routes.ts
export const rmvkRoutes = {
    RMVK: {
        params: {
            relativePath: 'rmvk',
        },
        ERM_CTI: {
            params: {
                relativePath: 'erm-cti',
            },
        },
        ERM_BCMA: {
            params: {
                relativePath: 'erm-bcma',
            },
        },
        ERM_VKTR: {
            params: {
                relativePath: 'erm-vktr',
            },
        },
        ERM_EC_DOCS_REQUEST: {
            params: {
                relativePath: 'erm-ec-docs-request',
            },
        },
        ERM_CURRENCY_CONTRACT: {
            params: {
                relativePath: 'erm-currency-contract',
            },
        },
        ERM_CRR: {
            params: {
                relativePath: 'erm-crr',
            },
        },
        ERM_DOCUMENTS_REFERENCE: {
            params: {
                relativePath: 'erm-documents-reference',
            },
        },
        ERM_EC_LETTER_FROM_BANK: {
            params: {
                relativePath: 'erm-ec-letter-from-bank',
            },
        },
        ERM_EC_LETTER_TO_BANK: {
            params: {
                relativePath: 'erm-ec-letter-to-bank',
            },
        },
    },
};
