// @vtb/ui-kit3/src/components/form/FieldSet/FieldSet.stories.tsx
import * as React from 'react';
import { Meta, Story } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import styled from 'styled-components';

import { RadioButton as Radio } from '~/components/RadioButton';
import { FieldSet, PropsType as FieldSetPropsType } from '~/components/form/FieldSet/index';
import { CheckboxField } from '~/components';

const Desc = styled.div`
  font-family: 'VTB Group UI';
  font-size: 16px;
  line-height: 24px;
`;

const Description = () => (
  <Desc>
    Радиальные кнопки применяются, когда есть список опций, из которых пользователь может выбрать только один вариант
  </Desc>
);

export default {
  title: 'Form Field Components/FieldSet',
  decorators: [withDesign],
  component: FieldSet,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=64810%3A2',
    },
    componentSubtitle: <Description />,
    actions: {
      argTypesRegex: 'onChange',
    },
    layout: 'centered',
  },
  argTypes: {
    dimension: {
      options: ['m', 's'],
      control: { type: 'radio' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    label: {
      control: { type: 'text' },
    },
    name: {
      control: { type: 'text' },
    },
  },
} as Meta;
const Container = styled.div`
  > * {
    margin-top: 24px;
  }
`;
const Template1: Story<FieldSetPropsType> = (props: FieldSetPropsType) => {
  return (
    <Container>
      <FieldSet legend={'Группа чекбоксов:'}>
        <CheckboxField name="check1">Text 1</CheckboxField>
        <CheckboxField name="check2">Text 2</CheckboxField>
        <CheckboxField name="check3">Text 3</CheckboxField>
      </FieldSet>
      <FieldSet disabled legend={'Disabled группа чекбоксов:'}>
        <CheckboxField name="check1">Text 1</CheckboxField>
        <CheckboxField name="check2" defaultChecked>
          Text 2
        </CheckboxField>
        <CheckboxField name="check3">Text 3</CheckboxField>
      </FieldSet>
      <FieldSet dimension="s" legend={'Группа маленьких чекбоксов:'}>
        <CheckboxField name="check1">Text 1</CheckboxField>
        <CheckboxField name="check2">Text 2</CheckboxField>
        <CheckboxField name="check3">Text 3</CheckboxField>
      </FieldSet>
    </Container>
  );
};

const Template2: Story<FieldSetPropsType> = ({ disabled, dimension }: FieldSetPropsType) => {
  const values = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15'];
  const [selected, setSelected] = React.useState('');
  return (
    <Container>
      <FieldSet
        style={{ display: 'flex', flexDirection: 'column' }}
        onChange={(e) => {
          setSelected((e.target as HTMLInputElement).value);
        }}
        legend="Управляемая группа радиокнопок"
        disabled={disabled}
        dimension={dimension}
      >
        <Radio value={values[0]} name="test1" checked={values[0] === selected} readOnly>
          First option
        </Radio>
        <Radio value={values[1]} name="test1" checked={values[1] === selected} readOnly>
          Second option
        </Radio>
        <Radio value={values[2]} name="test1" checked={values[2] === selected} readOnly>
          Third option
        </Radio>
      </FieldSet>
      <FieldSet
        style={{ display: 'flex', flexDirection: 'column' }}
        legend="Неуправляемая группа радиокнопок"
        disabled={disabled}
        dimension={dimension}
      >
        <Radio value={values[3]} name="test2">
          First option
        </Radio>
        <Radio value={values[4]} name="test2">
          Second option
        </Radio>
        <Radio value={values[5]} name="test2">
          Third option
        </Radio>
      </FieldSet>
      <FieldSet style={{ display: 'flex', flexDirection: 'column' }} legend="Disabled группа радиокнопок" disabled>
        <Radio value={values[6]} name="test3">
          First option
        </Radio>
        <Radio value={values[7]} name="test3">
          Second option
        </Radio>
        <Radio value={values[8]} name="test3">
          Third option
        </Radio>
      </FieldSet>
      <FieldSet
        style={{ display: 'flex', flexDirection: 'column' }}
        legend="Группа радиокнопок с одной задизейбленной кнопкой"
      >
        <Radio value={values[9]} name="test4">
          First option
        </Radio>
        <Radio value={values[10]} disabled name="test4">
          Second option
        </Radio>
        <Radio value={values[11]} name="test4">
          Third option
        </Radio>
      </FieldSet>
      <FieldSet
        style={{ display: 'flex', flexDirection: 'column' }}
        legend="Группа радиокнопок маленького размера"
        dimension="s"
      >
        <Radio value={values[12]} name="test5">
          First option
        </Radio>
        <Radio value={values[13]} name="test5">
          Second option
        </Radio>
        <Radio value={values[14]} name="test5">
          Third option
        </Radio>
      </FieldSet>
    </Container>
  );
};

export const Checkbox = Template1.bind({});
Checkbox.args = {};
Checkbox.storyName = 'Checkbox Fieldset';
Checkbox.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};

export const RadioGroup = Template2.bind({});
RadioGroup.args = {};
RadioGroup.storyName = 'RadioGroup Fieldset';
RadioGroup.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};
