// @vtb/ui-kit3/src/components/MenuButton/MenuButton.stories.tsx
import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import { ReactComponent as MinusCircleOutline } from '@openvtb/admiral-icons/build/service/MinusCircleOutline.svg';
import styled from 'styled-components';
import { T } from '#ui-kit3/components/T';

import { MenuButton, MenuButtonItem, MenuButtonProps } from '../MenuButton';

const WrapperVertical = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  gap: 20px;
`;
const Wrapper = styled.div`
  display: flex;
`;
const Separator = styled.div`
  height: 20px;
  width: 20px;
`;

export default {
  title: 'Example/MenuButton',
  decorators: [withDesign],
  component: MenuButton,
  parameters: {
    layout: 'centered',
    design: [
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=108393%3A69948',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24698',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24648',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24734',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24784',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24834',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24673',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24759',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24809',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=39%3A24859',
      },
    ],
  },
  argTypes: {
    dimension: {
      options: ['xl', 'l', 'm', 's'],
      control: { type: 'radio' },
    },
    appearance: {
      options: ['primary', 'secondary', 'ghost'],
      control: { type: 'radio' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    loading: {
      control: { type: 'boolean' },
    },
    skeleton: {
      control: { type: 'boolean' },
    },
  },
} as ComponentMeta<typeof MenuButton>;

const items = [
  {
    id: '1',
    display: 'Option one',
  },
  {
    id: '2',
    display: 'Option two',
  },
  {
    id: '3',
    display: 'Option three',
  },
  {
    id: '4',
    display: 'Option four',
  },
  {
    id: '5',
    display: 'Option five',
    disabled: true,
  },
  {
    id: '6',
    display: 'Option six',
  },
  {
    id: '7',
    display: 'Option seven',
  },
] as MenuButtonItem[];

const Template1: ComponentStory<typeof MenuButton> = (args) => {
  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <MenuButton
      {...args}
      selected={selected}
      onChange={(id) => {
        console.log(`selected: ${id}`);
        setSelected(id);
      }}
      options={items}
      onOpen={() => console.log('open menu')}
      onClose={() => console.log('close menu')}
    >
      test
    </MenuButton>
  );
};

const Template2: ComponentStory<typeof MenuButton> = (args) => {
  const items: Array<MenuButtonItem> = [
    {
      id: '1',
      display: 'Option one',
    },
    {
      id: '2',
      display: (
        <div style={{ display: 'flex', width: '115px', justifyContent: 'space-between', alignItems: 'center' }}>
          <MinusCircleOutline width={20} height={20} />
          Option two
        </div>
      ),
      disabled: true,
    },
    {
      id: '3',
      display: 'Option three',
    },
    {
      id: '4',
      display: 'Option four',
    },
  ];

  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <MenuButton
      {...args}
      selected={selected}
      onChange={(id) => {
        console.log(`selected: ${id}`);
        setSelected(id);
      }}
      options={items}
      onOpen={() => console.log('open menu')}
      onClose={() => console.log('close menu')}
    >
      test
    </MenuButton>
  );
};

const Template3: ComponentStory<typeof MenuButton> = (args) => {
  const items: Array<MenuButtonItem> = [
    {
      id: '1',
      display: 'Option one',
    },
    {
      id: '2',
      display: (
        <div style={{ display: 'flex', width: '115px', justifyContent: 'space-between', alignItems: 'center' }}>
          <MinusCircleOutline width={20} height={20} />
          Option two
        </div>
      ),
      disabled: true,
    },
    {
      id: '3',
      display: 'Option three',
    },
    {
      id: '4',
      display: 'Option four',
    },
  ];

  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <>
      <Wrapper>
        <WrapperVertical>
          <T font="Body/Body 1 Long" as="div">
            Size XL
          </T>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="xl"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="xl"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="xl"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="xl"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="xl"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="xl"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
        </WrapperVertical>
        <Separator />
        <WrapperVertical>
          <T font="Body/Body 1 Long" as="div">
            Size L
          </T>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="l"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="l"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="l"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="l"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="l"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="l"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
        </WrapperVertical>
        <Separator />
        <WrapperVertical>
          <T font="Body/Body 1 Long" as="div">
            Size M
          </T>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="m"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="m"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="m"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="m"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="m"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="m"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
        </WrapperVertical>
        <Separator />
        <WrapperVertical>
          <T font="Body/Body 1 Long" as="div">
            Size S
          </T>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="s"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="primary"
            dimension="s"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="s"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="secondary"
            dimension="s"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="s"
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
          <MenuButton
            {...args}
            appearance="ghost"
            dimension="s"
            loading
            selected={selected}
            onChange={(id) => {
              console.log(`selected: ${id}`);
              setSelected(id);
            }}
            options={items}
            onOpen={() => console.log('open menu')}
            onClose={() => console.log('close menu')}
          >
            Button 56
          </MenuButton>
        </WrapperVertical>
      </Wrapper>
    </>
  );
};

export const MenuButtonBase = Template1.bind({});
MenuButtonBase.args = {};
MenuButtonBase.storyName = 'MenuButton. Базовый пример.';
MenuButtonBase.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};

export const MenuButtonOptions = Template2.bind({});
MenuButtonOptions.args = {};
MenuButtonOptions.storyName = 'MenuButton. Задизейбленные и кастомизированные опции.';
MenuButtonOptions.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};

export const MenuButtonVariants = Template3.bind({});
MenuButtonVariants.args = {};
MenuButtonVariants.storyName = 'MenuButton. Размеры и стили.';
MenuButtonVariants.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};
