// @vtb/ui-kit3/src/components/Checkbox/Checkbox.stories.tsx
import React, { useEffect, useState } from 'react';
import { Meta, Story } from '@storybook/react';
import { Checkbox } from './index';

import { withDesign } from 'storybook-addon-designs';
import { ALL_DIMENSIONS_VALUES } from '~/components/Checkbox/CheckboxDimension';
import styled from 'styled-components';

export default {
  title: 'Atoms/Checkbox',
  decorators: [withDesign],
  component: Checkbox,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=64585%3A231',
    },
  },
  argTypes: {
    dimension: {
      options: ALL_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },
    indeterminate: {
      options: [false, true],
      control: { type: 'radio' },
    },
    checked: {
      options: [false, true],
      control: { type: 'radio' },
    },
    disabled: {
      options: [undefined, true],
      control: { type: 'radio' },
    },
  },
} as Meta;
const Container = styled.div`
  display: flex;
  flex-direction: column;

  > * {
    margin-top: 16px;
  }
`;
const CheckboxDemo: Story<React.ComponentProps<typeof Checkbox>> = (props) => {
  const args = (Object.keys(props) as Array<keyof typeof props>).reduce((acc, key) => {
    if (props[key] !== undefined) acc[key] = props[key];

    return acc;
  }, {} as Record<any, any>);

  const [checked, setChecked] = useState<boolean>(args.checked ?? false);
  useEffect(() => {
    setChecked(Boolean(args.checked));
  }, [args.checked]);
  return (
    <Container>
      <Checkbox
        {...args}
        checked={checked}
        onChange={(e) => {
          setChecked(e.target.checked);
        }}
      />
      <Checkbox dimension="s" />
      <Checkbox disabled defaultChecked />
      <Checkbox error />
      <Checkbox dimension="s" indeterminate />
    </Container>
  );
};

export const Example = CheckboxDemo.bind({});
