// @vtb-ermo/notifications/lib/NotificationsContext.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationsProvider = exports.NotificationsContext = void 0;
var react_1 = __importStar(require("react"));
exports.NotificationsContext = react_1.createContext({});
var NotificationsProvider = function (_a) {
    var children = _a.children;
    var _b = __read(react_1.useState([]), 2), notifications = _b[0], setNotifications = _b[1];
    var publish = react_1.useCallback(function (item) {
        setNotifications(function (prevState) {
            var _a = item.memorable, memorable = _a === void 0 ? true : _a, _b = item.notifiable, notifiable = _b === void 0 ? true : _b;
            return prevState.some(function (n) { return n.id === item.id; })
                ? prevState
                : __spreadArray(__spreadArray([], __read(prevState)), [__assign(__assign({}, item), { alert: notifiable, new: true, memorable: memorable })]);
        });
    }, []);
    var unPublish = react_1.useCallback(function (id) {
        setNotifications(function (prevState) { return prevState.filter(function (i) { return i.id !== id; }); });
    }, []);
    return (react_1.default.createElement(exports.NotificationsContext.Provider, { value: {
            notifications: notifications,
            setNotifications: setNotifications,
            publish: publish,
            unPublish: unPublish,
        } }, children));
};
exports.NotificationsProvider = NotificationsProvider;
