// @vtb/ui-kit3/src/components/input/DateInput/DateInput.stories.tsx
import React, { ChangeEvent, useState } from 'react';
import { Meta, Story } from '@storybook/react';
import { DateInput } from '../DateInput';
import { withDesign } from 'storybook-addon-designs';
import { INPUT_DIMENSIONS_VALUES } from '~/components/input/inputTypes';

export default {
  title: 'Input/DateInput',
  component: DateInput,
  decorators: [withDesign],
  parameters: {
    design: {
      type: 'figma',
      url:
        'https://www.figma.com/file/nwmAE2QMjzDD1FECyC2EUL/01-%E2%9C%85-UI-Kit-2.1-Internet-bank?node-id=65740%3A76866',
    },
  },
  argTypes: {
    dimension: {
      options: INPUT_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },
    defaultIsCalendarOpen: { type: 'boolean' },

    onChange: {
      action: 'onChange',
    },
  },
} as Meta;

const Template: Story<React.ComponentProps<typeof DateInput>> = (props) => {
  const cleanProps = (Object.keys(props) as Array<keyof typeof props>).reduce((acc, key) => {
    if (props[key] !== undefined) acc[key] = props[key];

    return acc;
  }, {} as Record<any, any>);

  const [localValue, setValue] = useState<string>(String(props.value) ?? '');
  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.currentTarget.value;
    setValue(inputValue);
    props.onChange?.(e);
  };
  return <DateInput {...cleanProps} value={localValue} onChange={handleChange} />;
};

export const DateInputStory = Template.bind({});

DateInputStory.args = {};
DateInputStory.argTypes = { defaultIsCalendarOpen: { type: 'boolean' } };
DateInputStory.storyName = 'DateInput (input type="date")';
