// @openvtb/react-ui-kit/src/OverflowMenu/styled-components/IndicatorsWrapper.ts
import styled from 'styled-components';

import { DEFAULT_THEME } from '../../common';

interface IIndicatorsWrapperProps {
  menuIsOpen?: boolean;
  size: string;
  disabled?: boolean;
}

const wrapperWidth = {
  big: 36,
  small: 32,
  micro: 28,
};

export const IndicatorsWrapper = styled.div<IIndicatorsWrapperProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  outline: none;
  ::before {
    content: '';
    position: absolute;
    width: ${({ size }) => wrapperWidth[size]}px;
    height: ${({ size }) => wrapperWidth[size]}px;
    outline: none;
    border-radius: 50%;
    pointer-events: none;
    background: ${({ menuIsOpen, theme }) => (menuIsOpen ? theme.color.opacity.blackFocus : '')};
  }
  &:hover::before {
    background-color: ${({ menuIsOpen, theme, disabled }) =>
      menuIsOpen || disabled ? '' : theme.color.opacity.blackHover};
  }
  &:active::before {
    background-color: ${({ theme, disabled }) =>
      disabled ? '' : theme.color.opacity.blackPressed};
  }
`;

IndicatorsWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
