// @openvtb/react-ui-kit/src/OverflowMenu/styled-components/OptionWrapper.ts
import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../common';
import { OPTION_HEIGHT, OPTION_PADDING } from '../constants';
import type { Size } from '../constants';

type PropsType = {
  size: Size;
  disabled?: boolean;
  selected?: boolean;
};

export const OptionWrapper = styled.div<PropsType>`
  display: flex;
  align-items: center;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  user-select: none;
  box-sizing: border-box;
  ${({ theme, size, disabled }) => css`
    height: ${OPTION_HEIGHT[size]};
    padding: ${OPTION_PADDING[size]};
    cursor: ${disabled ? 'default' : 'pointer'};
    background-color: ${theme.color.neutral.white};
    color: ${theme.color.neutral[90]};
    & svg {
      fill: ${theme.color.neutral[50]};
    }
  `}

  ${({ selected, theme }) =>
    selected &&
    css`
      color: ${theme.color.primary[60]};
      & svg {
        fill: ${theme.color.primary[60]};
      }
    `}

  ${({ theme, disabled }) =>
    disabled
      ? css`
          color: ${theme.color.neutral[30]};
          & svg {
            fill: ${theme.color.neutral[30]};
          }
        `
      : css`
          &:focus,
          &:hover {
            color: ${theme.color.primary[60]};
            background-color: ${theme.color.neutral[5]};
            & svg {
              fill: ${theme.color.primary[60]};
            }
          }
        `}

  &:focus {
    outline: none;
  }
  & > div {
    overflow: hidden;
  }
`;

OptionWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
