// @openvtb/react-ui-kit/src/Inputs/Mask/DropdownIndicator.tsx
import React, { FC, Dispatch, MouseEvent } from 'react';

import { ReactComponent as ChevronDownOutline } from '../../Icons/system/ChevronDownOutline.svg';
import { ReactComponent as ChevronUpOutline } from '../../Icons/system/ChevronUpOutline.svg';
import type { Size } from '../common/types';

import { StyledDropdownIndication, IconFlag } from './StyledComponents';

interface IDropdownIndicatorProps {
  setOpened?: Dispatch<React.SetStateAction<boolean>>;
  opened?: boolean;
  size: Size;
  phoneCode: string;
  setFocused: Dispatch<React.SetStateAction<boolean>>;
  disabled?: boolean;
}

export const DropdownIndicator: FC<IDropdownIndicatorProps> = ({
  opened,
  setOpened,
  size,
  phoneCode,
  children,
  setFocused,
  disabled,
}) => {
  const Icon = opened ? ChevronUpOutline : ChevronDownOutline;
  const iconSize = size === 'micro' ? 20 : 24;
  const withPadding = phoneCode !== '+7 (';
  const handleIconClick = (event: MouseEvent) => {
    event?.preventDefault();
    setOpened && setOpened(!opened);
    setFocused(true);
  };
  return (
    <StyledDropdownIndication onMouseDown={(e) => e.preventDefault()} disabled={disabled}>
      <IconFlag disabled={disabled} withPadding={withPadding}>
        {children}
      </IconFlag>

      <Icon width={iconSize} height={iconSize} onMouseDown={handleIconClick} />
    </StyledDropdownIndication>
  );
};
